{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Description:\n",
    "\n",
    "* This notebook goes through the analysis of blasting spacers vs a database containing potential protospacers.\n",
    "* Any blast datbase can be used.\n",
    "* A script is provided to easily blast all of the genomes in CLdb."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Before running this notebook:\n",
    "\n",
    "* __run the [Setup](./Setup.pynb) notebook__"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# User-defined variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# directory where you want the spacer blasting to be done\n",
    "## CHANGE THIS!\n",
    "workDir = \"/home/nyoungb2/t/CLdb_Methanosarcina/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Init"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The rpy2.ipython extension is already loaded. To reload it, use:\n",
      "  %reload_ext rpy2.ipython\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "from IPython.display import FileLinks\n",
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "blastDir = os.path.join(workDir, 'arrayBlast')\n",
    "if not os.path.isdir(blastDir):\n",
    "    os.makedirs(blastDir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Blast vs all E. coli genomes in the CLdb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Spacer blast"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Selecting spacer sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique spacers:\t1104\n",
      "\n",
      ">NA|spacer|NA|628|1\n",
      "TTTTACCGGGGATGTGACAGTCAGGGAATACGGCAT\n",
      ">NA|spacer|NA|804|1\n",
      "TACCCAACTGTTCAAATTTCAAAGCCACTACGTGA\n",
      ">NA|spacer|NA|923|1\n",
      "TACTGTTCGGCGGTTCTTCGCATTGTTGCGTCAAT\n"
     ]
    }
   ],
   "source": [
    "!cd $blastDir; \\\n",
    "    CLdb -- array2fasta -cluster -cut 1 > spacers_cut1.fna\n",
    "\n",
    "## getting number of spacer sequences\n",
    "!printf 'Number of unique spacers:\\t'\n",
    "!cd $blastDir; \\\n",
    "    grep -c \">\" spacers_cut1.fna; \\\n",
    "    echo; \\\n",
    "    head -n 6 spacers_cut1.fna"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Blast\n",
    "\n",
    "* The array blast commands are sub-sub commands.\n",
    "* __SO__, that's why when you type: '`CLdb -- arrayBlast -h`', you get then following:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage:\r\n",
      "    arrayBlast [options] subcommand -- [subcommand_options]\r\n",
      "\r\n",
      "  Options:\r\n",
      "    --list\r\n",
      "        List all subcommands.\r\n",
      "\r\n",
      "    --perldoc\r\n",
      "        Get perldoc of subcommand.\r\n",
      "\r\n",
      "    -v Verbose output\r\n",
      "    -h This help message\r\n",
      "\r\n",
      "  For more information:\r\n",
      "    CLdb --perldoc -- arrayBlast\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!CLdb -- arrayBlast -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "add_crRNA\r\n",
      "add_proto\r\n",
      "align_proto\r\n",
      "filter_arrayHits\r\n",
      "get_PAM\r\n",
      "get_SEEDstats\r\n",
      "get_align\r\n",
      "get_proto\r\n",
      "run\r\n",
      "srl2csv\r\n",
      "srl2json\r\n",
      "xml2json\r\n",
      "xml2srl\r\n"
     ]
    }
   ],
   "source": [
    "# listing all arrayBlast subcommands\n",
    "!CLdb -- arrayBlast --list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Sub-sub-command help:\n",
    "\n",
    "`CLdb arrayBlast -- run -- -h`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Making BLAST databases...\n",
      "...making blast databases in /home/nyoungb2/t/CLdb_Methanosarcina//arrayBlast/blast_db/\n",
      "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/t/CLdb_Methanosarcina//arrayBlast/blast_db//Methanosarcina_acetivorans_C2A.fasta\n",
      "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/t/CLdb_Methanosarcina//arrayBlast/blast_db//Methanosarcina_barkeri_str_fusaro.fasta\n",
      "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/t/CLdb_Methanosarcina//arrayBlast/blast_db//Methanosarcina_horonobensis_HB_1.fasta\n",
      "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/t/CLdb_Methanosarcina//arrayBlast/blast_db//Methanosarcina_mazei_C16.fasta\n",
      "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/t/CLdb_Methanosarcina//arrayBlast/blast_db//Methanosarcina_mazei_Go1.fasta\n",
      "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/t/CLdb_Methanosarcina//arrayBlast/blast_db//Methanosarcina_mazei_LYC.fasta\n",
      "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/t/CLdb_Methanosarcina//arrayBlast/blast_db//Methanosarcina_mazei_S_6.fasta\n",
      "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/t/CLdb_Methanosarcina//arrayBlast/blast_db//Methanosarcina_mazei_SarPi.fasta\n",
      "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/t/CLdb_Methanosarcina//arrayBlast/blast_db//Methanosarcina_mazei_WWM610.fasta\n",
      "\n",
      "Blasting sequences...\n",
      "Number of blast hits to Methanosarcina_mazei_Go1.fasta:\t28622\n",
      "Number of blast hits to Methanosarcina_mazei_S_6.fasta:\t27769\n",
      "Number of blast hits to Methanosarcina_mazei_SarPi.fasta:\t33535\n",
      "Number of blast hits to Methanosarcina_mazei_C16.fasta:\t29628\n",
      "Number of blast hits to Methanosarcina_mazei_LYC.fasta:\t31244\n",
      "Number of blast hits to Methanosarcina_mazei_WWM610.fasta:\t28519\n",
      "Number of blast hits to Methanosarcina_barkeri_str_fusaro.fasta:\t25491\n",
      "Number of blast hits to Methanosarcina_horonobensis_HB_1.fasta:\t28834\n",
      "Number of blast hits to Methanosarcina_acetivorans_C2A.fasta:\t24641\n"
     ]
    }
   ],
   "source": [
    "# spacer blast\n",
    "## Really just a thin wrapper around blastn that will blast the spacers against all genomes in CLdb\n",
    "### If you want to blast some else, just run the blast yourself (use 'perldoc CLdb_arrayBlast.pl' for more info)\n",
    "\n",
    "!cd $blastDir; \\\n",
    "    CLdb -- arrayBlast -- run \\\n",
    "    -query spacers_cut1.fna \\\n",
    "    -fork 10 \\\n",
    "    > spacers_cut1_blastn.xml\n",
    "    \n",
    "# NOTE: using 10 CPUs    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note:**\n",
    "\n",
    "* Using spacer clustering cutoff of 1, so just 'unique' spacer sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-rw-r-- 1 nyoungb2 nyoungb2 2.5M Jan  3 20:21 spacers_cut1_blastn.srl\r\n",
      "-rw-rw-r-- 1 nyoungb2 nyoungb2 9.6M Jan  3 20:20 spacers_cut1_blastn.xml\r\n"
     ]
    }
   ],
   "source": [
    "# converting .xml to .srl\n",
    "## This retains all of the info in the xml, but produces a smaller file that is faster to load and write.\n",
    "## For large blast runs (xml files > 100 Mb), this will take a little while\n",
    "\n",
    "!cd $blastDir; \\\n",
    "    CLdb -- arrayBlast -- xml2srl \\\n",
    "    spacers_cut1_blastn.xml \\\n",
    "    > spacers_cut1_blastn.srl\n",
    "\n",
    "# checking output file\n",
    "!cd $blastDir; \\\n",
    "    ls -thlc spacers_cut1_blastn.*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note:**\n",
    "\n",
    "* You could pipe the output from `CLdb -- arrayBlast -- run` directly into `CLdb -- arrayBlast -- xml2srl`\n",
    "  * Example: `CLdb -- arrayBlast -- run -query spacers_cut1.fna | CLdb -- arrayBlast -- xml2srl`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DR Blast\n",
    "\n",
    "* BLASTing DR sequences against the same subjects.\n",
    "* This is needed to filter out spacer blast hits to CRISPR arrays."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique DRs:\t38\n",
      "\n",
      ">NA|DR|NA|5|1\n",
      "GTTTCAATCCTTGTTTTAGTGGATCTTGCTCTCGAAT\n",
      ">NA|DR|NA|20|1\n",
      "ATTCGTGAGCATGATCCATTAAGAGCCTATTCGAAAA\n",
      ">NA|DR|NA|27|1\n",
      "GCTTCAATTCTGCCACAACCTTTCGGTTATGGAAAC\n"
     ]
    }
   ],
   "source": [
    "# selecting DRs, so we can blast them.\n",
    "## This is needed blast filter out spacers that hit CRISPR arrays\n",
    "\n",
    "!cd $blastDir; \\\n",
    "    CLdb -- array2fasta \\\n",
    "    -cluster -cut 1 -r \\\n",
    "    > DRs_cut1.fna\n",
    "\n",
    "## getting number of spacer sequences\n",
    "!printf 'Number of unique DRs:\\t'\n",
    "!cd $blastDir; \\\n",
    "    grep -c \">\" DRs_cut1.fna; \\\n",
    "    echo; \\\n",
    "    head -n 6 DRs_cut1.fna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Making BLAST databases...\n",
      "...making blast databases in /home/nyoungb2/t/CLdb_Methanosarcina//arrayBlast/blast_db/\n",
      "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/t/CLdb_Methanosarcina//arrayBlast/blast_db//Methanosarcina_acetivorans_C2A.fasta\n",
      "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/t/CLdb_Methanosarcina//arrayBlast/blast_db//Methanosarcina_barkeri_str_fusaro.fasta\n",
      "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/t/CLdb_Methanosarcina//arrayBlast/blast_db//Methanosarcina_horonobensis_HB_1.fasta\n",
      "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/t/CLdb_Methanosarcina//arrayBlast/blast_db//Methanosarcina_mazei_C16.fasta\n",
      "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/t/CLdb_Methanosarcina//arrayBlast/blast_db//Methanosarcina_mazei_Go1.fasta\n",
      "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/t/CLdb_Methanosarcina//arrayBlast/blast_db//Methanosarcina_mazei_LYC.fasta\n",
      "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/t/CLdb_Methanosarcina//arrayBlast/blast_db//Methanosarcina_mazei_S_6.fasta\n",
      "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/t/CLdb_Methanosarcina//arrayBlast/blast_db//Methanosarcina_mazei_SarPi.fasta\n",
      "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/t/CLdb_Methanosarcina//arrayBlast/blast_db//Methanosarcina_mazei_WWM610.fasta\n",
      "\n",
      "Blasting sequences...\n",
      "Number of blast hits to Methanosarcina_barkeri_str_fusaro.fasta:\t17138\n",
      "Number of blast hits to Methanosarcina_acetivorans_C2A.fasta:\t21463\n",
      "Number of blast hits to Methanosarcina_mazei_S_6.fasta:\t32573\n",
      "Number of blast hits to Methanosarcina_horonobensis_HB_1.fasta:\t40840\n",
      "Number of blast hits to Methanosarcina_mazei_Go1.fasta:\t47480\n",
      "Number of blast hits to Methanosarcina_mazei_C16.fasta:\t64637\n",
      "Number of blast hits to Methanosarcina_mazei_WWM610.fasta:\t47727\n",
      "Number of blast hits to Methanosarcina_mazei_LYC.fasta:\t79088\n",
      "Number of blast hits to Methanosarcina_mazei_SarPi.fasta:\t86164\n"
     ]
    }
   ],
   "source": [
    "# DR blast, just like spacer blast\n",
    "\n",
    "!cd $blastDir; \\\n",
    "    CLdb -- arrayBlast -- run \\\n",
    "    -query DRs_cut1.fna \\\n",
    "    -fork 10 \\\n",
    "    > DRs_cut1_blastn.xml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-rw-r-- 1 nyoungb2 nyoungb2 5.4M Jan  3 20:23 DRs_cut1_blastn.srl\r\n",
      "-rw-rw-r-- 1 nyoungb2 nyoungb2  17M Jan  3 20:22 DRs_cut1_blastn.xml\r\n"
     ]
    }
   ],
   "source": [
    "# converting .xml to .srl\n",
    "## This retains all of the info in the xml, but produces a smaller file that is faster to load and write.\n",
    "## For large blast runs (xml files > 100 Mb), this will take a little while\n",
    "\n",
    "!cd $blastDir; \\\n",
    "    CLdb -- arrayBlast -- xml2srl \\\n",
    "    DRs_cut1_blastn.xml \\\n",
    "    > DRs_cut1_blastn.srl\n",
    "\n",
    "\n",
    "# checking output file\n",
    "!cd $blastDir; \\\n",
    "    ls -thlc DRs_cut1_blastn.*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filtering out spacer hits to CRISPRs\n",
    "\n",
    "* Based on DR hits falling adjacent to spacer hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "### Summary for filtering of 'bad' DR hits ###\n",
      "Total DR hits: 22538\n",
      "DR hits with E-values < 10: 0\n",
      "DR hits with hit lengths (fraction of total) < 0.66: 0\n",
      "DR hits used for filtering: 22538\n",
      "### DR filtering of spacer blast hits ###\n",
      "Total spacer blast hits: 2585\n",
      " NOTE: Deleting the blast hits to arrays\n",
      "Spacer blast hits hitting an array defined just by DR-blast hits: 2414\n",
      " NOTE: Deleting the blast hits to arrays\n",
      "Spacer blast hits hitting a mobile genetic element: 171\n",
      "\n",
      "-rw-rw-r-- 1 nyoungb2 nyoungb2 1.9M Jan  3 20:23 spacers_cut1_blastn_filt.srl\n"
     ]
    }
   ],
   "source": [
    "# filtering spacer hits to arrays\n",
    "## This is based on the premise that DR hits adjacent to spacer hits signify that the spacer is hitting a CRISPR array\n",
    "\n",
    "!cd $blastDir; \\\n",
    "    CLdb -- arrayBlast -- filter_arrayHits \\\n",
    "    spacers_cut1_blastn.srl \\\n",
    "    DRs_cut1_blastn.srl \\\n",
    "    > spacers_cut1_blastn_filt.srl\n",
    "    \n",
    "    \n",
    "# checking output file\n",
    "!cd $blastDir; \\\n",
    "    echo; \\\n",
    "    ls -thlc spacers_cut1_blastn_filt.srl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "source": [
    "## Converting blast .srl to .csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "code_folding": [],
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Decoding the srl file...\n",
      "qseqid\tsseqid\tpident\tlength\tmismatch\tgapopen\tqstart\tqend\tsstart\tsend\tevalue\tbitscore\t\n",
      "NA|spacer|NA|841|1\tNC_007355\t95.6521739130435\t23\tundef\t5\t10\t32\t3027136\t3027114\t0.000347095\t38.1576\n",
      "NA|spacer|NA|465|1\tNC_007355\t96.2962962962963\t27\tundef\t5\t5\t31\t857001\t856975\t1.48854e-06\t46.087\n",
      "NA|spacer|NA|465|1\tNC_007355\t100\t19\tundef\t5\t13\t31\t4703448\t4703466\t0.000362872\t38.1576\n",
      "NA|spacer|NA|465|1\tNC_007355\t100\t19\tundef\t5\t13\t31\t1036282\t1036264\t0.000362872\t38.1576\n",
      "NA|spacer|NA|280|1\tNC_007355\t95.8333333333333\t24\tundef\t5\t2\t25\t3935219\t3935242\t9.58273e-05\t40.14\n",
      "NA|spacer|NA|722|1\tNC_007355\t100\t19\tundef\t5\t3\t21\t3312027\t3312009\t0.000347095\t38.1576\n",
      "NA|spacer|NA|184|1\tNC_007355\t91.8918918918919\t37\tundef\t5\t1\t37\t2848699\t2848735\t9.9483e-08\t50.0517\n",
      "NA|spacer|NA|446|1\tNC_007355\t88.8888888888889\t36\tundef\t5\t1\t36\t3417664\t3417699\t9.18345e-05\t40.14\n",
      "NA|spacer|NA|446|1\tNC_007355\t88.8888888888889\t36\tundef\t5\t1\t36\t3928350\t3928315\t9.18345e-05\t40.14\n"
     ]
    }
   ],
   "source": [
    "!cd $blastDir; \\\n",
    "    CLdb -- arrayBlast -- srl2csv \\\n",
    "    spacers_cut1_blastn_filt.srl \\\n",
    "    > spacers_cut1_blastn_filt.txt\n",
    "    \n",
    "# assessing table of blast hits    \n",
    "!cd $blastDir; \\\n",
    "    head spacers_cut1_blastn_filt.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of blast hits: 171\n"
     ]
    }
   ],
   "source": [
    "#!printf \"Number of blast hits: \"\n",
    "nhits = !cd $blastDir; wc -l spacers_cut1_blastn_filt.txt\n",
    "nhits = nhits[0].split(' ')[0]\n",
    "print \"Number of blast hits: {}\".format(int(nhits) - 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Conclusions\n",
    "\n",
    "* **We have some BLAST hits!**\n",
    "* Let's add some information to the blast results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## adding crRNA info\n",
    "\n",
    "* Adding crRNA region to .srl (actually the DNA template, refering to it as 'crDNA')\n",
    "  * How much of the adjacent DR sequences are included in the crDNA is determined by the user (default: 10bp on either side)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Decoding .srl file...\n",
      "Extracting all queries with blast hits from blast hit file...\n",
      "...Number of total unique query IDs:\t1103\n",
      "Getting info from CLdb on each blast query spacer...\n",
      "Getting spacer regions from each genome...\n",
      " Processing genome: Methanosarcina_mazei_Go1.fasta...\n",
      " Processing genome: Methanosarcina_mazei_SarPi.fasta...\n",
      " Processing genome: Methanosarcina_acetivorans_C2A.fasta...\n",
      " Processing genome: Methanosarcina_barkeri_str_fusaro.fasta...\n",
      " Processing genome: Methanosarcina_mazei_WWM610.fasta...\n",
      " Processing genome: Methanosarcina_horonobensis_HB_1.fasta...\n",
      " Processing genome: Methanosarcina_mazei_S_6.fasta...\n",
      " Processing genome: Methanosarcina_mazei_LYC.fasta...\n",
      " Processing genome: Methanosarcina_mazei_C16.fasta...\n",
      "\n",
      "-rw-rw-r-- 1 nyoungb2 nyoungb2 2.3M Jan  3 20:24 spacers_cut1_blastn_filt_crDNA.srl\n"
     ]
    }
   ],
   "source": [
    "!cd $blastDir; \\\n",
    "    CLdb -- arrayBlast -- add_crRNA \\\n",
    "    < spacers_cut1_blastn_filt.srl \\\n",
    "    > spacers_cut1_blastn_filt_crDNA.srl\n",
    "   \n",
    "## checking output\n",
    "!cd $blastDir; echo; \\\n",
    "    ls -thlc spacers_cut1_blastn_filt_crDNA.srl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Adding protospacer info\n",
    "\n",
    "* Similar to adding crRNA info, but getting information from the blast database(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieving proto info from blast db: 'Methanosarcina_mazei_LYC.fasta'\n",
      "Retrieving proto info from blast db: 'Methanosarcina_acetivorans_C2A.fasta'\n",
      "Retrieving proto info from blast db: 'Methanosarcina_mazei_Go1.fasta'\n",
      "Retrieving proto info from blast db: 'Methanosarcina_mazei_WWM610.fasta'\n",
      "Retrieving proto info from blast db: 'Methanosarcina_horonobensis_HB_1.fasta'\n",
      "Retrieving proto info from blast db: 'Methanosarcina_barkeri_str_fusaro.fasta'\n",
      "Retrieving proto info from blast db: 'Methanosarcina_mazei_C16.fasta'\n",
      "Retrieving proto info from blast db: 'Methanosarcina_mazei_SarPi.fasta'\n",
      "Retrieving proto info from blast db: 'Methanosarcina_mazei_S_6.fasta'\n",
      "Retrieving proto sequences from blast db: 'Methanosarcina_mazei_LYC.fasta'\n",
      "Retrieving proto sequences from blast db: 'Methanosarcina_mazei_S_6.fasta'\n",
      "Retrieving proto sequences from blast db: 'Methanosarcina_mazei_C16.fasta'\n",
      "Retrieving proto sequences from blast db: 'Methanosarcina_mazei_Go1.fasta'\n",
      "Retrieving proto sequences from blast db: 'Methanosarcina_horonobensis_HB_1.fasta'\n",
      "Retrieving proto sequences from blast db: 'Methanosarcina_acetivorans_C2A.fasta'\n",
      "Retrieving proto sequences from blast db: 'Methanosarcina_barkeri_str_fusaro.fasta'\n",
      "Retrieving proto sequences from blast db: 'Methanosarcina_mazei_WWM610.fasta'\n",
      "Retrieving proto sequences from blast db: 'Methanosarcina_mazei_SarPi.fasta'\n",
      "\n",
      "-rw-rw-r-- 1 nyoungb2 nyoungb2 2.3M Jan  3 20:24 spacers_cut1_blastn_filt_crDNA_proto.srl\n"
     ]
    }
   ],
   "source": [
    "!cd $blastDir; \\\n",
    "    CLdb -- arrayBlast -- add_proto -workers 24 \\\n",
    "    < spacers_cut1_blastn_filt_crDNA.srl \\\n",
    "    > spacers_cut1_blastn_filt_crDNA_proto.srl   \n",
    "    \n",
    "## checking output\n",
    "!cd $blastDir; echo; \\\n",
    "    ls -thlc spacers_cut1_blastn_filt_crDNA_proto.srl   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Aligning crRNA & protospacer\n",
    "\n",
    "* Alignment added to the .srl file.\n",
    "* These are individual alignments between the spacer and protospacer (using clustalw)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Aligning protospacers from blast db: 'Methanosarcina_mazei_S_6.fasta'\n",
      "Aligning protospacers from blast db: 'Methanosarcina_acetivorans_C2A.fasta'\n",
      "Aligning protospacers from blast db: 'Methanosarcina_barkeri_str_fusaro.fasta'\n",
      "Aligning protospacers from blast db: 'Methanosarcina_mazei_LYC.fasta'\n",
      "Aligning protospacers from blast db: 'Methanosarcina_mazei_C16.fasta'\n",
      "Aligning protospacers from blast db: 'Methanosarcina_mazei_SarPi.fasta'\n",
      "Aligning protospacers from blast db: 'Methanosarcina_horonobensis_HB_1.fasta'\n",
      "Aligning protospacers from blast db: 'Methanosarcina_mazei_Go1.fasta'\n",
      "Aligning protospacers from blast db: 'Methanosarcina_mazei_WWM610.fasta'\n",
      "\n",
      "-rw-rw-r-- 1 nyoungb2 nyoungb2 2.3M Jan  3 20:25 spacers_cut1_blastn_filt_crDNA_proto_aln.srl\n"
     ]
    }
   ],
   "source": [
    "# aligning crRNA and protospacer\n",
    "\n",
    "!cd $blastDir; \\\n",
    "    CLdb -- arrayBlast -- align_proto \\\n",
    "    < spacers_cut1_blastn_filt_crDNA_proto.srl \\\n",
    "    > spacers_cut1_blastn_filt_crDNA_proto_aln.srl\n",
    "    \n",
    "## checking output\n",
    "!cd $blastDir; echo; \\\n",
    "    ls -thlc spacers_cut1_blastn_filt_crDNA_proto_aln.srl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting alignments (fasta)\n",
    "\n",
    "* You can parse out the alignments you want (parse by subtype, taxon_name, etc.)\n",
    "  * **OR** you can parse them after writing (add subtype, taxon_name, e-value, etc. to the sequence name)\n",
    "* The alignments will be oriented to the crDNA (5'-3' for the crRNA)\n",
    "* See `CLdb -- arrayBlast --perldoc -- get_align` for info on what `-outfmt` does"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Decoding .srl file...\n",
      "Getting alignments...\n",
      "\n",
      "-rw-rw-r-- 1 nyoungb2 nyoungb2 35K Jan  3 20:25 blastn_crDNA-proto_aln.fna\n"
     ]
    }
   ],
   "source": [
    "# getting alignment of protospacer and crDNA and writing as a fasta\n",
    "\n",
    "!cd $blastDir; \\\n",
    "    CLdb -- arrayBlast -- get_align \\\n",
    "    -outfmt taxon_name,subtype,evalue \\\n",
    "    < spacers_cut1_blastn_filt_crDNA_proto_aln.srl \\\n",
    "    > blastn_crDNA-proto_aln.fna\n",
    "    \n",
    "\n",
    "## checking output\n",
    "!cd $blastDir; echo; \\\n",
    "    ls -thlc blastn_crDNA-proto_aln.fna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">crDNA|17|56|1ebv46LCvRJMH|0.000298084\r\n",
      "tgctcgcgaaTCCCTGGTTTATGTCCTGCT.CGTCAAAATCTATGTTgtttcaatcc\r\n",
      ">proto|17|56|1ebv46LCvRJMH|0.000298084\r\n",
      ".agttaattacACCTGGTTTATGTCCTGCTCCGATTACCGCAACAAtggttttgtt.\r\n",
      ">crDNA|17|89|1ebv46LCvRJMH|0.000298084\r\n",
      "tgctcgcgaaTCCCTGGTTTATGTCCTGCT.CGTCAAAATCTATGTTgtttcaatcc\r\n",
      ">proto|17|89|1ebv46LCvRJMH|0.000298084\r\n",
      ".agttaattacACCTGGTTTATGTCCTGCTCCGATTACCGCAACAAtggttttgtt.\r\n",
      ">crDNA|7|7|1eX1MJpd4Qk75|0.000325182\r\n",
      "aggattgaaaCTTCTTCGATTGTTTTTATGCCGTT.TTTATCAAAAATAgtcgagaagt\r\n"
     ]
    }
   ],
   "source": [
    "# let's take a look at the sequences \n",
    "!cd $blastDir; \\\n",
    "    head -n 10 blastn_crDNA-proto_aln.fna"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* __The alignment is really just for each crDNA-protospacer pair (every 4 lines)__\n",
    "* First, the crRNA (sequence name starts with \"crRNA\")\n",
    "* Second, the protospacer (sequence name start with \"proto\")\n",
    "* Example: \n",
    "  * ~~~\n",
    ">crDNA\n",
    "ATAGACA\n",
    ">proto\n",
    "ATAGACA\n",
    "~~~\n",
    "\n",
    "* **Note:** the lower case letters are sequence adjacent to the actual protospacer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**If you wanted each crRNA-protospacer pair individually, you could use the `split` bash command like this:**\n",
    "\n",
    ">WARNING: this produces a lot of files, so we will make them in a new directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of files produced: 179\n"
     ]
    }
   ],
   "source": [
    "# file/directory\n",
    "inFile = os.path.join(blastDir, 'blastn_crDNA-proto_aln.fna')\n",
    "outDir = os.path.join(blastDir, 'blastn_crDNA-proto_aln_split')\n",
    "if not os.path.isdir(outDir):\n",
    "    os.makedirs(outDir)\n",
    "outPre = os.path.join(outDir, 'split')    \n",
    "\n",
    "# spliting alignment\n",
    "!split -l 4 -d -a 4 $inFile $outPre\n",
    "\n",
    "# checking out files\n",
    "!printf 'Number of files produced: '\n",
    "!find $outDir -name \"split*\" | wc -l"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting PAMs\n",
    "\n",
    "* **Looking at the alignments can be very useful, but it can help to extract what we specifically want and summarize.**\n",
    "  * Let's pull out the potential PAM regions from the alignment \n",
    "  * We are going to assume that the PAM is:\n",
    "    * 4bp long \n",
    "    * 1bp away from the 5' end of the protospacer ('right' side of the alignment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\r\n",
      ">proto|18|37|uWdGwRtuvgdp|3.13923e-13\r\n",
      "taat\r\n",
      ">proto|18|74|1vQhyRa3ZkVOT|1.31536e-06\r\n",
      "taaa\r\n",
      ">proto|10|18|1VdAVgxl34ddZ|3.90827e-07\r\n",
      "taaa\r\n",
      ">proto|17|33|u4GEEvxNLPSL|0.000307282\r\n",
      "atat\r\n",
      ">proto|11|78|1beRkenSGXl|0.000347095\r\n",
      "tata\r\n"
     ]
    }
   ],
   "source": [
    "!cd $blastDir; \\\n",
    "    CLdb -- arrayBlast -- get_PAM -PAM 1 4 -f - \\\n",
    "    < blastn_crDNA-proto_aln.fna \\\n",
    "    > blastn_crDNA-proto_aln_PAM.fna\n",
    "\n",
    "\n",
    "## checking output\n",
    "!cd $blastDir; echo; \\\n",
    "    head blastn_crDNA-proto_aln_PAM.fna"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Summary table of each unique PAM sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     42 taaa\r\n",
      "     31 ttaa\r\n",
      "     12 tggt\r\n",
      "      7 taac\r\n",
      "      7 ctgc\r\n",
      "      7 cagg\r\n",
      "      6 tatt\r\n",
      "      6 ccgg\r\n",
      "      5 tccg\r\n",
      "      5 taat\r\n",
      "      5 gggc\r\n",
      "      5 atat\r\n",
      "      5 aaaa\r\n",
      "      4 tgaa\r\n",
      "      4 caat\r\n",
      "      3 aagg\r\n",
      "      2 tgac\r\n",
      "      2 gaca\r\n",
      "      2 ataa\r\n",
      "      1 ttgc\r\n",
      "      1 tctg\r\n",
      "      1 tcca\r\n",
      "      1 tcaa\r\n",
      "      1 tatg\r\n",
      "      1 tata\r\n",
      "      1 taag\r\n",
      "      1 gggg\r\n",
      "      1 gctg\r\n",
      "      1 cgca\r\n",
      "      1 ccgc\r\n",
      "      1 caag\r\n",
      "      1 attg\r\n",
      "      1 atga\r\n",
      "      1 agtc\r\n",
      "      1 aggg\r\n",
      "      1 agac\r\n",
      "      1 aatt\r\n",
      "      1 aaga\r\n"
     ]
    }
   ],
   "source": [
    "!cd $blastDir; \\\n",
    "    egrep -v \"^>\" blastn_crDNA-proto_aln_PAM.fna | \\\n",
    "    sort | uniq -c | sort -k 1 -n -r "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Spacer (crDNA) - protospacer mismatch distribution\n",
    "\n",
    "* Let's look at potential spacer targeting by assessing mismatches of the crDNAs and protospacers.\n",
    "* This script will make 2 summary files of mismatches for each crDNA-protospacer alignment\n",
    "* Given that spacers can vary in length, the alignment is relative to a user-defined 'SEED' region.\n",
    "* I'm going to use the default SEED region: '-SEED -8 -1'\n",
    "  * This SEED region covers the last 8 bp of the alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SEED summary file written: 'blastn_crDNA-proto_aln_SEEDstats-summary.txt'\n",
      "SEED by-position table written: 'blastn_crDNA-proto_aln_SEEDstats-byPos.txt'\n",
      "\n",
      "-rw-rw-r-- 1 nyoungb2 nyoungb2 3.3K Jan  3 20:28 blastn_crDNA-proto_aln_SEEDstats-byPos.txt\n",
      "-rw-rw-r-- 1 nyoungb2 nyoungb2  55K Jan  3 20:28 blastn_crDNA-proto_aln_SEEDstats-summary.txt\n"
     ]
    }
   ],
   "source": [
    "!cd $blastDir; \\\n",
    "    CLdb -- arrayBlast -- get_SEEDstats \\\n",
    "    --SEED -8 -1 \\\n",
    "    -prefix blastn_crDNA-proto_aln \\\n",
    "    < blastn_crDNA-proto_aln.fna \n",
    "    \n",
    "## checking output files\n",
    "!cd $blastDir; echo; \\\n",
    "    ls -thlc blastn_crDNA-proto_aln_SEEDstats*.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "region\tpos_rel_SEED\tpos_rel_align\tcount\tmismatches\tmismatch_norm\n",
      "SEED\t1\t10\t179\t94\t0.525139664804469\n",
      "SEED\t2\t11\t179\t48\t0.268156424581006\n",
      "SEED\t3\t12\t179\t38\t0.212290502793296\n",
      "SEED\t4\t13\t179\t65\t0.363128491620112\n",
      "SEED\t5\t14\t179\t63\t0.35195530726257\n",
      "--------------------------------------------------\n",
      "alignment\tcrDNA\tprotospacer\tregion\tcount\tmismatches\tmismatch_norm\n",
      "9\tcrDNA|11|78|1beRkenSGXl|0.000347095\tproto|11|78|1beRkenSGXl|0.000347095\tprotospacer\t40\t10\t0.25\n",
      "9\tcrDNA|11|78|1beRkenSGXl|0.000347095\tproto|11|78|1beRkenSGXl|0.000347095\tnonSEED\t32\t8\t0.25\n",
      "9\tcrDNA|11|78|1beRkenSGXl|0.000347095\tproto|11|78|1beRkenSGXl|0.000347095\tSEED\t10\t2\t0.2\n",
      "164\tcrDNA|11|49|2A8bCEvyJhoO3|6.57658e-12\tproto|11|49|2A8bCEvyJhoO3|6.57658e-12\tSEED\t10\t2\t0.2\n",
      "164\tcrDNA|11|49|2A8bCEvyJhoO3|6.57658e-12\tproto|11|49|2A8bCEvyJhoO3|6.57658e-12\tprotospacer\t39\t5\t0.128205128205128\n"
     ]
    }
   ],
   "source": [
    "# checking files\n",
    "!cd $blastDir; \\\n",
    "    head -n 6 blastn_crDNA-proto_aln_SEEDstats-byPos.txt\n",
    "    \n",
    "print '-' * 50\n",
    "\n",
    "\n",
    "!cd $blastDir; \\\n",
    "    head -n 6 blastn_crDNA-proto_aln_SEEDstats-summary.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plotting mismatches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "library(ggplot2)\n",
    "library(gridExtra)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAyAAAAHgCAMAAABq2fnHAAAC/VBMVEUAAAAAv8QBAQECAgIDAwME\nBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUWFhYX\nFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMlJSUmJiYnJycoKCgpKSkqKior\nKyssLCwtLS0uLi4vLy8wMDAyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6Ojo7Ozs8PDw9PT0+Pj4/\nPz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tMTExNTU1OTk5PT09QUFBRUVFS\nUlJTU1NUVFRVVVVWVlZXV1dZWVlaWlpbW1tcXFxdXV1eXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVm\nZmZnZ2doaGhpaWlra2tsbGxtbW1ubm5vb29wcHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6\nenp7e3t8fHx9fX1+fn5/f3+AgICBgYGCgoKDg4OEhISFhYWGhoaHh4eIiIiJiYmKioqLi4uMjIyN\njY2Ojo6Pj4+QkJCRkZGSkpKTk5OUlJSVlZWWlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6fn5+g\noKChoaGioqKjo6OkpKSlpaWmpqanp6eoqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKz\ns7O0tLS1tbW2tra3t7e4uLi5ubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXG\nxsbHx8fIyMjJycnKysrLy8vMzMzNzc3Ozs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ\n2dna2trb29vc3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs\n7Ozt7e3u7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4dm34+Pj5+fn6+vr7trH7+/v8/Pz9\n/f3+/v7///+nmVqcAAAgAElEQVR4nO2dC3gUVZqwGw1jnMTQDBFBcYgjQ+aXEYYoKBcDoyzrZb3A\nSHZch/zzz8A4ju44IDrMThYzow6yi/vHlSCXAEF0g6MLclEJd00GgoBCSExMkEtunUsnIemQzna6\nz7Onbp2qStXpOtWn6arO9z481fVV1fnOR3e9qepLnXIgAAB0cUS7AACwMiAIABAAQQCAAAgCAARA\nEAAgAIIAAAEQBAAIqASp9Hpy1vVEpxQAsB5KQZbFN740afxTUaoFACyHUpBhJwIjz1UlR6kWALAc\nSkGcZ49MQHVJUaoFACyHUpCFqbeuq548J0q1AIDlUArSWlDQW7GiPkq1AIDlkAvi8432YdxmTrE6\nW/rT7tFYSI2bRZLWbhZZ3Exq8TLJwiRJVxuLLExK6bxE24KVAmTkgsTFOeI4Mkzkaa3pT3O3xkJq\nGmsZJGnwM0hSU9/AIkugnkGSOheDJDXeZhZZmlgk6WijbcFKATLKU6xZpvOAIIYBQbSwhyDmAUEM\nA4JoYQ9BCu9O5TCRBwQxDAiihT0EGb3kVBnGRB6iIINkmI4GGY4uyzEc1bCJzHVuNopymQIDSJAb\nPWbzgCDUexeLyFJlDgBBXnu112Setqb+tPeIM/Ld13Q0yHCkeNkMR01sInOdm40sVWZYdHVSNmgI\nf+c3glKQqYlDxkbgPYji7zscQRhGlipzABxBygRM5AFBqPcnFpGFy4xJQRDqrQ2YygOCUO9BLCIL\nlxmTglyckTC0ZFq1iTwgCPUexCKycJkxKcisp7tSfIvuNZEHBKHeg1hEFi4zJgX5thuloMYEKaQ4\n3wJBqPcgFpGFy4xJQcZ/iAXZ8UMhoDrfAkGo9yAWkYXLjElBDiTPTZoz/GMhoDrfAkGo9yAWkYXL\njElBUFNe9to6cV51vkUGBKHeg1hEFi4zNgVxVXII84rzrVCAINR7EIvIwmXGpCDPOFLGYIRAcb4l\n52Ap8r2yatWngR0bNneIy0AQ6j2IRWThMmNSkMRieSQ/3+rDv35ZKXK9j+eqC9DxPeJSEIR6D2IR\nWbjMmBQk7VzoFgF/YSkqf+udgtYDRci1Hi+pPH369AWNi4Yv+cQZ+e5rOhpkOFK8UIajFjaRuc7N\nRhYuk5Ju2gEMXBHxoR9KQYpuWJiF4WadAkNH9W+ztxSdP41Obdl1BnnfwPF/b9y48VR3f3r84ox8\n9zUdDTIcKV4ow1E3m8hc52YjC5dJSa+PskHrldBDLciUtCVLMdys2/1vPy6qK7pvff82WBBMz+v7\n8BFknbgMTrEou4NTLDueYiXLx1K5mSvh4s3922BBDh9F1VuqtqKT8B4EBNGOYlKQ57bIglGH8OTg\nd/u3wYJ43lm/qTmwa/OWTnEZCEK9B7GILFxmTAoy5epRfYM2/Md3XljzwndyjOUBQaj3IBaRhcuM\nSUGUF0wdejbjt58ZzAOCUO9BLCILlxmTgugTakAgEIR6D2IRWbjMmBREpkHazjQBYU2oAYFAEOo9\niEVk4TJjUhCZBiVNB0o4DghrQg0IBIJQ70EsIguXGZOCyDRQDfUeakAgEIR6D2IRWbjMmBREpoFq\nqPdQAwKBINR7EIvIwmXGpCAKDRRDvYcaEAgEod6DWEQWLjO6guTHmRRCDXFcrN46X986RdAPEIR6\nD2IRWbjM6ArSXmlKh/4QPuZ1zRucOHieSyPQoNXVn7YecUa++5qOBhmOFC+U4cjFJjLXudnIwmVS\n4umgbKC6EGNNwjMvNtyfdE9F59whmY4T+AjSlTl0ZDZac+28xEdIf9qpBEnhGTfjNe4XJI9mupDr\nZ+IdPRWBliB1/XF7xRn57ms6GmQ4UrxQhqM6NpG5zs1GFi6Tks52yga1KkEcz5f9ZkLDE5NWXPfp\nb3lBXksoznMcWuN4L8/xOStB8qZvP7H9ntyPHlyAgyTul4tNQ4Q1ikBLEDjFousOTrHYnmKtie9F\nM4ampoz8xxnoK16QJ9OR/9r/XJOALjgOsxJkNKdlzVjk4a4C+T73S9094jt2RQCC6EfmOgdBwhYk\nEaEF05vz/nWZs3gxL8jyhJKN+AiSyFSQEjw5eiNq4AR5zzk/a77zr8IaRQCC6EfmOgdBWAhSMzNp\n5uGWWYlzeEE8850jXkZsBckbvviNxcNzD9z4Fy6qWrlkZZW0ShGAILqRuc5jUBBl5xEXJEjhMt8u\nxwXzRqhQfYr15e/nv3AMlRUJEQw9Sh+Z6xwEYSZI9fRbRi03aYMGhI95ZUOPqn65qAEIQtkdCBIh\nQRgjF0SlgWzoUdUvFzUAQSi7A0HsJ0hJU4mAEMqGHlX9clEDEISyOxCErSCX1URAEBWyoUdVv1zU\nAASh7A4EsZ8g4ySEUDnUu2y7g6WIH3UUhh7ViMx1DoLYQZATEmJMGHqUH3UUhh7ViMx1DoLYQRCJ\nTf2WyN++80OP8qOOSkOPgiAmugNB7ChI2YKMjIyHR3CzivMt5dv3vaWIH3VUGnp00+uvv/55b3/8\nAXFGvvuajgYZjhTPk+Gol01krnOzkW3KDEnAb2AjOe20ggT+MPGu+2srh0ydOnVxcDppegmNIHdm\nLnogf9IOblZ9vlXp9eSs65EE4UcdlYYebcA618PP3em6i/mfuys7Z/9z93paQT6bHkD/9VRlOr83\nB6dnvk8ebUEpSHxTx32ofpoQdK89iPJXeoVgWXzjS5PGPyUJwo86CkOPakTmOodTrAicYm39+ZNT\nt7XNuW/mGX6u4pZCX2+rWhD0xCcUgqQcQmktnmFCsDDtJCq68+dCMOxEYOS5qmRJEH7UURh6VCMy\n1zkIEglBZgaOPP5yFtr7ED+Hip4cl1FR6UxPT88Rp5wgSzZQCLL6mnN/mHjP34lOcL9NrHIKgfPs\nkQmoDr4oDBmZ6xwEiYQgWejs3MxC1P0Dfu5UNQq8N5mXQn4E+ac9iITqU6zzXb63c9qE+TF/w5Mi\n8X5sC1NvXVc9mXBFIQhC1x0IEnlBsrEW+Aiy70F+bkeGH1VOVAtSnkrzHkTBu8m/WfHPyeJnvr6C\ngt6KFZdAkFCRuc5BkEgJ0vbYLPwehJvz/XHqndOP859fzQ1OJ087RvSDODZvefbCrC+CEYxqYigy\n1zkIYofvQRqQ/qglIUc1AUHougNB7CfI0LrvXuIRwskCQhByVBMQhK47EMR+giwaepVw504hLC4u\nLipI3yoEMKqJwchc5yCIHQTBx4l+6xtvFx5hVBODkbnOQZBwBfkfNeGaIaJ+k65+K35c/OoDRjUx\nGJnrHAQJV5BBahjIwaEURPFWnHsDkjb4WXEVjGpiLDLXOQhiD0EUb8WLOcqkcU0CFfvL/fp52pr6\n094jzsjrNh0NMhwpXijDURObyFznZiPblBmSrk4DG8lpiIYg+m/FyyYkT7x+PNz+IGRkrnM4gtjj\nCKJ4K664b+ddS3pQz5IpIEioyFznIIg9BFG8FVfct9PZiCcu+Jg3ZGSucxAkAoKYu0KKKIjirbji\nvp1P5OLJm4+DIKEic52DIBEQxNwVUmRB5Cju2/mLweMfGe94ICNDZ+QfEISyOxAk4oKYu0LKuCCK\nGxbmBwFBLLXn2aTMaAhi7gop44KEum8nCKIVmescBImAIOaukCIKcvtpmrZyQBDK7kCQiAti7gop\nFUpBsjO7dbbrkE1BEEJkrnMQJAKCmLtCiihIuvOaMcG3HTJg8GrDkbnOQRB7fA9ySqDfVv0Gr/a9\nsmrVpzA2r0ZkrnMQxB6CKG4qVThNNpL1LOVmrvfxBMbm1YjMdR7zgigj2woiu6kUQqOzS9WfYgVH\n7S1/652CVmls3tPHjh075+5Ph0+ckddtOhpkOFK8UIYjN5vIXOdmI5uWqUF3l+ZifRqjIYjsplII\n3dAlWyMbtZfj/Gl0aos0Nu/ugoKCUk9/vH5xRl636WiQ4UjxQhmOPGwic52bjWxapga+Hs3F+rij\nIYjsplIIvST/Jl02aq9Iz+vS2LwccIpF2R2cYrE9xYoUSkFkN5VKTR17leybdMWovQgdPoqqt8DY\nvBqRuc5BEHsIIrupVFnZMf6LdPFTY8WovQh53lm/qRnG5tWIzHUOgthDENlNpVRffShG7dUABKHs\nDgSxpSC9Zfu/Fj7nVX/1IR+1VwMQhLI7EMSOgnzxg+SJyXeIF4TM0thcFxCEsjsQxI6C3MldWLso\n3UQeEISyOxDEjoIk9V1YK79vpwFAEMruQBA7CjJvDZ6sfoybVd63MyQgCGV3IIj9BHniiTmOiXMn\nXr1ACDfSlACCUHYHgthPENVVtTMTb3tmu/7nVkpAEMruQBD7CYJxVXKIQU/Jf8xz3m0sDwhC2R0I\nYkdBnnGkjMEIQdenyx8b+X9+aSxPG9wnna67AXafdGUUgfukRwqlIInFsuBb1z29rclontba/rR4\nxRn57ms6GmQ4UrxQhqNaNpG5zs1GNi1Tg852zcUEWClARilI2jlZsPyx0Wm/zq9EhoBTLMru4BTL\njqdYRTcszMJIYaDimUTSXT5lgCCU3YEgdhRkStqSpRgh2Pni9KTpWQeN5QFBKLsDQewoSHKLLLjj\n+Y/0h/lRA4JQdjegBVFGNhLkuS1m84AglN2BIMHIRoJMuXpUqta4WKEBQSi7A0GCkY0EoRqOVwEI\nQtkdCBKMbCSIeUAQyu5AkGBkI0EmC5jIA4JQdgeC6ESWFqS4uLioIH2rwbYw9KhGZK5zECQYWVoQ\nnsbbDbaFoUc1InOdgyA6kRUFOa4/hrsSaehRDhCEsjsQxFBkMUG4NyBpg5812FYaejQnKyurmLSh\nfPc1HQ0yHCleKMMRYhOZ69xsFOtlEjD+JXZY9HsPUlxcFtDZVo009Gi3x+NpruuP26uxkJqmegZJ\nXH4GSepcLhZZAiyyNDQySFLnbWGRReu1p6aznbJBVH7NS4fhoUfDorGWQZIGP4MkNfUNLLIE6hkk\nqXMxSFLjbWaRpYlFEjt8zJsuYbCt4aFHwwIE0QIEiYIgH/Jkf3uCiTwgiGFAEC3sIAjHpd85/81n\nIo+WIGWHGDxxNfUsBDn7EYMkNbUsSqnZfZZFljoWSQ6Us8jCQviaoydpW7DY/UOjEmTbdx89zyz3\nGao7tkeU5r9Eu4I+/mx0pJjIs+6raFcQZMehaFegjUKQC4+N3s4wNwiiCQiihR0EWfmdxUw/XAZB\nNAFBtLCDII6r4gUY5T73MaNE4dP+brQr6CP4yV/02X0h2hUE+Vv/m49bArkgwe9golYNAFgMVteD\nAEBMAoIAAIEICiK/XCS6HCy1SjG+DzbkXrRGLd4t69e2WKMUhLr+3SqvkJoICiK/XCSa+NcvK7VK\nMeU7Uc0aa9Ry9AA6vssapSC0609WeYXURFAQ+eUi0STgLyy1SjG1jci91hq1NLSjE3utUQq6+EGO\nVV4hNREURLpcJPrsLbVOMRdzK6xSy7uvtlijFP/GjhzrvEJKIiiIdLlI9MGCWKSYQOH6OovU0u1H\n1ZusUUrxMZRjkWelHxEURH65SHTBglikmNNb/VZ5Yj45js7nWaOU9zfl/2mzNUrpRyQ/xZJdLhJd\nsCAWKWbbylWr3rZGLe0b175Va41SMDlWeYXUwPcgAEAABAEAAiAIABAAQbQIjE3uQejEOFSSZrRJ\nnN51mFyKOF+IRLirvkRa2yrSfzFryLB/qEAo2cFRGXx0Ohzxd+3Xr0q3RkAXEESLz0eO+Yjfa5t2\nGm2iu/NxKeJ8IRIpBNHaVp6+9+alVbWLxgVQ8j43pjf46Dzk/iYn8ZhuVSAIPSCIFosWLc0MHkHy\nUlI2pKCyqa/ddMteVDbld8lTi+647lmEDv0oYfZFafksx+gOhPJ/8aRzSjlCfx07ZI4L+X41NPkl\nLgVeeSBNWio2wJxKz75dzMIJ8tYt104uF7edvRqh5RniSgyXXmiOOedox5Y85EbJJUK50qPzBJ68\n+BNuVuibayak5foSagSoAEE08N/8xSlntyjIl9cfqZmGBUl8pfP5u1HZ1W83p93wTbHD1ZS8veXX\nM6Tlwl/n/MFvup6fEKh2ftKcmYEKUs8ej/86eIolLpUaYEGcmafFLLir89ccbMxcIG676mGEpnwg\nruSI84nNMb4fzt7t4WY0Bfnse9ys0DduJqbl+oIjiAlAEA0O/Qih27aJgvx+MULbsCBDfOhUKiob\nhdCShQilVG6Yi1BXQq+4XBRkPEI9wypWzkfI9a3egluPBBq9QUHEpVIDLMg13UjMgrvqOoc6FmWI\n29Ze11U3tEtcyW0b5xObc0H3m3+fPPsoFiPR6XTe1PfIC1LJXxAq9I2biWm5vkAQE4AgGjx17YgR\n8T8VBZn/Bj7ZwoKMRagslfuHlmYhNKYyOyklJWVorbhcFOQRPJm4fwl3I+3Eut61E0Zme4KCiEul\nBliQMQiJWXBXvn+5494HJEHQXbtXz5dWctvG+cTmeOLtwo7kxR9FydvOnj37DQo+yo4gQt+4mZiW\n6wsEMQEI0p+e6w/U1e1P7BQEeeF5hLan8Pu0QpC1c/BuWBIQl/cdQXzXl6/MxH/sB/sqqtG5STl9\nRxBhqdQA8ccRMQvuasvEZpQfFOQvT9+/W1rJbYuPIEJzPL/5x9ySGatJ70GEvnEzMS1/zAJB6AFB\n+rP7e3ifDKRsFQQ5NrykNl1DkLrhOxsXTUVBQdy4Zb4jt/GF8f6vhxS2/OxxtGJyffWEtbwgbjwV\nlyoFEbPgrnLu8TRMeljcFn016qYeaSVHnFtsjmkY+seK8jeTKsRPr7zBR+ch97k3Ej/nthH6xs3E\ntIIg7ig9ozYGBOnP/Oe46bNzxU+xcm8cmzuuvyDo43EJM6uC+/u8JO5TrAd+knTXGYTeG5v0aAO6\n9GjisF/1cCnmJXGfYglLlYKIWXBXrbOGTflwxCZxW3TbwuBKDpxeaM5R8eCIpLt3S9+D5Acfue9B\nJgvfgwh942Y1Qlq+L75GgAoQJBRlB/COOsPYtvkZka0FuOKAIKEourHB89ByY9uCIDEHCBKSV24a\n9UuDv8MGQWIOEAQACIAgAEAABAEAAiAIABAAQQCAAAgCAARAEAAgAIIAAAEQBAAIgCAAQAAEAQAC\nIAgAEABBAIAACAIABEAQACAAggAAARAEAAiAIABAAAQBAAIgCAAQMCjIwVIU2LFhcwc/iWxFAGAh\nDAniX7+sFFUXoON7+EmkawIAy2BIkIC/sBQdKEKu9fwEL6k6ffp0TYcGPR6tpdGCUTWdTLJ0eC31\n3Hi7ol2BnG7qatoibIaIwVOsvaVo1xnkfYOf4HjHxo0bT3ZrEOjRWkoNmyzd/lisxutlksbvY5KG\nUTW91NW0RtSLIMYF2YcPHuv4ibisrUYDX6PWUmqamGSp8bLJw6ia7hYmaVx1TNJ4WpmkaWBTTYfm\n3kQkYk4oMC5I1VZ0cg8/EZeBIJSAIPrEgCCBXZu3dPITcRkIQgkIoo/dBdECBKEktgW5LIc6DQhC\nBwiiDwgiwE4CEpESZJAc6v87CELAMoIojQBB1IAglIAg+oAgdIAg+oAgAuwkIAGCGAAE0QQEIQOC\nUAKC6AOC0AGC6AOCCLCTgEQ4gjRo0NsiPCoE0dqQTAt9Ey163EzSMKrGq/mMUdPkYpKmqz3cDHIj\nlNFl6lyeS7Qt6thJQAIEMQAIosmAF6TS68lZ16O7Gk6xKIFTLH3seIq1LL7xpUnjn9JdD4JQAoLo\nY0dBhp0IjDxXlay7HgShBATRx46COM8emYDqknTXgyCUgCD62FGQham3rquePEd3PQhCCQiijx0F\n8RUU9FasuKS7HgShBATRx46CINRbGyCsBUEoAUH0saMgF2ckDC2ZVq27HgShBATRx46CzHq6K8W3\n6F4p/HTVqlV/9r2yatWn4gIQhBIQRB87CvJtN0pBjQmyJRX7XO/3RSAIJSCIPnYUZPyHWJAdP+xb\n0LvRW/7WOwXSgEQgCCUgiD52FORA8tykOcM/7lvw2Zfo/Gl0agsyMHCcQhD6ccSsNVSbtaqxzMBx\nciOU0WXqXLYcOK4pL3ut7CdhgTd7uYee15Ew9OiFFg1624VHhSBaG5Jpp2+ihe8SkzSMqunpYJKm\nzc0kjbcz3AxyI5TRZepc3R7aFq7IiiFB8WvemgKEDh9F1VvEGE6xKIFTLH3seIpVOG0cRzDeexIh\nzzvrNzWLMQhCCQiijx0FGZ1dWobRXQ+CUAKC6GNHQW7oIjcFQSgBQfSxiSC3n5YFL73aS2wKglAC\ngujDWpD8OLNGqFAKkp3ZLc2mpo69asjY1NRU3aYgCCUgiD6sBWmvDFMMCaUg6c5rxohOlEnoNgVB\nKAFB9AlbkDUJz7zYcH/SPRWdc4dkOk7gI0hX5tCR2WjNtfMSH/GxEuSUgBB0rz2I8ld6QZABLAhx\nr7eWII7ny34zoeGJSSuu+/S3vCCvJRTnOQ6tcbyX5/iclSCKX7gvTDuJiu78OQgyoARR7uc2EiS+\nF80Ympoy8h9noK94QZ5MR/5r/3NNArrgOMxKEMUv3IdV4UmVM+KCaG0KgugDggioBElEaMH05rx/\nXeYsXswLsjyhZCM+giSyFETxC/cxf8OTojEgCAhiE0FqZibNPNwyK3EOL4hnvnPEy4itIIpfuL+b\n/JsV/5y8SV+QJg16W4VHxV6vtWEfyk0pGobGp1kjNW4mWZp62pmkaWlmkqa7Q3OxYj9XRYRN1Q1p\n6eqkbdGgs1sWLvPtclwwb4QKpSCKX7h3lGcvzPqiQ7dpZI4gYb55UQJHEH1i7AgSpHr6LaOWM7ID\nqQWR/cLd5xvtw7gjP+wPCEKHLQSh1cUm36TLfuEeF+eI48jQbQqCUAKC6GMXQXrL9n8tfc47K0RT\nEIQSEESfsAW5rCYignzxg+SJyXdUyRcR3qRrVQ2C6AOC6GMTQe5c0oN6FqULQdmCjIyMh0eAICCI\n9l4/AAVJasQT1xDRlsxFD+RP2gGCgCAgiMi8NXiy+jEhiG/quA/VT4uuIOHpAoLoA4JQC/LEE3Mc\nE+dOvHqBEKYcQmktnmEgCAgSE4IE/jDxrvtrK4dMnTp1cXA6aXqJcUHyJYRw9TXn/jDxnr+T1vKD\nKgZ2bNgsfXUIglACguhzBQT5bHoA/ddTlencfN/0zPc9hgXBuCo5xOB8l+/tnLbgKm5QxeoCdHwP\nCGIOEESfSAiy9edPTt3WNue+mWf4uYpbCn29rWpB0BOfUAjyjCNlDEbLFsQPqnigCLnWc3K0tLQ0\n1mvgaxYeFfu51oZ9KDdVNTSeRoueFvo2GriZZKn3tjJJ0+xikqarTXOxYicjRvX6Ub9NQ+K5RFu/\n6iaeWoLMDBx5/OUstPchfg4VPTkuo6LSmZ6eniNOOUGWbKAQJLFY1xZhUMVdZ5D3DRzkvfrqq0V+\nDZD4qNjPtTbsQ7mpqqHxNJrVBOjbaMAmiz9gh2oUOxkx8utH/TY1Ww0B1Y1rtATJQmfnZhai7h/w\nc6eqUeC9ybwU8iPIP+1BJJSCpJ3Ts4Wn5/V9+AiyTozgFIsSOMXSJyKnWNlYC3wE2fcgP7cjw48q\nJ6oFKU+leQ9SdMPCLIyGLeKgilVb0Ukz70FI+zkIQgcIImBMkLbHZuH3INyc749T75x+nP/8am5w\nOnnaMaIfKkGmpC1ZitGwRRxUMbBr85ZOEMQcFhXE+H5uM0GYoBQkuUXPFg1AEEpAEH1sIshzW2SB\nwhYNQBBKQBB9bCLIlKtHpQbHilPYAoKEz5UUJPQOCoKYEEQxVpzCFmsIoloZktgWhLQPDjhB/kdN\nRARRwHJkRRCkBgQhEbYgg9RERJDJAmJU6fXkrOsBQewviNbuCoKYEKS4uLioIH2rECyLb3xp0vin\nQBAQBASR03i78DjsRGDkuapkECR8QWh3FyIgiEDUBDkujvTjPHtkAqojDPtTq4GvUXhU1KoM1U2U\nmxIbqlaGpKfZwEahaQo3gWJ3CTdZY71GVnKPWv17WjXTkBqq0hhvGJrOdtqnIWrvQdIGPysEC1Nv\nXVc9eY5u0/ZGDXrdwqOiVmWoXkmKiA1D42szsFFo3OEmUOwuzNKQkio3JUWN+hGzhqHp6qB9MupD\nCmLuCimiIMUcZeK4P76Cgt6KFZc0GgmYP8VSriRFA/0US9lQlYaUlNSQmCYiDUMTiVMsc1dIEQWh\nAgQxDO3uotOQuIMab2h6PzfdMDSREMTcFVIEQdIluCBtZ5oACAKChNswNBF5k27qCimCIB/yZH97\nAheUNJUIgCAgSLgNQxMJQcxdIUUQhOPS75z/Jt3RTXnJrb0EoX2BiIAgYTYMTSQEMXeFFFmQbd99\n9Lw0r7rkFgQxD3HPMt6QIg2pITFNRBqGJhKCmLtCiiTIhcdGb++L+l9yC4IYhmLPikgaUkNimog0\nDI0tvihc+Z3F8tvlqC65BUFooNizIpKG1JCYJiINQ2MLQRxXxQsIoeqSWxCEBoo9KyJpSA2JaSLS\nMDS2EKROQghVl9z6PtiQe5EfXtGGglC+XCpAkDAbhsYWgqhQXXJbvhPVrOGHV7SmIBSvJS0gSJgN\nQ2NHQVSX3NY2IvdafnhFHPxtz5495Zc08HuER0WtylC9khQRGypRvCSq6JIyosVD34RRNabTkBoS\n00SkYWh6vLTPcDMjA0JAEKTfJbcXcyv44RURCBISRtWYTkNqSEwTkYahsaMgqktuA4Xr+TcnPa+L\nC+AUiwCjakynITUkpolIw9DY5SaeBE5v9YvDK4oLQBACjKoxnYbUkJgmIg1DEwOCbFu5atXb/PCK\n4gIQhACjahjtoBFJM8AFaUCuEE2jLogyongtaQFB2DbUwH6CDK37rvBmSLcpCEKAUTUUaUibRiTN\nABdk0dCrnDy6TUEQAoyqoUhD2jQiaQa4IAg9GqJpbAqi9VqCIGwbamBHQRDqrfMR1oIghtMwqoaY\nhrRpRNIMeEFc8wYnDp6n/1bdvoLQvpZXUhCK2sz+pxilGfCCPJrpQq6f6Q/7M+AFIb3sFNUYb8jo\nP8UozeRfW0EAABoUSURBVIAXJIn7sWLTEN31IAjhZaeoxnhDRv8pRmkGvCDf5y5u36N/+4OoDxyn\njBQvAqvB0VRpSIOzkdYRqzHekJjGeENGacw31CDsgeMiBUGQ95zzs+Y7/6q7fiAcQSjSwBHEZP+2\nPYKgqpVLVlbprwZBNF5oEIS6f/sKEgIQROOFNiFIRKq5AmlAkBCAIBovNAhC3T8IMjAEGXBpItM/\nCAKCxEgaECQEIMjATgOChAAEGdhprkT/IAgIYts0IIgGgR0bNkvDk4IgAzsNCKJBdQE6Lt1eAQQZ\n2GlAEA0OFCHXenEeBBnYaUAQDXadQd438GNuVlbWYdKGip1XGapXkiKKhopnVhUh/Ui9KaSxUv8E\nOkgr2UEpyD58BFmHH7s9Hk9znQa+Jq2l1GjmpqenhUkaRtV0u5mkaWxgkqarjUkaF5tqOttpW9RG\nxId+UApStRWdNPQeJEzY3FvQMvcoFOhuYZLGVcckjaeVSZoGNtXEyk9NArs2b+kU50EQSkAQfWJF\nEDkgCCUgiD4DRpDCr8N+rjjqmWSp2V/OJE19LZM0n55hkobNHllTdIpJmjo2z82Rk9RN2ElAIgxB\nNMm5wDhhWKzRv4l1FMg/Fe0K5GwtiXYFcj4kfiYaRUCQKwcIog8IEhVAEH1AEEOwFuSDRsYJw2Ln\nFTpRNcaes9GuQM7Br6JdgZxiS/3xkMFaEACIKUAQACAAggAAAbaCyC8XiT4HS61TkO+DDbkXLVON\nd8v6tS2WqQahrn+3ziulgq0g8stFoo1//bJS6xRUvhPVrLFMNUcPoOO7LFMNQrv+ZJ1XSgVbQeSX\ni0SbgL+w1DoF1TYi91rLVNPQjk7stUw16OIHOdZ5pVSwFUS6XMQa7C21UkEXcyssVM27r7ZYphr/\nxo4cK71SCtgKIl0uYg2wIJYpKFC4vs461XT7UfUmy1RTfAzlWOe5UcFWEPnlItEHC2KZgk5v9Vvo\n6fnkODqfZ5lq3t+U/6fNlqlGBeNPsWSXi0QfLIhlCtq2ctWqty1TTfvGtW/VWqYaTI51XikV8D0I\nABAAQQCAAAgCAARAECWBsck9CJ0Yh0rSjDaJ07uZPJcizhciEe6qL5HWtor0X8waMuwfKhBKdnBU\nBh+dDkf8Xfv1q9KtEQgBCKLk85FjPuL32qadRpvo7nxcijhfiEQKQbS2lafvvXlpVe2icQGUvM+N\n6Q0+Og+5v8lJPKZbFQhiFhBEyaJFSzODR5C8lJQNKahs6ms33bIXlU35XfLUojuuexahQz9KmH1R\nWj7LMboDofxfPOmcUo7QX8cOmeNCvl8NTX6JS4FXHkiTlooNMKfSs28Xs3CCvHXLtZPLxW1nr0Zo\neYa4EsOlF5pjzjnasSUPuVGyeLWT9Og8gScv/oSbFfrmmglpub6EGgETgCAK/Dd/ccrZLQry5fVH\naqZhQRJf6Xz+blR29dvNaTd8U+xwNSVvb/n1DGm58Nc5f/CbrucnBKqdnzRnZqCC1LPH478OnmKJ\nS6UGWBBn5mkxC+7q/DUHGzMXiNuuehihKR+IKznifGJzjO+Hs3d7uBlNQT77Hjcr9I2biWm5vuAI\nYhoQRMGhHyF02zZRkN8vRmgbFmSID51KRWWjEFqyEKGUyg1zEepK6BWXi4KMR6hnWMXK+Qi5vtVb\ncOuRQKM3KIi4VGqABbmmG4lZcFdd51DHogxx29rruuqGdokruW3jfGJzLuh+8++TZx/FYiQ6nc6b\n+h55QSrjuU2EvnEzMS3XFwhiGhBEwVPXjhgR/1NRkPlv4JMtLMhYhMpSuX9oaRZCYyqzk1JSUobW\nistFQR7Bk4n7l+ANUGJd79oJI7M9QUHEpVIDLMgYhMQsuCvfv9xx7wOSIOiu3avnSyu5beN8YnM8\n8XZhR/Lij6LkbWfPnv0GBR9lRxChb9xMTMv1BYKYBgSR03P9gbq6/YmdgiAvPI/Q9hR+n1YIsnYO\n3g1LAuLyviOI7/rylZn4j/1gX0U1Ojcpp+8IIiyVGiD+OCJmwV1tmdiM8oOC/OXp+3dLK7lt8RFE\naI7nN/+YWzJjNek9iNA3biam5Y9ZIIhZQBA5u7+H98lAylZBkGPDS2rTNQSpG76zcdFUFBTEjVvm\nO3IbXxjv/3pIYcvPHkcrJtdXT1jLC+LGU3GpUhAxC+4q5x5Pw6SHxW3RV6Nu6pFWcsS5xeaYhqF/\nrCh/M6lC/PTKG3x0HnKfeyPxc24boW/cTEwrCOKO0jNqe0AQOfOf46bPzhU/xcq9cWzuuP6CoI/H\nJcysCu7v85K4T7Ee+EnSXWcQem9s0qMN6NKjicN+1cOlmJfEfYolLFUKImbBXbXOGjblwxGbxG3R\nbQuDKzlweqE5R8WDI5Lu3i19D5IffOS+B5ksfA8i9I2b1Qhp+b74GgETgCD6lB3AO+oMY9vmZ0S2\nFiBKgCD6FN3Y4HloubFtQZAYBQQh8MpNo35p8CfYIEiMAoIAAAEQBAAIgCAAQAAEAQACIAgAEABB\nAIAACAIABEAQACAAggAAARAEAAiAIABAAAQBAAIgCAAQAEEAgAAIAgAEQBAAIACCAAABEAQACIAg\nAEAABAEAAiAIABAAQQCAQHiCdLZo4HVrLaWGTZaODiZp2BTT2s0kDaNqutqYpGFTTLuHugkjA0IQ\nniBtNRoE6rWWUtPEJMuldiZp2BTT4GeSps7FJI23mUkaNk9NSxd1E0YGhMCgIAdLuWlgx4bNHfxE\nXAyCUAGC6GJvQfzrl/GCVBeg43v4ibgCBKECBNHF3oIE/IW8IAeKkGs9P8HBhaqqqpomDQItWkup\naWOSxeNhkoZNMW4/kzTNrUzS9LQzScPmqbnkpW3REFEvghg8xdrLC7LrDPK+wU9w8NfVq1cf69EA\n+bSWUsMmi9/PJA2j/xJikoZRNYFeJmnYFNMrvU6X5ZBatEXQChlUguzDB491/ERcDKdYVMApli7B\nUyyFIMQmETJCBZUgVVvRyT38RFwMglABgugSA4I05KLArs1bOvmJuBgEoQIE0cXugugAglABgugC\nglADgugBgvAwMiAEIIghQBBdQBACIAgVIIguIAg1IIgeIAgPIwNCAIIYAgTRBQQhAIJQAYLoAoJQ\nA4LoEWOCKJWIMUG0fkXG6MeKbiZZGP1YkU0xrH6syOYJZvRjxbCfGoUSqoiAtX6sqAMcQaiAI4gW\nMX0E0SocBNEDBNECBDEHCKIHCMLDyIAQyAXZ76NtDYJQAYJoYR9B3vzpwu0eqtYgCBUgiBb2EQSh\nlvyfZb7tNt4aBKECBNHCToJgPP/9y5+Ks5VeT866HlJrEIQKEEQLmwmCEd+KLItvfGnS+KdIrUEQ\nKkAQLewniMiwE4GR56qSkTgeFn78dNWqVX/2vbJq1afiJiAIFSCIFrYVxHn2yARUl4TE8bCEhRX7\nXO/3bQKCUAGCaGErQeRvOxam3rquevIcJI6HxS/r3egtf+udglYQxAwgiBZ2EkTxtqO1oKC3YkU9\nEsfD4pd99iU6fxqd2oJnPykoKDjl0QB1aS2lxsski8/HJA2bYroCbNJ0M0njZ/OfCjuLQglVRIDi\nw1Z2gvS97UA+32gfxs2dYgWHwgq82cs99LyOJ2eOHTt21q1BoE1rKTUdTLJ0dzFJw6aYtgCbNO1M\n0vjY/KfCzqJQQhURaIyGIH1vO1BcnCOOIwOJ42FxC2sKEDp8FFVvEbeHUywq4BRLCzudYvW97cDM\nCi7mh8JqyEVo70mEPO+s39QMgpgBBNEiIoLkx0VEEB//tuOS4dYgCBUgiBYREaS90pQO/SF8zFt4\ndyoHqTUIQgUIogUbQdYkPPNiw/1J91R0zh2S6TiBjyBdmUNHZqM1185LfIT6V7h6ghROG8chBKOX\nnCrDgCA1IAgBqwjieL7sNxManpi04rpPf8sL8lpCcZ7j0BrHe3mOz1kJMjq7tM+JG0P/shcEoQIE\n0YKRIPG9aMbQ1JSR/zgDfcUL8mQ68l/7n2sS0AXHYVaC3NAlC157tRcEEQFBdLGKIIkILZjenPev\ny5zFi3lBlieUbMRHkESWgrwkd2Jq4pCx8B5EAATRxcRTo3SAnSA1M5NmHm6ZlTiHF8Qz3zniZcRS\nkNTUsVfJnCgTAEFqQBACFhEkSOEy3y7HBfNGqJALUlamcqK3NkBuDYJQAYIIRFSQ6um3jFoevhgS\nylOs7rUHUf5KrxBcnJEwtGRaNal1W60GgQatpdQ0McnScYlJGjbFuPxM0tQ3MknjbWGSxsRTI3dA\nGV1WRSTYSUBC9U162klUdOfPhWDW010pvkX3klq3uTQINGktpaaFSRZPJ5M0bIpp9jNJ09jMJE1P\nK5M0Jp4auQPK6LIqIlAfMScUqH6sWIUnVU4h+LYbpaDGBFJrOMWiAk6xBCJxinVZTUQEGfM3PCka\nIwTjP8SC7PghCFIDghAYUIK8m/ybFf+cvEkIDiTPTZoz/GMQpAYEITCgBOkoz16Y9UWHGDXlZa+t\nI7YGQagAQQTsKoj8GikOVyUHCFIDghAYQILIr5HCPONIGYMBQWpAEAK2ESTwh4l33V9bOWTq1KmL\ng9NJ00uMCyK/RgqTWExuCoLQEtOCGN+zoyTIZ9MD6L+eqkzn5vumZ75P/k2u1vUg4pv0tHMgiAQI\noot1Bdn68yenbmubc9/MM/xcxS2Fvt5WtSDoiU8oBClbkJGR8fAIISi6YWEWBvUNHMePGScFIAg1\nIEj/TSMpyMzAkcdfzkJ7H+LnUNGT4zIqKp3p6ek54pQTZMkGCkHuzFz0QP6kHUIwJW3JUgzqGziO\nHzNONoocWZBBcmifdhBEH6sIYnbPvmKCZKGzczMLUfcP+LlT1Sjw3mReCvkR5J/2IBJKQeKbOu5D\n9dOEILlFWiwNHMePGScFbS0tLY31GgTEpQpBtDYk46ZvokFnB5M0bIpp8jNJ42pmksYb7n9KsT8S\nI0JDchoCqi8gtATJxlrgI8i+B/m5HRl+VDlRLUh5Ks17kJRDKK3FM0wInpMG9wkOHMePGScFea++\n+mqRXwMkPioE0dqQTIC+iVYWRmmYZAk+NWHCphoUbhrF/kiMCA3JaQiohhbRE6TtsVn4PQg35/vj\n1DunH+c/v5obnE6edozoh0qQ1dec+8PEe/5OCKZcPUoctCE4cBw3ZpwsgFMsOgbQKZYyIjS00/cg\nmPNdvrdz2oT5votDpIHj+DHjgqPIgSC0gCD9G9pMEO0vz6WB4/gx4/gABDEDCNK/ob0EUXx5Hu64\nWCCIGhCkf0N7CaL48jzccbFAEDUgSP+GrAT5HzUREUTx5Xm442KBIGpAkP4NWQkySE34bvAoBen7\n8hyFPy4WCKIGBOnf0F6C9H15jsIfFwsEUQOC9G9oL0H6vjxH4Y+LBYKoAUH6N7SXIH1fnhsDBKEC\nBOnf0F6C9H15jknhGTfjtU6NdiCICQauIEYjpoKYu0KKKIjirCpv+vYT2+/J/ejBBfqCNGkQaBEe\nFeVqbUjGTd9EA4+HSRo2xbj9TNI0tzBJ09MeZgLFvkyMmoxG/Rrq0xBSEHNXSBEFUdwGejQ3dl3N\nWOQZpS+IltlwBNHDFkcQ0p93/S2teAQxd4UUURDFbaBHc8eiozeihsgLorUpCKIHCMITUhBzV0gR\nBZHdBhqfYg1f/Mbi4bkHbvzLFRYkzEOPAhBEl5gXxNwVUkRBZLeBxnz5+/kvHENlRfqtQRAqQJAr\nKoi5K6SIgihuA20AEIQKEOSKCmLuCimiILLbQKftTBMAQWpAEAMSWE8QJujeBrqkqUSA1PpKCBKe\nLiCILiCIEeSCGDpoKABBqABB7C2I4qAxTgIEqQFBQBCOsxJccEICyQaO+2BD7kV+9DgQxAwgiL0F\nccQnC8g34MYhlcaKK9+Jatbwo8fRC0Lcz0EQGkAQHkYGhEAuyFM337tKNhxX3zik0lhxtY3IvZYf\nPQ4Hf9uzZ0/5JQ0CHcKjYs9WRSpImxIbhqTHS91ECw+TLJ0BJmk6Opmk6e0SZ+S7JCm6bDoy3QWB\n5isvCPIf/u2t9/z/i2LUNw6pNFYcQhdzK/jR46ImCK0uMS0IaWcKvaOBINSCYALHXky9W5jtG4dU\nGisuULieP8L0vC5ufeVPsYhpNIjpUyzFzqTaiLROIHiKpdxSPxrwp1j8vv/xL0feJ8z2jUMqjRV3\neqtfHD0OBDFD9ATR2hIEoRbEu/v/jZy9tlGM+sYhlQaO27Zy1aq3+dHjQBBKjL/uoQFBeK68IP93\n5IN58jM7+Tik2pgXRD8CQXQbWuyHUlegCxKREUKN4mPewYkChluDIEYx/roTGoIgctjLoIVckDoJ\nw60tLYjhZ9oAIEjINBHpggR7GbTQ/bGiIUAQo5jdC0AQXRgZEAIQxBBh3uuYalAc/TQgiBxGBoRg\nwAhi+HnXBAQJmSYiXZBgZEAIYkkQiheFFhAkZJqIdEGCkQEhAEEMAYKETBORLkgwMiAEYQqiNaKX\n9sBxyqhJP1JvSkxDGMiMOFgZLSYGjlNWQ4oMpyEPuWa6e+NdmI2uwMBxkcLeRxBlpHhuB94RxGxk\n5S5IMDIgBCCIIUCQ6HRBgpEBIQBBQuzZAiBIdLogwciAEAxQQSheWx4dQUgvJm0X2u2svPdegS5A\nEMsLQnzBjK4DQUx2AYLYTBBiFyAI8y5AEBDEvnvvFegCBIklQaAL5l3YRBBpXCz+UQo4QBDoIqJd\n2EQQaVws/lEKOEAQ6CKiXdhEEGlcLP5RCjhAEOgiol3YRBBpXCz+UQpys7KyDpNaKfZeVYT0I/Wm\nxiPFc4uQ0eiy6Qi6iHQXBDpIK9lhSBBpXCz+UQq6PR5Pc50GAZfWUmo0c1PTcYlJGjbFNPqZpGlo\nZJLG28IkDZunprWbtkVtZMWQMCSINC4W/ygFHMRTrDCBm3jqYZX7pAuweWpauqibRE4KOcY+xRLH\nxeIf+Ym4AgShAgTRxd6C6AKCUAGC6DKQBNn9TdhPFwcbzY59ziQNm2KqP2aSpraOSZqD5UzSsHlq\nTn9K3YSRASEITxBNXm5ln9M0H++LdgUyGl+LdgVy1pdHuwIZJ7eE3iYqgCBXEBBEFxAkSoAguoAg\nRoiAIFuu0Fc4hig5Ee0KZLQVRLsCOR+dj3YFMqr2RrsCHSIgCADEDiAIABAAQQCAAHNB5JeLRJ2D\npZaph7/HvFWKQd4t69e2WKYa1PXvlnmd1DAXRH65SJTxr19Wapl6+HvMW6UYdPQAOr7LMtWgXX+y\nzOukhrkg8stFokzAX1hqmXr4e8xbpRjU0I5O7LVMNRc/yLHM66SGuSB991S3AHtLLVTPxdwK6xSD\n3n21xSrV+Dd25FjodVLCXBDpchFLgAWxSj38PeatUgzq9qPqTVappvgYyrHOU6OCuSDyy0WiDhbE\nKvXw95i3SjHok+PofJ5Vqnl/U/6fNlulGDXsP8WSXS4SdbAgVqmHv8e8VYpB7RvXvlVrmWoQPoJY\nqBgF8D0IABAAQQCAAAgCAARAEBWBsck9CJ0Yh0rSjDaJ8+ms4FLE+UIkwl31JdLaVpH+i1lDhv1D\nBULJDo7K4KPT4Yi/a79+Vbo1AmRAEBWfjxzzEb/XNu002kR35+NSxPlCJFIIorWtPH3vzUuraheN\nC6DkfW5Mb/DRecj9TU7iMd2qQBCTgCAqFi1amhk8guSlpGxIQWVTX7vplr2obMrvkqcW3XHdswgd\n+lHC7IvS8lmO0R0I5f/iSeeUcoT+OnbIHBfy/Wpo8ktcCrzyQJq0VGyAOZWefbuYhRPkrVuunVwu\nbjt7NULLM8SVGC690BxzztGOLXnIjZJLhHKlRyd33cuLP+Fmhb65ZkJari+hRoAeEESJ/+YvTjm7\nRUG+vP5IzTQsSOIrnc/fjcqufrs57YZvih2upuTtLb+eIS0X/jrnD37T9fyEQLXzk+bMDFSQevZ4\n/NfBUyxxqdQAC+LMPC1mwV2dv+ZgY+YCcdtVDyM05QNxJUecT2yO8f1w9m4PN6MpyGff42aFvnEz\nMS3XFxxBzAKCKDn0I4Ru2yYK8vvFCG3DggzxoVOpqGwUQksWIpRSuWEuQl0JveJyUZDxCPUMq1g5\nHyHXt3oLbj0SaPQGBRGXSg2wINd0IzEL7qrrHOpYlCFuW3tdV93QLnElt22cT2zOBd1v/n3y7KNY\njESn03lT3yMvSGU8t4nQN24mpuX6AkHMAoIoeeraESPifyoKMv8NfLKFBRmLUFkq9w8tzUJoTGV2\nUkpKytBacbkoyCN4MnH/ErwBSqzrXTthZLYnKIi4VGqABRmDkJgFd+X7lzvufUASBN21e/V8aSW3\nbZxPbI4n3i7sSF78UZS87ezZs9+g4KPsCCL0jZuJabm+QBCzgCAKeq4/UFe3P7FTEOSF5xHansLv\n0wpB1s7Bu2FJQFzedwTxXV++MhP/sR/sq6hG5ybl9B1BhKVSA8QfR8QsuKstE5tRflCQvzx9/25p\nJbctPoIIzfH85h9zS2asJr0HEfrGzcS0/DELBDEJCKJg9/fwPhlI2SoIcmx4SW26hiB1w3c2LpqK\ngoK4cct8R27jC+P9Xw8pbPnZ42jF5PrqCWt5Qdx4Ki5VCiJmwV3l3ONpmPSwuC36atRNPdJKjji3\n2BzTMPSPFeVvJlWIn155g4/OQ+5zbyR+zm0j9I2biWkFQdxRekbtDgiiYP5z3PTZueKnWLk3js0d\n118Q9PG4hJlVwf19XhL3KdYDP0m66wxC741NerQBXXo0cdivergU85K4T7GEpUpBxCy4q9ZZw6Z8\nOGKTuC26bWFwJQdOLzTnqHhwRNLdu6XvQfKDj9z3IJOF70GEvnGzGiEt3xdfI0APCEKg7ADeUWcY\n2zY/I7K1ANEBBCFQdGOD56HlxrYFQWITEITEKzeN+qXB32CDILEJCAIABEAQACAAggAAARAEAAiA\nIABAAAQBAAIgCAAQAEEAgMD/Au2oq3DWZRfZAAAAAElFTkSuQmCC\n"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i blastDir -w 800\n",
    "\n",
    "# loading table\n",
    "infile = file.path(blastDir, 'blastn_crDNA-proto_aln_SEEDstats-byPos.txt')\n",
    "\n",
    "tbl = read.delim(infile)\n",
    "tbl = tbl[tbl$region != 'protospacer', ]\n",
    "\n",
    "# mismatches and counts, & normalized mismatches\n",
    "p.mis = ggplot(tbl, aes(pos_rel_SEED, mismatches, group=region, fill=region)) +\n",
    "    geom_bar(stat='identity', position='dodge') +\n",
    "    labs(y='Number\\nof mismatches', x='Alignment position relative to SEED start')\n",
    "p.count = ggplot(tbl, aes(pos_rel_SEED, count, group=region, fill=region)) +  \n",
    "    geom_bar(stat='identity', position='dodge') +\n",
    "    labs(y='Number of alignments\\nwith nucleotide\\nat position', x='Alignment position relative to SEED start')\n",
    "p.norm = ggplot(tbl, aes(pos_rel_SEED, mismatch_norm, group=region, fill=region)) +\n",
    "  geom_bar(stat='identity', position='dodge') +\n",
    "    labs(y='Mismatchs /\\nalignments', x='Alignment position relative to SEED start')\n",
    "\n",
    "grid.arrange(p.count, p.mis, p.norm, ncol=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Notes on plots:**\n",
    "\n",
    "* The top plot shows that some spacer-protospacer are not quiet as long as the rest, which is why the 'Number of alignment with nucleotide at position' drops at the higher positions (> ~30).\n",
    "* There are fewer mismatches in the middle of the spacer-protospacer alignments."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Array Blast vs NCBI's nt database\n",
    "\n",
    "* See the [E. coli example](../Ecoli/arrayBlast.ipynb) for conducting a spacer blast vs the nt database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "hide_input": true,
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
