{
 "metadata": {
  "name": "",
  "signature": "sha256:4936b7f9bbea0f1c0461091d9da5a0a6008a099ccc4abef2b3c997d889337743"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": []
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### Before running this notebook:\n",
      "\n",
      "* __run the [Setup](./Setup.pynb) notebook__"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from IPython.display import FileLinks\n",
      "%load_ext rmagic\n",
      "%load_ext rpy2.ipython"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "The rmagic extension is already loaded. To reload it, use:\n",
        "  %reload_ext rmagic\n"
       ]
      }
     ],
     "prompt_number": 50
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# This tutorial assumes that this IPython notebook is in the ./CLdb/examples/ecoli_ipynb/ directory.\n",
      "## these are the relative paths to everything we will be working with\n",
      "CLdbDir = \"../CLdb/\"\n",
      "scriptDir = \"../../../bin/\""
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Checking on files in the CLdb directory\n",
      "FileLinks(CLdbDir)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "../CLdb/<br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/CLdb.sqlite' target='_blank'>CLdb.sqlite</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/GIs.txt' target='_blank'>GIs.txt</a><br>\n",
        "../CLdb/array/<br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/array/Ecoli_K12_DH10B_a2.txt' target='_blank'>Ecoli_K12_DH10B_a2.txt</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/array/Ecoli_K12_W3110_a1.txt' target='_blank'>Ecoli_K12_W3110_a1.txt</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/array/Ecoli_K12_MG1655_a1.txt' target='_blank'>Ecoli_K12_MG1655_a1.txt</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/array/Ecoli_BL21_DE3_a2.txt' target='_blank'>Ecoli_BL21_DE3_a2.txt</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/array/Ecoli_K12_MG1655_a2.txt' target='_blank'>Ecoli_K12_MG1655_a2.txt</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/array/Ecoli_BL21_DE3_a1.txt' target='_blank'>Ecoli_BL21_DE3_a1.txt</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/array/Ecoli_K12_W3110_a2.txt' target='_blank'>Ecoli_K12_W3110_a2.txt</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/array/Ecoli_K12_DH10B_a1.txt' target='_blank'>Ecoli_K12_DH10B_a1.txt</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/array/Ecoli_0157_H7_a1.txt' target='_blank'>Ecoli_0157_H7_a1.txt</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/array/Ecoli_0157_H7_a2.txt' target='_blank'>Ecoli_0157_H7_a2.txt</a><br>\n",
        "../CLdb/clustering/<br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/clustering/spacers.cdhit.clstr' target='_blank'>spacers.cdhit.clstr</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/clustering/spacers.cdhit' target='_blank'>spacers.cdhit</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/clustering/DRs.cdhit' target='_blank'>DRs.cdhit</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/clustering/DRs.fna' target='_blank'>DRs.fna</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/clustering/DRs.cdhit.clstr' target='_blank'>DRs.cdhit.clstr</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/clustering/spacers.fna' target='_blank'>spacers.fna</a><br>\n",
        "../CLdb/fasta/<br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/fasta/Escherichia_coli_BL2_DE3.fasta' target='_blank'>Escherichia_coli_BL2_DE3.fasta</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/fasta/Escherichia_coli_K_2_W3.fasta' target='_blank'>Escherichia_coli_K_2_W3.fasta</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/fasta/Escherichia_coli_K_2_MG_655.fasta' target='_blank'>Escherichia_coli_K_2_MG_655.fasta</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/fasta/Escherichia_coli_K_2_DH_B.fasta' target='_blank'>Escherichia_coli_K_2_DH_B.fasta</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/fasta/Escherichia_coli_O_57_H7.fasta' target='_blank'>Escherichia_coli_O_57_H7.fasta</a><br>\n",
        "../CLdb/genbank/<br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/genbank/Escherichia_coli_BL2_DE3.gbk' target='_blank'>Escherichia_coli_BL2_DE3.gbk</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/genbank/Escherichia_coli_K_2_W3.gbk' target='_blank'>Escherichia_coli_K_2_W3.gbk</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/genbank/Escherichia_coli_K_2_MG_655.gbk' target='_blank'>Escherichia_coli_K_2_MG_655.gbk</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/genbank/Escherichia_coli_K_2_DH_B.gbk' target='_blank'>Escherichia_coli_K_2_DH_B.gbk</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/genbank/Escherichia_coli_O_57_H7.gbk' target='_blank'>Escherichia_coli_O_57_H7.gbk</a><br>\n",
        "../CLdb/loci/<br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/loci/140419_loci.txt' target='_blank'>140419_loci.txt</a><br>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 3,
       "text": [
        "../CLdb/\n",
        "  CLdb.sqlite\n",
        "  GIs.txt\n",
        "../CLdb/array/\n",
        "  Ecoli_K12_DH10B_a2.txt\n",
        "  Ecoli_K12_W3110_a1.txt\n",
        "  Ecoli_K12_MG1655_a1.txt\n",
        "  Ecoli_BL21_DE3_a2.txt\n",
        "  Ecoli_K12_MG1655_a2.txt\n",
        "  Ecoli_BL21_DE3_a1.txt\n",
        "  Ecoli_K12_W3110_a2.txt\n",
        "  Ecoli_K12_DH10B_a1.txt\n",
        "  Ecoli_0157_H7_a1.txt\n",
        "  Ecoli_0157_H7_a2.txt\n",
        "../CLdb/clustering/\n",
        "  spacers.cdhit.clstr\n",
        "  spacers.cdhit\n",
        "  DRs.cdhit\n",
        "  DRs.fna\n",
        "  DRs.cdhit.clstr\n",
        "  spacers.fna\n",
        "../CLdb/fasta/\n",
        "  Escherichia_coli_BL2_DE3.fasta\n",
        "  Escherichia_coli_K_2_W3.fasta\n",
        "  Escherichia_coli_K_2_MG_655.fasta\n",
        "  Escherichia_coli_K_2_DH_B.fasta\n",
        "  Escherichia_coli_O_57_H7.fasta\n",
        "../CLdb/genbank/\n",
        "  Escherichia_coli_BL2_DE3.gbk\n",
        "  Escherichia_coli_K_2_W3.gbk\n",
        "  Escherichia_coli_K_2_MG_655.gbk\n",
        "  Escherichia_coli_K_2_DH_B.gbk\n",
        "  Escherichia_coli_O_57_H7.gbk\n",
        "../CLdb/loci/\n",
        "  140419_loci.txt"
       ]
      }
     ],
     "prompt_number": 3
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\"\n",
      "\n",
      "BlastDir='arrayBlast'\n",
      "\n",
      "if [ -d $1$BlastDir ]; then\n",
      "    rm -r $1$BlastDir\n",
      "fi    \n",
      "\n",
      "mkdir $1$BlastDir"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "# Getting 'unique' spacer sequences (OTUs at 100% sequence identity)\n",
      "cd $1\n",
      "perl $2'CLdb_array2fasta.pl' -da CLdb.sqlite -clust -cut 1 > ./arrayBlast/spacers_cut1.fna\n",
      "\n",
      "## getting number of spacer sequences\n",
      "printf 'Number_unique_spacers:\\t'; grep -c \">\" ./arrayBlast/spacers_cut1.fna\n",
      "\n",
      "## looking at spacers\n",
      "head -n 6 ./arrayBlast/spacers_cut1.fna"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Number_unique_spacers:\t39\n",
        ">NA|spacer|NA|22|1\n",
        "TTTTACAGTGTCAACAATCGTTCCCTGATTGT\n",
        ">NA|spacer|NA|13|1\n",
        "GACAGAACGGCCTCAGTAGTCTCGTCAGGCTC\n",
        ">NA|spacer|NA|6|1\n",
        "GAAATGCTGGTGAGCGTTAATGCCGCAAACACA\n"
       ]
      }
     ],
     "prompt_number": 5
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "# spacer blast\n",
      "## Really just a thin wrapper around blastn that will blast the spacers against all genomes in CLdb\n",
      "### If you want to blast some else, just run the blast yourself (use 'perldoc CLdb_arrayBlast.pl' for more info)\n",
      "\n",
      "cd $1\n",
      "perl $2'CLdb_arrayBlast.pl' -da CLdb.sqlite -query ./arrayBlast/spacers_cut1.fna > ./arrayBlast/spacers_cut1_blastn.xml"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "Making BLAST databases...\n",
        "...making blast databases in /home/nyoungb2/perl/projects/CLdb/examples/ecoli/CLdb//arrayBlast/blast_db/\n",
        "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/perl/projects/CLdb/examples/ecoli/CLdb//arrayBlast/blast_db//Escherichia_coli_BL2_DE3.fasta\n",
        "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/perl/projects/CLdb/examples/ecoli/CLdb//arrayBlast/blast_db//Escherichia_coli_K_2_DH_B.fasta\n",
        "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/perl/projects/CLdb/examples/ecoli/CLdb//arrayBlast/blast_db//Escherichia_coli_K_2_MG_655.fasta\n",
        "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/perl/projects/CLdb/examples/ecoli/CLdb//arrayBlast/blast_db//Escherichia_coli_K_2_W3.fasta\n",
        "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/perl/projects/CLdb/examples/ecoli/CLdb//arrayBlast/blast_db//Escherichia_coli_O_57_H7.fasta\n",
        "\n",
        "Blasting sequences...\n",
        "Number of blast hits to Escherichia_coli_BL2_DE3.fasta:\t1290\n",
        "Number of blast hits to Escherichia_coli_K_2_DH_B.fasta:\t1317\n",
        "Number of blast hits to Escherichia_coli_K_2_MG_655.fasta:\t1317\n",
        "Number of blast hits to Escherichia_coli_K_2_W3.fasta:\t1317\n",
        "Number of blast hits to Escherichia_coli_O_57_H7.fasta:\t912\n"
       ]
      }
     ],
     "prompt_number": 6
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "# converting .xml to .srl\n",
      "## This retains all of the info in the xml, but produces a smaller file that is faster to load and write.\n",
      "## For large blast runs (xml files > 100 Mb), this will take a little while\n",
      "\n",
      "cd $1\n",
      "perl $2'blast-xml2srl.pl' ./arrayBlast/spacers_cut1_blastn.xml > ./arrayBlast/spacers_cut1_blastn.srl"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 7
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "# selecting DRs, so we can blast them.\n",
      "## This is needed blast filter out spacers that hit CRISPR arrays\n",
      "\n",
      "cd $1\n",
      "perl $2'CLdb_array2fasta.pl' -da CLdb.sqlite -clust -cut 1 -r > ./arrayBlast/DRs_cut1.fna"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 8
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "# DR blast, just like spacer blast\n",
      "\n",
      "cd $1\n",
      "perl $2'CLdb_arrayBlast.pl' -da CLdb.sqlite -query ./arrayBlast/DRs_cut1.fna > ./arrayBlast/DRs_cut1_blastn.xml"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "Making BLAST databases...\n",
        "...making blast databases in /home/nyoungb2/perl/projects/CLdb/examples/ecoli/CLdb//arrayBlast/blast_db/\n",
        "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/perl/projects/CLdb/examples/ecoli/CLdb//arrayBlast/blast_db//Escherichia_coli_BL2_DE3.fasta\n",
        "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/perl/projects/CLdb/examples/ecoli/CLdb//arrayBlast/blast_db//Escherichia_coli_K_2_DH_B.fasta\n",
        "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/perl/projects/CLdb/examples/ecoli/CLdb//arrayBlast/blast_db//Escherichia_coli_K_2_MG_655.fasta\n",
        "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/perl/projects/CLdb/examples/ecoli/CLdb//arrayBlast/blast_db//Escherichia_coli_K_2_W3.fasta\n",
        "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/perl/projects/CLdb/examples/ecoli/CLdb//arrayBlast/blast_db//Escherichia_coli_O_57_H7.fasta\n",
        "\n",
        "Blasting sequences...\n",
        "Number of blast hits to Escherichia_coli_BL2_DE3.fasta:\t4730\n",
        "Number of blast hits to Escherichia_coli_K_2_DH_B.fasta:\t4825\n",
        "Number of blast hits to Escherichia_coli_K_2_MG_655.fasta:\t4882\n",
        "Number of blast hits to Escherichia_coli_K_2_W3.fasta:\t4882\n",
        "Number of blast hits to Escherichia_coli_O_57_H7.fasta:\t1747\n"
       ]
      }
     ],
     "prompt_number": 9
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "# converting .xml to .srl\n",
      "## This retains all of the info in the xml, but produces a smaller file that is faster to load and write.\n",
      "## For large blast runs (xml files > 100 Mb), this will take a little while\n",
      "\n",
      "cd $1\n",
      "perl $2'blast-xml2srl.pl' ./arrayBlast/DRs_cut1_blastn.xml > ./arrayBlast/DRs_cut1_blastn.srl"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 10
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "# filtering spacer hits to arrays\n",
      "## This is based on the premise that DR hits adjacent to spacer hits signify that the spacer is hitting a CRISPR array\n",
      "\n",
      "cd $1\n",
      "perl $2'CLdb_arrayBlastDRFilter.pl' ./arrayBlast/spacers_cut1_blastn.srl \\\n",
      "    ./arrayBlast/DRs_cut1_blastn.srl > ./arrayBlast/spacers_cut1_blastn_filt.srl"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "### Summary for filtering of 'bad' DR hits ###\n",
        "Total DR hits: 1014\n",
        "DR hits with E-values < 10: 0\n",
        "DR hits with hit lengths (fraction of total) < 0.66: 0\n",
        "DR hits used for filtering: 1014\n",
        "### DR filtering of spacer blast hits ###\n",
        "Total spacer blast hits: 79\n",
        "Spacer blast hits hitting an array (filtered out): 79\n",
        "\tDeleting the blast hits to arrays\n",
        "Spacer blast hits hitting a protospacer: 0\n"
       ]
      }
     ],
     "prompt_number": 11
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "__No spacers hit any protospacers on the Ecoli genomes in the CLdb!__\n",
      "\n",
      "__Maybe we will get some hits when we blast against NCBI's nt database.__"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### Array Blast vs nt"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "# The arrayBlast wrapper only works with the genomes in CLdb.\n",
      "# We will need to do this part 'manually'\n",
      "## The blast output just needs to be in .xml format.\n",
      "## NOTE: you will need the BLAST nt database \n",
      "\n",
      "num_threads=20\n",
      "\n",
      "cd $1'arrayBlast'\n",
      "\n",
      "# spacer blast\n",
      "blastn -task blastn-short -db nt -query spacers_cut1.fna \\\n",
      "    -outfmt 5 -evalue 1e-5 -num_threads $num_threads > spacers_cut1_vNR.xml\n",
      "    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 12
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "num_threads=20\n",
      "\n",
      "cd $1'arrayBlast'\n",
      "\n",
      "# DR blast\n",
      "blastn -task blastn-short -db nt -query DRs_cut1.fna \\\n",
      "    -outfmt 5 -evalue 1e-5 -num_threads $num_threads > DRs_cut1_vNR.xml"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 13
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "!ls -thlc $CLdbDir'arrayBlast'"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "total 18M\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2  15M Jun 22 08:09 DRs_cut1_vNR.xml\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 1.6M Jun 22 08:08 spacers_cut1_vNR.xml\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2  43K Jun 22 08:06 spacers_cut1_blastn_filt.srl\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 247K Jun 22 08:06 DRs_cut1_blastn.srl\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 783K Jun 22 08:06 DRs_cut1_blastn.xml\r\n",
        "drwxrwxr-x 2 nyoungb2 nyoungb2 4.0K Jun 22 08:06 blast_db\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2  528 Jun 22 08:06 DRs_cut1.fna\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2  62K Jun 22 08:06 spacers_cut1_blastn.srl\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 227K Jun 22 08:06 spacers_cut1_blastn.xml\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 2.0K Jun 22 08:06 spacers_cut1.fna\r\n"
       ]
      }
     ],
     "prompt_number": 14
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "# converting .xml to .srl\n",
      "\n",
      "cd $1\n",
      "\n",
      "perl $2'blast-xml2srl.pl' ./arrayBlast/spacers_cut1_vNR.xml > ./arrayBlast/spacers_cut1_vNR.srl\n",
      "perl $2'blast-xml2srl.pl' ./arrayBlast/DRs_cut1_vNR.xml > ./arrayBlast/DRs_cut1_vNR.srl"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 15
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "# filtering spacer hits to arrays\n",
      "## This is based on the premise that DR hits adjacent to spacer hits signify that the spacer is hitting a CRISPR array\n",
      "\n",
      "cd $1\n",
      "perl $2'CLdb_arrayBlastDRFilter.pl' \\\n",
      "    ./arrayBlast/spacers_cut1_vNR.srl \\\n",
      "    ./arrayBlast/DRs_cut1_vNR.srl \\\n",
      "    > ./arrayBlast/spacers_cut1_blastn_filt.srl"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "### Summary for filtering of 'bad' DR hits ###\n",
        "Total DR hits: 20159\n",
        "DR hits with E-values < 10: 0\n",
        "DR hits with hit lengths (fraction of total) < 0.66: 0\n",
        "DR hits used for filtering: 20159\n",
        "### DR filtering of spacer blast hits ###\n",
        "Total spacer blast hits: 1573\n",
        "Spacer blast hits hitting an array (filtered out): 863\n",
        "\tDeleting the blast hits to arrays\n",
        "Spacer blast hits hitting a protospacer: 710\n"
       ]
      }
     ],
     "prompt_number": 16
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "__OK. It's looks like we have some hits to some putative protospacers!__\n",
      "\n",
      "__Let the investigation begin...__"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "# The srl file can be converted to a json or csv file if you would like to use tools external to CLdb.\n",
      "\n",
      "cd $1\n",
      "\n",
      "$2'blast-srl2json.pl' < ./arrayBlast/spacers_cut1_blastn_filt.srl > ./arrayBlast/spacers_cut1_blastn_filt.json\n",
      "\n",
      "# viewing some of the json file\n",
      "cat ./arrayBlast/spacers_cut1_blastn_filt.json | python -mjson.tool 2>/dev/null | head -n 40"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "{\n",
        "    \"0\": {\n",
        "        \"BlastOutput_db\": \"nt\", \n",
        "        \"BlastOutput_iterations\": {\n",
        "            \"Iteration\": [\n",
        "                {\n",
        "                    \"Iteration_hits\": {\n",
        "                        \"Hit\": [\n",
        "                            {\n",
        "                                \"Hit_accession\": \"KC765378\", \n",
        "                                \"Hit_def\": \"Escherichia coli strain 90-0009 CRISPR2a repeat region\", \n",
        "                                \"Hit_hsps\": {\n",
        "                                    \"Hsp\": {\n",
        "                                        \"1BFNxsbQbvEd6\": {\n",
        "                                            \"CLdb_array-hit\": 0, \n",
        "                                            \"Hsp_align-len\": \"32\", \n",
        "                                            \"Hsp_bit-score\": \"63.9282\", \n",
        "                                            \"Hsp_evalue\": \"3.97536e-08\", \n",
        "                                            \"Hsp_gaps\": \"0\", \n",
        "                                            \"Hsp_hit-frame\": \"-1\", \n",
        "                                            \"Hsp_hit-from\": \"286\", \n",
        "                                            \"Hsp_hit-to\": \"255\", \n",
        "                                            \"Hsp_hseq\": \"TTTTACAGTGTCAACAATCGTTCCCTGATTGT\", \n",
        "                                            \"Hsp_identity\": \"32\", \n",
        "                                            \"Hsp_midline\": \"||||||||||||||||||||||||||||||||\", \n",
        "                                            \"Hsp_num\": \"1\", \n",
        "                                            \"Hsp_positive\": \"32\", \n",
        "                                            \"Hsp_qseq\": \"TTTTACAGTGTCAACAATCGTTCCCTGATTGT\", \n",
        "                                            \"Hsp_query-frame\": \"1\", \n",
        "                                            \"Hsp_query-from\": \"1\", \n",
        "                                            \"Hsp_query-to\": \"32\", \n",
        "                                            \"Hsp_score\": \"32\"\n",
        "                                        }\n",
        "                                    }\n",
        "                                }, \n",
        "                                \"Hit_id\": \"gi|528889517|gb|KC765378.1|\", \n",
        "                                \"Hit_len\": \"922\", \n",
        "                                \"Hit_num\": \"1\"\n",
        "                            }, \n",
        "                            {\n"
       ]
      }
     ],
     "prompt_number": 17
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "!ls -thlc $CLdbDir'arrayBlast' | sort -k 9"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "total 24M\r\n",
        "drwxrwxr-x 2 nyoungb2 nyoungb2 4.0K Jun 22 08:06 blast_db\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 247K Jun 22 08:06 DRs_cut1_blastn.srl\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 783K Jun 22 08:06 DRs_cut1_blastn.xml\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2  528 Jun 22 08:06 DRs_cut1.fna\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 4.8M Jun 22 08:10 DRs_cut1_vNR.srl\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2  15M Jun 22 08:09 DRs_cut1_vNR.xml\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 638K Jun 22 08:10 spacers_cut1_blastn_filt.json\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 385K Jun 22 08:10 spacers_cut1_blastn_filt.srl\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2  62K Jun 22 08:06 spacers_cut1_blastn.srl\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 227K Jun 22 08:06 spacers_cut1_blastn.xml\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 2.0K Jun 22 08:06 spacers_cut1.fna\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 584K Jun 22 08:09 spacers_cut1_vNR.srl\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 1.6M Jun 22 08:08 spacers_cut1_vNR.xml\r\n"
       ]
      }
     ],
     "prompt_number": 18
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "# adding crRNA region to .srl (actually the DNA template, refering to it as 'crDNA')\n",
      "## How much of the adjacent DR sequences are included in the crDNA is determined by the user (default: 10bp on either side)\n",
      "\n",
      "cd $1\n",
      "perl $2'CLdb_arrayBlastAddcrRNA.pl' -da CLdb.sqlite \\\n",
      "    < ./arrayBlast/spacers_cut1_blastn_filt.srl \\\n",
      "    > ./arrayBlast/spacers_cut1_blastn_filt_crDNA.srl\n",
      "    \n",
      "## TODO: add summary info on crDNA sequences obtained"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "Decoding .srl file...\n",
        "Extracting all queries with blast hits from blast hit file...\n",
        "...Number of total unique query IDs:\t39\n",
        "Getting info from CLdb on each blast query spacer...\n",
        "Getting spacer regions from each genome...\n",
        " Processing genome: Escherichia_coli_K_2_DH_B.fasta...\n",
        " Processing genome: Escherichia_coli_K_2_MG_655.fasta...\n",
        " Processing genome: Escherichia_coli_O_57_H7.fasta...\n",
        " Processing genome: Escherichia_coli_BL2_DE3.fasta...\n",
        " Processing genome: Escherichia_coli_K_2_W3.fasta...\n"
       ]
      }
     ],
     "prompt_number": 19
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "# adding protospacer region to .srl\n",
      "\n",
      "cd $1\n",
      "perl $2'CLdb_arrayBlastAddProto.pl' -workers 24 \\\n",
      "    < ./arrayBlast/spacers_cut1_blastn_filt_crDNA.srl \\\n",
      "    > ./arrayBlast/spacers_cut1_blastn_filt_crDNA_proto.srl      "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "Retrieving proto info from blast db: 'nt'\n",
        "Retrieving proto sequences from blast db: 'nt'\n"
       ]
      }
     ],
     "prompt_number": 20
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "# aligning crRNA and protospacer\n",
      "## alignment added to the .srl file.\n",
      "## These are individual alignments between the spacer and protospacer (using clustalw).\n",
      "\n",
      "cd $1\n",
      "perl $2'CLdb_arrayBlastAlignProto.pl' \\\n",
      "    < ./arrayBlast/spacers_cut1_blastn_filt_crDNA_proto.srl \\\n",
      "    > ./arrayBlast/spacers_cut1_blastn_filt_crDNA_proto_aln.srl"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "Aligning protospacers from blast db: 'nt'\n"
       ]
      }
     ],
     "prompt_number": 21
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "# getting alignment of protospacer and crDNA and writing as a fasta\n",
      "## You can parse out the alignments you want (parse by subtype, taxon_name, etc.)\n",
      "### OR you can parse them after writing (add subtype, taxon_name, e-value, etc. to the sequence name)\n",
      "## The alignments will be oriented to the crDNA (5'-3' for the crRNA)\n",
      "\n",
      "cd $1\n",
      "perl $2'CLdb_arrayBlastGetAlign.pl' \\\n",
      "    -da CLdb.sqlite -outfmt taxon_name,subtype,evalue \\\n",
      "    < ./arrayBlast/spacers_cut1_blastn_filt_crDNA_proto_aln.srl \\\n",
      "    > ./arrayBlast/all_crDNA-proto_aln.fna"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "Decoding .srl file...\n",
        "Getting metadata info from CLdb...\n",
        "Getting alignments...\n"
       ]
      }
     ],
     "prompt_number": 22
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# let's take a look at the sequences \n",
      "!head -n 20 $CLdbDir'/arrayBlast/all_crDNA-proto_aln.fna'"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        ">crDNA|J|10|1BFNxsbQbvEd6|Escherichia_coli_BL21_DE3|I-E|3.97536e-08\r\n",
        "gggataaaccGACAATCAGGGAACGATTGTTGACACTGTAAAAgtgttccccg\r\n",
        ">proto|J|10|1BFNxsbQbvEd6|Escherichia_coli_BL21_DE3|I-E|3.97536e-08\r\n",
        ".ggataaaccgACAATCAGGGAACGATTGTTGACACTGTAAAAgtgttccccg\r\n",
        ">crDNA|J|10|1H7g3Ri62Kvuk|Escherichia_coli_BL21_DE3|I-E|3.97536e-08\r\n",
        "gggataaaccGACAATCAGGGAACGATTGTTGACACTGTAAAAgtgttccccg\r\n",
        ">proto|J|10|1H7g3Ri62Kvuk|Escherichia_coli_BL21_DE3|I-E|3.97536e-08\r\n",
        ".ggataaaccgACAATCAGGGAACGATTGTTGACACTGTAAAAgtgttccccg\r\n",
        ">crDNA|J|10|1HxIJTv3VwnRy|Escherichia_coli_BL21_DE3|I-E|3.97536e-08\r\n",
        "gggataaaccGACAATCAGGGAACGATTGTTGACACTGTAAAAgtgttccccg\r\n",
        ">proto|J|10|1HxIJTv3VwnRy|Escherichia_coli_BL21_DE3|I-E|3.97536e-08\r\n",
        ".ggataaaccgACAATCAGGGAACGATTGTTGACACTGTAAAAgtgttccccg\r\n",
        ">crDNA|J|10|1HWPcIyJKjguo|Escherichia_coli_BL21_DE3|I-E|3.97536e-08\r\n",
        "gggataaaccGACAATCAGGGAACGATTGTTGACACTGTAAAAgtgttccccg\r\n",
        ">proto|J|10|1HWPcIyJKjguo|Escherichia_coli_BL21_DE3|I-E|3.97536e-08\r\n",
        ".ggataaaccgACAATCAGGGAACGATTGTTGACACTGTAAAAgtgttccccg\r\n",
        ">crDNA|J|10|1J9r0YWTNLRMo|Escherichia_coli_BL21_DE3|I-E|3.97536e-08\r\n",
        "gggataaaccGACAATCAGGGAACGATTGTTGACACTGTAAAAgtgttccccg\r\n",
        ">proto|J|10|1J9r0YWTNLRMo|Escherichia_coli_BL21_DE3|I-E|3.97536e-08\r\n",
        ".ggataaaccgACAATCAGGGAACGATTGTTGACACTGTAAAAgtgttccccg\r\n"
       ]
      }
     ],
     "prompt_number": 23
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "__The alignment is really just for each crDNA-protospacer pair (every 4 lines):__\n",
      "\n",
      "~~~\n",
      ">crDNA\n",
      "ATAGACA\n",
      ">proto\n",
      "ATAGACA\n",
      "~~~\n",
      "\n",
      "__If you wanted each pair individually, you could use something like the following:__"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# WARNING: this will produce a lot of files!\n",
      "!split -l 4 -d -a 4 $CLdbDir'arrayBlast/all_crDNA-proto_aln.fna' $CLdbDir'arrayBlast/crDNA-proto_aln_split'\n",
      "\n",
      "## Let look at one of them\n",
      "!find $CLdbDir'arrayBlast/' -name 'crDNA-proto_aln_split[0-9]*' | head -n 1 | xargs -I % cat %"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        ">crDNA|J|6|22ByLIOpuxUAQ|Escherichia_coli_BL21_DE3|I-E|3.97536e-08\r\n",
        "gggataaaccGGCGACCACGCAGAGCGGCGACAACCTGACGTTgtgttccccg\r\n",
        ">proto|J|6|22ByLIOpuxUAQ|Escherichia_coli_BL21_DE3|I-E|3.97536e-08\r\n",
        ".ggataaaccgGCGACCACGCAGAGCGGCGACAACCTGACGTTgtgttccccg\r\n",
        "find: `standard output': Broken pipe\r\n",
        "find: write error\r\n"
       ]
      }
     ],
     "prompt_number": 36
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Let's remove those files\n",
      "!find $CLdbDir'arrayBlast' -name 'crDNA-proto_aln_split[0-9]*' | xargs -I % rm %"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 38
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "# Looking at the alignments can be very useful, but it can help to extract what we specifically want and summarize.\n",
      "## Let's pull out the potential PAM regions from the alignment \n",
      "## We are going to assume that the PAM is:\n",
      "### 4bp (really 3bp) long \n",
      "### adjacent to the 5' end of the protospacer (right side of the alignment)\n",
      "\n",
      "cd $1\n",
      "perl $2'CLdb_arrayBlastGetPAM.pl' -PAM 1 4 -f - \\\n",
      "    < ./arrayBlast/all_crDNA-proto_aln.fna \\\n",
      "    > ./arrayBlast/all_crDNA-proto_aln_PAM.fna\n",
      "\n",
      "head ./arrayBlast/all_crDNA-proto_aln_PAM.fna"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        ">proto|G|3|21WKGmhD8OS0s|Escherichia_coli_O157_H7|I-E|3.97536e-08\n",
        "gtgt\n",
        ">proto|D|3|acrDRh1RR5Na|Escherichia_coli_K-12_DH10B|I-E|1.08991e-08\n",
        "gtgt\n",
        ">proto|J|7|2PvuEIE5YqPoo|Escherichia_coli_BL21_DE3|I-E|3.97536e-08\n",
        "gtgt\n",
        ">proto|B|2|1z8G9sfx4ozre|Escherichia_coli_K-12_MG1655|I-E|1.08991e-08\n",
        "gtgt\n",
        ">proto|F|3|4bpkkVPh0mck|Escherichia_coli_K-12_W3110|I-E|1.08991e-08\n",
        "gtgt\n"
       ]
      }
     ],
     "prompt_number": 39
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#summarizing each PAM sequence\n",
      "!egrep -v \"^>\" $CLdbDir'./arrayBlast/all_crDNA-proto_aln_PAM.fna' | sort | uniq -c | sort -k 1 -n -r "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "   1201 gtgt\r\n",
        "    371 gagt\r\n",
        "      6 \r\n",
        "      4 gagc\r\n",
        "      3 aatt\r\n",
        "      3 aagt\r\n",
        "      2 cgtg\r\n",
        "      2 cggt\r\n",
        "      2 acgg\r\n"
       ]
      }
     ],
     "prompt_number": 40
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# and if we reverse the sequence...\n",
      "!egrep -v \"^>\" $CLdbDir'./arrayBlast/all_crDNA-proto_aln_PAM.fna' | rev | sort | uniq -c | sort -k 1 -n -r "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "   1201 tgtg\r\n",
        "    371 tgag\r\n",
        "      6 \r\n",
        "      4 cgag\r\n",
        "      3 ttaa\r\n",
        "      3 tgaa\r\n",
        "      2 tggc\r\n",
        "      2 gtgc\r\n",
        "      2 ggca\r\n"
       ]
      }
     ],
     "prompt_number": 41
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "__We get the (less prevalent) PAM region ('TGT') for known I-E CRISPR systems as shown in [Rorek et al., 2013](http://www.annualreviews.org/doi/abs/10.1146/annurev-biochem-072911-172315)!__"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### spacer (crDNA) - protospacer mismatch distribution\n",
      "\n",
      "Let's look at potential spacer targeting by assessing mismatches of the crDNAs and protospacers"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "# This script will make 2 summary files of mismatches for each crDNA-protospacer alignment\n",
      "# Given that spacers can vary in length, the alignment is relative to a user-defined 'SEED' region.\n",
      "# I'm going to use the default SEED region: '-SEED -8 -1'\n",
      "## This SEED region covers the last 8 bp of the alignment\n",
      "\n",
      "cd $1\n",
      "perl $2'CLdb_arrayBlastSEEDstats.pl' \\\n",
      "    --SEED -8 -1 \\\n",
      "    < ./arrayBlast/all_crDNA-proto_aln.fna \\\n",
      "    -prefix ./arrayBlast/all_crDNA-proto_aln\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "SEED summary file written: './arrayBlast/all_crDNA-proto_aln_SEEDstats-summary.txt'\n",
        "SEED by-position table written: './arrayBlast/all_crDNA-proto_aln_SEEDstats-byPos.txt'\n"
       ]
      }
     ],
     "prompt_number": 52
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%R -i CLdbDir -w 800\n",
      "# plotting mismatches across the protospacer alignments\n",
      "library(ggplot2)\n",
      "library(gridExtra)\n",
      "\n",
      "# loading table\n",
      "filename = './arrayBlast/all_crDNA-proto_aln_SEEDstats-byPos.txt'\n",
      "fullname = paste(c(CLdbDir, filename), collapse='')\n",
      "\n",
      "tbl = read.delim( fullname )\n",
      "tbl = tbl[tbl$region != 'protospacer', ]\n",
      "\n",
      "# mismatches and counts, & normalized mismatches\n",
      "p.mis = ggplot(tbl, aes(pos_rel_SEED, mismatches, group=region, fill=region)) +\n",
      "    geom_bar(stat='identity', position='dodge') +\n",
      "    labs(y='Number\\nof mismatches', x='Alignment position relative to SEED start')\n",
      "p.count = ggplot(tbl, aes(pos_rel_SEED, count, group=region, fill=region)) +  \n",
      "    geom_bar(stat='identity', position='dodge') +\n",
      "    labs(y='Number of align-\\nments with position', x='Alignment position relative to SEED start')\n",
      "p.norm = ggplot(tbl, aes(pos_rel_SEED, mismatch_norm, group=region, fill=region)) +\n",
      "  geom_bar(stat='identity', position='dodge') +\n",
      "    labs(y='Mismatchs /\\nalignments', x='Alignment position relative to SEED start')\n",
      "\n",
      "grid.arrange(p.count, p.mis, p.norm, ncol=1)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "display_data",
       "text": [
        "NULL\n"
       ]
      },
      {
       "metadata": {},
       "output_type": "display_data",
       "png": "iVBORw0KGgoAAAANSUhEUgAAAyAAAAHgCAMAAABq2fnHAAAC9FBMVEUAAAAAv8QBAQECAgIDAwME\nBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUWFhYX\nFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMlJSUmJiYnJycoKCgpKSkqKior\nKyssLCwtLS0uLi4vLy8wMDAyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6Ojo7Ozs8PDw9PT0+Pj4/\nPz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tMTExNTU1OTk5PT09QUFBRUVFS\nUlJTU1NUVFRVVVVWVlZXV1dZWVlaWlpbW1tcXFxdXV1eXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVm\nZmZnZ2doaGhpaWlra2tsbGxtbW1ubm5vb29wcHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6\nenp7e3t8fHx9fX1+fn5/f3+AgICBgYGCgoKDg4OGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Pj4+Q\nkJCRkZGSkpKTk5OUlJSVlZWWlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKj\no6OkpKSlpaWmpqanp6eoqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2\ntra3t7e4uLi5ubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJ\nycnKysrLy8vMzMzNzc3Ozs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc\n3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u7u7v\n7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4dm34+Pj5+fn6+vr7trH7+/v8/Pz9/f3+/v7///8q\n/YX8AAAgAElEQVR4nO2de3wU5b3/txpqbGJYDlFBscRKSX9SoURBuRg4yuF4OV6gklOPJb/+Wqm1\netqCaO0pB6kWOfg7WAkSrlHsCbb+QJCqQW4KQhSKEJMmEuSSbLK5AsnGJGeTff75PbMzs9m5z+5+\nJ5nlfN4vX7Mzz3yf73zdPG9mb/OMhwEADPEMdAEAuBkIAoAJEAQAEyAIACZAEABMgCAAmABBADAB\nggBgAgQBwAQIAoAJEAQAEyAIACZAEABMSFSQGi0N53QaE6O+ljpjXTt1xhqfjzxlgDxlbR11xpoL\nfvKU9XaCSMa/JQ4I0txJ/HTV1DSSZ6zvoU9JP/ZC9II0UGes6ab/69jKSDL+LYEgdCkhCBXJIUjJ\nbdkC5v11CocgZEAQE2hFMMJEkJELj5VzzPvrFA5ByIAgJtCKYISJINcEbPTXKbxPkK8p0Tb0Q8BX\nSmo0DYkH1PR7AEXVAxFwcQny4gs91v0hCASJPyC5BZmcPnh0Yu9BIIhDAe4Y3okHJLcg5SLh9T1l\nLPj8ypUfhratf60tvJCCIAgEiT8gGQVZE7Xe4wuGH3vXLi5j/j/ztepidvj98EIKgSAQJP6AZBTE\nG1nzzxmUPmiOn6+FekvKWMWrbxS37t7P/GvDC958ZN++fWXnNASC8ppqsGob+iFA9Tc5p2lIPOBc\nvwdQVO2GACPaDff00TjggjyQ72f+H84Kr+8sY6ePs2Ob3vmcdb0cXvDGPVu3bj3arqGzR15TDVZt\nQz8EqP4m7ZqGxAPa+z2Aomo3BBjRYbinj+YBEuQ/I2sZQgmNg2VBON0vfcBPHmvCCykIL7HwEiv+\ngGR8idXHt4X3Ge+Ln2JxQfYdYtWbTmxmf30/vJCCIAgEiT8g6QRR/LrkTe/cRXO9f5IFCbyxdmNT\n6J3XNrWHFxDEqAGC2A5IOkGUvy45sXzh8hMW/SEIBIk/IOkEsfXrEgUQBILEH5B0gkT9uiRne46I\neX8IAkHiD0g6QaJ+XVLauLtUYLd5fwgCQeIPSDpBon5dEgyODHJaMsz7QxAIEn9A0gkSRUqKJ0Ug\nz7w/BIEg8QcknSBZYcZMe1H4FHeGnf4QBILEH5B0gqybuvXI1tsL/nLPo3b7QxAIEn9A0gkyslYY\n86NZYAQ+xYrhz9zvAe4Y3okHJJ8gpXxx6BpWP4KVNpaKmPeHIBAk/oCkE2TdVQteXnBVwe5rfi9s\nVXUFVqzpNu8PQSBI/AFJJwj77Ndzn/6Ele8X1henNjw3Yexj5v3rGzVc6JbXVINV29APAaq/SaOm\nIfGAxn4PoKjaDQFGtBju6aPeKSWUmPyad+iR0PBTJzLN++MMgjNI/AFJdgZRvS/3njw4jvnwRaH1\nn7nfA9wxvBMPSDJBVO/L52XfsKZ64izz/hAEgsQfkGSCqAgWF/dULrtg3h+CQJD4A5JMkDEycoM8\nq4kJEASCxB+QZIIckRE3+2Y1MQGCQJD4A5wTpCglMS8i6L3E2ig+RM9qYggEgSDxBzgnyPmq+HzQ\noPq5+6N5eXn3DRM3omc1MQSCQJD4AwgFWZ32xDP1d2XcXtk+e3C+5wg/g3TkDxm+hK2+fE76/Zbv\nFOwKckv+/LuLJmwTN6JnNYEgZn/mfg9wx/BOPIBSEM9T5T8fV//whGVXfPiLsCAvph1Y59m72vPm\nOs+nVIKkNrbdyeqmiBvRs5pAELM/c78HuGN4Jx5AKUhqD5s2JDtr+D9PY38LC/JILuu9/A+r09gZ\nzz4qQbL2spzmwFBpC7Oa2Pwz93uAO4Z34gGUgqQz9ujUpnX/vth7YEFYkKVppRv4GSSdUpBVl536\nzfjb/0HaClXuqugVV/eUsb6J3TG7u2kDBIk/IDFBaqZnTN/XPCN9VliQwFzvsN8xWkHY6Y7g6yvO\nievl4zLHXzlWuEA9PLt738TumN3dtAGCxB8QvyARShYH3/Gcid8Ic0GiuXVhN+teOIlJs7v3Tewu\nz+4OQXQbIEj8AQSCVE+9fsRSMj/MBPE28IU/Mnl138TumN3dtAGzu8cf4O7Z3euZ4mvzhwv44pWH\nZEH6JnaXZ3fH/UF0G3B/kPgD3H1/kCG+b14II27+eNDY+8d67s7LywsL0jexO2Z3N23AS6z4A+J/\nifWVGgcEmT/kEm8YcbMoQliQvondMbu7aQMEiT/A3YIw9kDM/SEIBCEMcLsgtn7hrgCCQBDCALcL\nYusX7gogCAQhDHC7ILZ+4a4AgkAQwgAnBQn9Zvytd9VWDZ48efKCyHLCVIuJ35SCKH7hXjJFcXkh\nBDEKgCBkAU4K8tHUEPuvx6pyhfW+5effNr9plFIQxS/cRy4pi7ofGwQxDIAgZAGkgmz+0SOTt5yb\ndef0z8NrldeXBHta1YKwh9+LQRDFL9yv7jB3A4LoN0CQ+ANoBZkeOvjQ7xaxnfeG19j+R8bkVVZ5\nc3NzV0hLQZCF62MQRPEL9+f67scGQcwCIAhZAK0gi9jJ2fklrPM74bVj1Sz05kTNGeRf3tcMahNB\n+sjOHn1J5H5sEMQsAIKQBdAKsoRrwc8gH9wTXtuW18uqxqsFqciO5T1IFOXln4Rvx/YJBLEKgCBk\nAfSCnHtwBn8PIqwFfzv5lqmHw59fzY4sJ06xGOCGguAehbYDIAhZAKkgJCgFuel4ZBX3KLQdAEHI\nAtwuyJL8zr4NW/coxO0PcPsDwoD4b3/wlZqE1RBRCpLrvWyU5ftyBTiD4AxCGOD2M8gxEWEV9yi0\nHQBByALiF+S/1TgiCGM9tSFxBfcotB0AQcgC4hfka2ocEeTstLQhpVOqxY0NdubPhiAQhDDA7YLM\neLwjKzj/DnFjevqNT2w9B0EsAyAIWYDbBflGC8tiDWnSVnfp/53jvQ2CWAVAELIAtwsy9m0uyLbv\nihsdHy59cPj/+gkEsQqAIGQBbhdkd+bsjFlXvStufP2Kx7dYzq0CQSAIYQCpIPFdIWUqCGtct6TQ\nJ60vfXBkzs+KLG5EAkEgCGEAqSDxXSFlLkhP+a4vQpGtUOUT6SZTL0IQ3QYIEn8AqSDxXSFlKsjR\n72SOz7xZuiBk+zNTM6Yu2gNBrAIgCFkAqSDxXSFlKsgtwnzV83PFjZuf+ktb1L7g8ytXfojbH0AQ\nJwNIBYnvCilTQTKi5qtW4/8zX+D2BxDEyQBSQeK7QspUkDmr+WLVg7qRFa++UdyK2x9AECcDSAWJ\n7wopE0EefniWZ/zs8Zc+qht5+jg7tklx+4M/rVq16uNuDT0heU1VsrahHwJUf4NuTUPiAd39HkBR\ntSsDJILaUaVB9RsPHUFIiBYkarLqMFVdgRVruqMCul9S3P7gzAkOrgfB9SB0ARLnYr8epD8E4fir\nBMT1xakNz00Y+5i8a98hVr0Jtz/ASywnA0hfYjkgyBOerFEccWPokdDwUycy5X2BN9ZubMLtDyCI\nkwFuFyT9QNSG9+TBccyHSRssAyAIWYDbBck5FbUxL/uGNdUTLWayhiAQhDAgfkGcQinI/qvnLeKI\nG63FxT2Vy+rM+0MQCEIY4HZBJuUsfJYjrGJeLNsBEIQswO2CZPbdWxfzYtkOgCBkAW4X5JebojZs\nzYsFQSAIYYDbBZl06YhszIsVawAEIQtwuyDlIuJGyW3Z1rZAEAhCGOB2QRSMXHgMd5iCIP0a4HZB\nJoqIG9fY+VkwBIEghAFuF+TAgQP7i3M3ixsv4g5T9gIgCFmA2wUJ03CT+Dg53cYdpmq1tHTKa6rB\nqm3ohwDV36BW05B4QG2/B1BU7coAiUadYaWBXgY9dAQ5LH03qHjHbkSdX8O5LnlNNVi1Df0QoPob\n+DUNiQf4+z2AompXBkg0a0eVBoufeFChfQ+SM+hJeTMyk7UxeImFl1iEAW5/iXVAoFyyQjGTtREQ\nBIIQBrhdEAWKmayNgCAQhDDA3YLkyoibypmsDYAgEIQwwN2CvB1myTfGiZuKmayNgCAQhDDA3YII\nXPiV9z+C4qpiJmsjIAgEIQxwvSBbvvnA6chG9EzWRkAQCEIY4HJBzjw4cmvUZvQUJ0ZAEAhCGOBu\nQZb/3YLoyXgVU5wYAUEgCGGAuwXxXJIqIm4qpjgxAoJAEMIAdwvikxE3FVOcGAFBIAhhgLsFUaGY\n4iQa3P7AtAGCxB+QVIJETXGiBLc/MG2AIPEHJJUgUVOcKMHtD0wbIEj8AUkliGKKk2jk2x8U8Bdg\nu8ySqwartqEfAlR/A6ZpSDyA9XsARdVJEWBCm9lOOsxeYhlNcSLf/qAzEAic8Wlo6dS2JUgTecaG\nXvKUfj95ylA9dca6RuqMvm76v46djAN3wZSM4QVTtm9/QIatU25M1PfQp6wjTxnyUWesbaDOWNNN\n/9dJjpdYhti+/QEZEIQKCBIj8QgSjU7hEIQMCGICyfi3xAFByj4kfrpqaupqqTNWv0udscZHPppr\ntp8iT0lv8e4K8pS2iiQZ/5YkKogOx4qsYwYc/7KBrsAOz/XTRzUJ8eoXA12Bk0AQNwNBBhwI4mYg\nyIDjgCAnS+hzktNaPNAV2KGoY6ArsMH2fvpGYmBwQBAALh4gCAAmQBAATKAXJPpqEbeyp8z9ZQbf\nWl9w1u1Vdm1aW9js9iITgl6Q6KtF3Env2sVl7i+zYjurWe32Kg/tZoffcXuRCUEvSPTVIu4k1FtS\n5v4yaxtYS6Hbq6w/z47sdHuRCUEviHy1iJvZWZYMZZ4tqHR/lX98odn9RSYAvSDy1SJuhgvi+jJD\nJWt9rq+ys5dVb3R7kQlBL0j01SJuhQvi+jKPb+51/5P53mF2ep3bi0wIBz7FirpaxK1wQVxf5pbl\nK1e+7vYqz28ofLXW7UUmBL4HAcAECAKACRAEABMgiJrQ6Mxuxo6MYaU5drukBA12CClSghaJ+KH6\nEunFKtIfnTF46D9VMpbpEaiKPHo9ntRbdxl2M64RmAJB1Hw6fNRfwqO2cbvdLoaDT0iRErRIpBBE\nLzY6fc91z56onT8mxDI/aOH0RB69e1u+XJH+iWFVECQ+IIia+fOfzY+cQdZlZa3PYuWTX7z2+p2s\nfNKvMifvv/mKJxnb+720mWfl9hmekW2MFf34Ee+kCsb+NHrwLD8L/nRI5nNCCr5zd47cKnXgHMtd\ncpOURRDk1esvn1ghxc5cxdjSPGknR0gvduec8pznltzbwjJLxXLlR+8Rvnjm+8KqeGyhm5hWOJZY\nI4gZCKKi97qjx7ydkiCfXXmwZgoXJP359qduY+WXvt6Uc/WXBzz+xsytzT+bJreL/zoXDXrF/9S4\nULX3vab8PFacffJw6heRl1hSq9yBC+LNPy5l4Yc6fdmehvxHpdiV9zE26S1pp0BKUOrOCX535o6A\nsKIryEffElbFY/NuUlrhWDiDxAkEUbH3e4zduEUS5NcLGNvCBRkcZMeyWfkIxhbOYyyrav1sxjrS\neqR2SZCxjHUPrVw+lzH/13uKbzgYauiKCCK1yh24IJd1MikLP1THKdY2P0+Krb2iwzekQ9opxKYE\npe7CRucr/5g58xAXI93r9V7b9xgWpCp8axfx2LyblFY4FgSJEwii4rHLhw1L/YEkyNyX+YstLsho\nxsqzhf/Ys4sYG1W1JCMrK2tIrdQuCXI/X4zftVC4X0S6r6dw3PAlgYggUqvcgQsyijEpCz9U8N9u\nvuNuWRB2645Vc+WdQmxKUOrOF10d3JF1qYdY5paTJ09+ySKPUWcQ8di8m5RWOBYEiRMIoqT7yt0+\n3670dlGQp59ibGtWeEwrBCmcxYdhaUhq7zuDBK+sWM7fwPgHBSur2akJK/rOIGKr3IGFzyNSFn6o\nTeObWFFEkN8/ftcOeacQy88gYne+/trfCy3TVpm9BxGPzbtJacPnLAgSHxBEyY5v8TEZytosCvLJ\nVaW1uTqC+K7a3jB/MosI0sJ7FnkKGp4e2/vF4JLmHz7Elk2sqx5XGBakhS+lVqUgUhZ+qBW3B+on\n3CfFsr+NuLZb3imQ0iJ159QP+W1lxSsZldKnV12RR+/ellMvp38qxIjH5t2ktKIgLQP0jCY5EETJ\n3F8KyydnS59iFVwzumCMVhD27pi06Sci431OhvAp1t3fz7j1c8beHJ3xQD278ED60J92CynmZAif\nYomtSkGkLPxQrTOGTnp72EYplt04L7JTgKcXuwtU3jMs47Yd8vcgRZFH4XuQieL3IOKxebcaMW34\nWOEaQcxAEDPKd/OBOs1ebFGes7WAAQGCmLH/mvrAvUvtxUKQixIIYsrz1474ic1fckOQixIIAoAJ\nEAQAEyAIACZAEABMgCAAmABBADABggBgAgQBwAQIAoAJEAQAEyAIACZAEABMgCAAmABBADABggBg\nAgQBwAQIAoAJEAQAEyAIACZAEABMgCAAmABBADAhPkFqtNQGanVaE6TegZQX6HPWOlBnUyt9Tp+P\nPue5RvqcfltRxCYYQCcIc0AQB578hm76nLUN9DnPtdPnrK+jzxlooc/ZYGsoEZtgAAQhAILQAkHs\nAkFIgSCxA0EIgCC0QBC7QBBSYhHkKyWGcRBED51yIQgtEMQKYhMMgCAEQBBaIIhdIAgpECR2IAgB\nEIQWCGIXCEIKBIkdW4J0bVpb2Bzatv41+T6QOuVCEFogiBUOWhGFLUEO7WaH36kuZofflxp0yoUg\ntEAQKxy0IgpbgtSfZ0d27t7P/Gv5xonjx49XNmtoYdq2hDnvQMoe+pxO1BnopM/Z2mo/ViWIYVxX\nO0FhKmw9nX6HzZCw+R7kjy80v/M563qZr27bsGFDaacWptOWKN0OpAzR53SizmAPfc6uLvuxKkEM\n43qDBIWpsPV0tjrqRQQzQaq6AivWdPOVzl5WvfEDfgZZI+3ROeHhJRYteIllRT/YwUwFWZza8NyE\nsY/xtfcOs9PrTmxmf8V7EF0gCC1JIsjQI6Hhp05k8rXzGwpfrQ2989om+ZbhOuVCEFogiBX9oYep\nIN6TB8cxX4beLp1yIQgtEMQKp5RQYiLIvOwb1lRPnKW3S6dcCEILBLHCKSWUmAgSLC7uqVx2QW+X\nTrkQhBYIYoVTSigx/Zi3pzakv0OnXAhCCwSxwgkdtJgIcnZa2pDSKdV6u3TKhSC0QBArnFJCiYkg\nMx7vyArOv0Nvl065EIQWCGKFU0ooMRHkGy0sizWk6e3SKReC0AJBrHBKCSUmgox9mwuy7bt6u3TK\nhSC0QBArnFJCiYkguzNnZ8y66l29XTrlQhBaIIgVTimhxOxTrMZ1Swp9unt0yoUgtEAQK5wRQk18\nVxTWavExn05rgjQ6kLKbPqevgT7nuXb6nPX19mNVghjGdbQSFKai0dZQIjbBABNBSqaMEdDbVefX\n0MAatI2J0kyfsiVIn7OhiT5nWwd9zsZG+7EqQQzjOs8TFKbC1p+9zikllJgIMnJJWTlHb5fOCQ8v\nsWjBSywrnFJCiYkgV3cY7tIpF4LQAkGscMAGHZSC3HQ8auO5F3qMeumUC0FogSBWmI7ropSEtOhD\nKciS/E55NTt79CWDR2dnZ+v10ikXgtACQawwHdfnqyjsYGpBcr2XjZKcKJeBIJZAEFriEWR12hPP\n1N+VcXtl++zB+Z4j/AzSkT9k+BK2+vI56fcHqQQ5JiJudBbuYUXLuyCIJRCElrgE8TxV/vNx9Q9P\nWHbFh78IC/Ji2oF1nr2rPW+u83xKJYjiF+7zcv7K9t/yIwhiCQShJS5BUnvYtCHZWcP/eRr7W1iQ\nR3JZ7+V/WJ3Gznj2UQmi+IX70BN8ccILQSyBILTEJUg6Y49ObVr374u9BxaEBVmaVrqBn0HSKQVR\n/MJ91Md8sX8UBLEEgtASryA10zOm72uekT4rLEhgrnfY7xitIIpfuP8x8+fL/jVzIwSxBILQktCn\nWCWLg+94zsRvhKkgil+4t1UsmbfoaJtOJwiiBILQkpAg1VOvH7GUzA+VIFG/cA8GRwY5LcK0P8G3\n1hecxezuhkAQWtz7TXrUL9xTUjwpAnl8vWI7q1mN2d0NgSC0uFiQnvJdX8if886ItNY2sJZCeXb3\nj99///3PLmhhOm2JEnAgZS99Tifq7Oqmz9neZj9WJYhhXLCToDAVtp7OJuXI/UqNI4Ic/U7m+Myb\nT0Q3iW/SzxZUyrO7QxCdpPQpIYgFAyLILQu7Wff8XHGj/NG8vLz7hvG1UMlaH8Ps7obgJRYt8bzE\n6h9BMhr4wj9YsiV//t1FE7bxteObexnD7O6GQBBa3CvInNV8sepBcSO1se1OVjeFr21ZvnLl65jd\n3RAIQotbBXn44Vme8bPHX/qouJm1l+U0B4bq9dIpF4LQAkGsiFWQ0G/G33pXbdXgyZMnL4gsJ0wt\ntS9IkYy4ueqyU78Zf/s/QBBLIAgtzgjy0dQQ+6/HqsLvsPuWn387YFsQjr9KQNo43RF8fcU5CGIJ\nBKGFSJDNP3pk8pZzs+6c/nl4rfL6kmBPq1oQ9vB7MQjyhCdrFEfPFghiDAShhUqQ6aGDD/1uEdt5\nb3iN7X9kTF5llTc3N3eFtBQEWbg+BkHSDxjaAkGMgSC0UAmyiJ2cnV/COr8TXjtWzUJvTtScQf7l\nfc3wNhYk55SRLRDEBAhCC5UgS7gW/AzywT3htW15vaxqvFqQiuxY3oPsv3reIo6OLRDEBAhCC6Ug\n5x6cwd+DCGvB306+Zerh8OdXsyPLiVM+MfVDJciknIXPcnRsgSAmQBBa3Po9CGOZzUa2QBATIAgt\n7hXkl5uMbIEgJkAQWtwryKRLR2RH5opT2KIEk1crwOTVtMQzeXX/CKKYK05hixKdyehx+wNacPsD\nK5RD8r/VOCKIsS1KdE54eIlFC15iWaEckl9T44ggE0WkraquwIo13RDEEghCi3sFOXDgwP7i3M3i\nxuLUhucmjH0MglgCQWhxryBhGm4SH4ceCQ0/dSITglgCQWhxuSCHM8RH78mD45gvA4JYAkFoca8g\nwhuQnEFPihvzsm9YUz1xFgSxJPkEsR7+F4Eg8V0hZSrIAYFyad6fYHFxT+WyCxDEEghCC5Eg8V0h\nZSqIXXTKhSC0QBArLAWJ7wopE0FyZYSNnO05IhDEEghiD7s5qd6DxHWFlIkgb4dZ8o1xwkZpY6kI\nBLEEgtijnwWJ7wopE0EELvzK+x/yHd2iLrndU8YwebUhEMQe/SxIfFdImQuy5ZsPnJbX+y657V27\nuIxh8mpD/kcIYtDFxYLEd4WUmSBnHhy5tW+r75LbUG9JGZMnr4YgGiCIPfpZEBKiBVn+dwuib5cT\nfcntzjImT169bcOGDaWdWphOW6J0O5AyRJ/TiTqDPfQ5u7rkNdVQ1YnVRBh06Q3aP771UUVsPZ2t\n/S+I55JUEXEz+pJbLog8efWJ48ePVzZraGHatoQ570DKHvqcTtQZ6KTP2doqr6mGqk6sJsKgS1e7\n/eNbH1XE1tPp739BfDLiZvQlt1wQTF5tCF5i2cP6qCJufYmlIvqSWy4IJq82BILY4yITxOSSW51y\nIQgtF4EgdlNocPEt2KIxueRWp1wIQgsEscIpJZTgklsCIIguF7sgJuiUC0FogSBWEJtgAAQhAILo\ncnELUs/8RrsgiAoIosvFLcgQ3zfF2+3q7dQpF4LQAkGscE6KaAwFmT/kEm8YvZ065UIQWiCIFc5J\nEY3Je5AHjHfplAtBaIEgVtDLoIfpm/QeX1B/h065ZoKovuO08T8few+7QBABCGIXE0H8cwalD5qj\n+1Zdp1wIQgsEscIpJZSYvcTK9zP/D3Wn/alv1NDEmrSNEqrhbhiXSA+7nAsSJpNoaqbP2d5Bn7M5\n8idSDVWdWE2EQZeuNsPD2U2hocXO/0y9U0ooMREkQ/ixYuNgvV06PkedQTT//OMMEjs4g1jhkBEq\nTAT5tvDj9vfj+C0WBCEAgljhkBEqTAR50zt30Vzvn/R26ZQLQWiBIFY4pYQSs0+xTixfuPyE7h6d\nchMShEApu1yMgtgdeBAkdpz4LRYEIQCCWEFsggEQhAAIYq8sCCJiOdytfYhdELs9/kcIYjAS3SbI\nV5oIEQgCQSyBIFYQm2AABCEAgsRXFgSBIHFjIoj12IUgZEAQAiBIfHVeXILYn90dghAAQaxwRgg1\n9gWxP7t7fwgSt1IQRDdFHAcRgSAy9md3hyAEQBArnBFCjX1B5NndCxYtWrTLNFI1VrUNlgHahpiP\ncRGjGlZ2GixTxHGQxOtUCxLTs9BmHUKBfUHk2d07A4HAGZ+GOlanbUyUJgdSBulz1jXS5zzfTp/T\n76fP2XGOPmejnaFU65wU0dgXBLO7GzLQ34PYpe8lFh1uucutU8TwKRZmdzcCgtCSnIJEo1MuBKEF\nglhBbIIBZIKc2XaG+GniOPAHrfyAPmeNjz7l0YP0OX0O1PnRMfqc9v7sxCYYEJ8gOnQvMpgiyGWc\nXDnQFdjjw63WMW6gOLabxiYdEMSlQBB3AEFcCgRxB2SCBDf0UKVylLokGXjH9w90BfbYXTnQFTgL\nmSAAXIxAEABMgCAAmEAlSPTVIi5mT1kyVBp8a33B2SSos2vT2sLmJKgzEagEib5axLX0rl1clgyV\nVmxnNauToM5Du9nhd5KgzkSgEiT6ahHXEuotKUuGSmsbWEthEtRZf54d2ZkEdSYClSDy1SIuZ2dZ\nclR6tqAyKer84wvNSVFn/FAJIl8t4nK4IElQaahkrS8Z6uzsZdUbk6DORKASJPpqERfDBUmCSo9v\n7k2KZ/S9w+z0uiSoMxHIPsWKulrExXBBkqDSLctXrnw9Ceo8v6Hw1dokqDMR8D0IACZAEABMgCAA\nmABBogiNzuxm7MgYVppjt0uK0Y/8hRQpQYtE/FB9ifRiFemPzhg89J8qGcv0CFRFHr0eT+qtuwy7\nGdcIrIEgUXw6fNRfwqO2cbvdLoaDT0iRErRIpBBELzY6fc91z56onT8mxDI/aOH0RB69e1u+XJH+\niUE3CJIQECSK+fOfzY+cQdZlZa3PYuWTX7z2+p2sfNKvMifvv/mKJxnb+720mWfl9hmekSeZmqQA\nABYWSURBVG2MFf34Ee+kCsb+NHrwLD8L/nRI5nNCCr5zd47cKnXgHMtdcpOURRDk1esvn1ghxc5c\nxdjSPGknR0gvduec8pznltzbwjJLxXLlR+8Rvnjm+8KqeGyhm5hWOJZYI4gPCNJH73VHj3k7JUE+\nu/JgzRQuSPrz7U/dxsovfb0p5+ovD3j8jZlbm382TW4X/3UuGvSK/6lxoWrve035eaw4++Th1C8i\nL7GkVrkDF8Sbf1zKwg91+rI9DfmPSrEr72Ns0lvSToGUoNSdE/zuzB0BYUVXkI++JayKx+bdpLTC\nsXAGSQQI0sfe7zF24xZJkF8vYGwLF2RwkB3LZuUjGFs4j7GsqvWzGetI65HaJUHGMtY9tHL5XMb8\nX+8pvuFgqKErIojUKnfgglzWyaQs/FAdp1jb/DwptvaKDt+QDmmnEJsSlLoLG52v/GPmzENcjHSv\n13tt32NYkKpUIUQ8Nu8mpRWOBUESAYL08djlw4al/kASZO7L/MUWF2Q0Y+XZwn/s2UWMjapakpGV\nlTWkVmqXBLmfL8bvWsgDWLqvp3Dc8CWBiCBSq9yBCzKKMSkLP1Tw326+425ZEHbrjlVz5Z1CbEpQ\n6s4XXR3ckXWph1jmlpMnT37JIo9RZxDx2LyblFY4FgRJBAgSofvK3T7frvR2UZCnn2Jsa1Z4TCsE\nKZzFh2FpSGrvO4MEr6xYzt/A+AcFK6vZqQkr+s4gYqvcgYXPI1IWfqhN45tYUUSQ3z9+1w55pxDL\nzyBid77+2t8LLdNWmb0HEY/Nu0lpw+csCJIAECTCjm/xMRnK2iwK8slVpbW5OoL4rtreMH8yiwjS\nwnsWeQoanh7b+8XgkuYfPsSWTayrHlcYFqSFL6VWpSBSFn6oFbcH6ifcJ8Wyv424tlveKZDSInXn\n1A/5bWXFKxmV0qdXXZFH796WUy+nfyrEiMfm3aS0oiAtA/SMXgxAkAhzfyksn5wtfYpVcM3ogjFa\nQdi7Y9Kmn4iM9zkZwqdYd38/49bPGXtzdMYD9ezCA+lDf9otpJiTIXyKJbYqBZGy8EO1zhg66e1h\nG6VYduO8yE4Bnl7sLlB5z7CM23bI34MURR6F70Emit+DiMfm3WrEtOFjhWsE8QFBDCjfzQfqNHux\nRXnO1gIGDghiwP5r6gP3LrUXC0EuXiCIEc9fO+InNn/GDUEuXiAIACZAEABMgCAAmABBADABggBg\nAgQBwAQIAoAJEAQAEyAIACZAEABMgCAAmABBADABggBgAgQBwAQIAoAJEAQAEyAIACZAEABMgCAA\nmABBADABggBgAgQBwAQCQWq0NLXqNCZGPXXC2kAtdUryGmtaG6kz+uqoM9ZfoM5o74lMfOTawRlB\nWtuJn7GaGvKRUsvIBSGvsaad/F8aH7nFDUHqjPaeyMRHrh0gCCEQhAgIEjsQhAYIEhsQhBAIQkSy\nCRLatv61Nvkx+Nb6grNyi4BO7RCEiAEV5CslRmEQhFUXs8Pvy48V21nNarlFQKd2CEIEBDHGOSmi\nsSXI7v3Mv1Z+rG1gLYVyi4BO7RCECAhijHNSRGNLkHc+Z10v9z2eLaiUW9a98MILe3s1hELatgSh\nz8jIMzrwfz2QT6RKEMO4gXkiLzgrhowtQT7g54s18mOoZK0v0nKuubn5yzoN5wPatgRpoU5Yz+qp\nU5LXWNdxnjqjv9F2qEoQo7CmIElh0dh5In3OiiFjS5ATm9lf35cfj2/u7WsR0Dn74SUWEXiJZYxz\nUkRj71Osd17b1F5fID5uWb5y5evhNWmvTu0QhAgIYoyDVkSB70EIgSBEQJDYgSA0QJDYgCCEQBAi\nIEjsQBAaIEhsQBBCIAgRECR2IAgNECQ2IAghEIQICBI7EIQGCBIbBILUN2po69S2JUgLdcIm1kSd\nkrzGxs426oxNzbZDVYIYhbX2kBQWjZ0nsj7xkWuHaEF2BeNKoSM3ziBE4AxiDI0AVkQL8soP5m0N\nxJ5Cp3YIQgQEMYbMAVOUL7Gai36Y/3pLjCl0aocgREAQY+gkMEPzHiTw/37yA2m1qiuwYk23ZQqd\n2iEIERDEGGoV9NF7ky69FVmc2vDchLGPWabQqR2CEAFBjKEVwQiTT7GGHgkNP3Ui0zKFTu0QhAgI\nYgyhBSaYCOI9eXAc82VYptCpHYIQAUGMIbTABJUg0W875mXfsKZ64izLFDq1QxAiIIgxDtigg1IQ\nxduO1uLinspldUw1LxZje8pY8PmVKz+EICogCBGuFSTqbUcwODLIaRFeYinmxepdu7iM+f/c10mn\ndghCBAQxZiAEiXrbkZLiSRHIY6p5sUK9JWWs4tU3ilt5i58XWu3XcKFD25YgrdQJG1gDdUryGv0d\nF6gzNjbZDlUJYhTWEiQpLBo7T2TdQAiieNsxI9KsmhdrZxk7fZwd28RbNr300kv7ezT0hrRtroMN\ndAE2CPUO4MFVghiFDdAf+7zpuC5KcUSQYPhth2ZKLsW8WGFBON0vSXt1zn54iUUEXmIZYzquz1fR\n+GH2MW/JbdkCTDUvVliQfYdY9SYIogKCEBGHIKvTnnim/q6M2yvbZw/O9xzhZ5CO/CHDl7DVl89J\nvz++X+HqCFIyZYyAuDFy4bFyDlPNixUWJPDG2o1NEEQFBCEiHkE8T5X/fFz9wxOWXfHhL8KCvJh2\nYJ1n72rPm+s8n1IJMnJJmeQE5xqbv+zVqR2CEAFBjFEJktrDpg3Jzhr+z9PY38KCPJLLei//w+o0\ndsazj0qQqzuiNl58oQeCxAQEISIeQdIZe3Rq07p/X+w9sCAsyNK00g38DJJOKchz0U5MTh88WnwP\nAkHsAUGIiFOQmukZ0/c1z0ifFRYkMNc77HeMUpDs7NGXRDlRLgJBbANBiEjkU6ySxcF3PGfiN0JF\ntCDl5SonempDNlLo1A5BiIAgxhgMx+qp149YmpATCpQvsToL97Ci5V3ixtlpaUNKp1RbptCpHYIQ\nAUGMoZPADNU36Tl/Zftv+ZG4MePxjqzg/DssU+jUDkGIgCDGOKGDFtWPFU/wxQmvuPGNFpbFGtIs\nU9RqORfQaUyMRuqEPuajTkleY23gHHXGOr/tUJUgRmGNQZLCFCltxKgE+UqNI4KM+pgv9o8SN8a+\nzQXZ9l3LFPUNGto6tG0J0kKdsJE1Uqckr7Ghs406Y2OT7VDViDMKa+0hKSwaO0+k6seK/SPIHzN/\nvuxfMzeKG7szZ2fMuupdyxQ6Zz+8xCICL7GMGQhB2iqWzFt0tE3aaly3pNDGrRJ1aocgREAQY5Sj\nsD8Eib5GSsBfJQBBbANBiHCpINHXSHGe8GSN4kAQ20AQIhwRJPSb8bfeVVs1ePLkyQsiywlTS+0L\nEn2NFCf9gKUbEEQBBCHCEUE+mhpi//VYVa6w3rf8/Nvmv8nVux5EepOecwqCxAYEIYJGkM0/emTy\nlnOz7pz+eXit8vqSYE+rWhD28HsxCFL+aF5e3n3DxI39V89bxIEgtoEgRBAJMj108KHfLWI77w2v\nsf2PjMmrrPLm5uaukJaCIAvXxyDILfnz7y6asE3cmJSz8FkOBLENBCGCSJBF7OTs/BLW+Z3w2rFq\nFnpzouYM8i/vxyBIamPbnaxuiriR2Sw3a+fFklsgSDQQhAgiQZZwLfgZ5IN7wmvb8npZ1Xi1IBXZ\nsbwHydrLcpoDQ8WNX8oXnevMiyW3QJBoIAgRhIKce3AGfw8irAV/O/mWqYfDn1/NjiwnTvnE1A+V\nIKsuO/Wb8bf/g7gx6dIR0qQN2nmx5JYj+/btKzunoaNb25Yg7eQZGXlG+hq7O6gznr9gO1Q14ozC\n2npJCovGzhPZaCkICapPsU53BF9fcU5c77s4RDsvltyyZ+vWrUfbNXQFtW0J0kGekZFnpK8x2EWe\nMmA7UjXijMI6eknqUqS0EdOsHLn9JIj+l+faebHkFgGdsx9eYhGBl1jGDIQgii/PzebFklsgSDQQ\nhAjXCqL48txsXixpDYIogCBExCHIf6txRBDFl+eYFytWIAgRcQjyNTUJadGHUhDFl+eYFytWIAgR\nrhVE8eU55sWKFQhChGsFyYz+7AzzYsUKBCHCtYL0fXkeAzq1QxAiIIgxylHYTy+xIl+ec7LCjJn2\nYrtOPwiiQ3IJYjn+k12Q+K6QMhVE8apq3dStR7beXvCXex6FIPaAIETQCBLfFVKmgihuAz2yVhj9\no1lghHmKOu394xy4R2EzdUIH7lFIXqOj9yhUjX9tqGWAiAP3KLTzRKqmE9ERJL4rpEwFUdwGeqRw\nLjp0Dau3EMSn5XxApzExmqgT1rE66pTkNfoC56gz1jfIa6rxrw21DBBpCto+uM2Mtp7IWktB4rtC\nylSQqNtA85dYVy14ecFVBbuv+b15Cp2zH15iEXGRvcSymZHoJVZ8V0iZChJ1G2jOZ7+e+/QnrHy/\nRQqd2iEIERDEGEtB4rtCylQQxW2g7aJTOwQhwlWCGHRwrSDxXSFlKkjUbaBztueIQBDbQBAL+lcQ\nEgxvA13aWCpimUKndghCBAQxRjkK+0MQ+ycNBTq1QxAiIIgxylHYH4IoThpjZCCIbSCIBUkuyEkZ\nYeOIDFNN+xNeBJ9fufJDCKICgliQ5IJ4UjNFogOEeUgV0/6EF/4/90Xo1A5BiEhyQWxm0ODSW7A9\ndt0dK6O+wO+bh1Qx7U94UfHqG8WtEEQFBIkvgwaXCsJ69/3ihtv/86y01TcPqWLan/Di9HF2TPhp\n/LYNGzaUdmoI9mjbEqSbPCMjz0hfY0+QOmNXl7ymGq3aUHWAQYfukOHBbGbQYOeJbFUP5X4QhBP6\n5Jns28TVvnlIFdP+yBvdL/HFiePHj1c2a2jv0rYlyHnqhC2MOiN9jc1d7dQZW1rlNdVo1YaqAww6\nnO8xPJjNDBrsPJH+gRGk+92fDL9TXO2bh1Qx7U94se8Qq5YvrtI5++ElFhF4iWXMAAjSteP/DJ9Z\n2CBt9c1Dqpj2J7wIvLF2YxMEUQFB4sugwaWC/O/h96xritqOnofUBJ3aIQgREMQYchd0UXzMOyhd\nJLYUOrVDECIgiDGkHhgSLUjkWpTYUujUDkGIgCDGkHpgiOGPFe2jU7uJIMqvO208D1+LuYc9IAgE\nsQEEIQSCxJdBAwSBIPaAIMYkPnLt4LQgmuENQWIBghiT+Mi1AwQhBILEmsGghItLkPpGDW2d8ppq\neKsbtF01aDIQ0cSayHJJtFAnbOxso87Y1CyvqUarNlQdYNChtcfwYJYZDEqw80TWJz5y7eC6M4hF\nB5xBjDD4910NziCxAUEIgSCxZoAgECQhIIgxiY9cOwy4IOqAxAWx2eFiF8RgOEOQ2IAghECQWDNA\nEAiSEBDEmMRHrh0gCCH9KYjFWIMgREAQQiBIrBkuEkG082LJLQI6tUMQIiCIMc5JEY0tQbTzYskt\nAjq196sglhn0gSAQxAa2BNHOiyW3COjUDkGIgCDGOCdFNLYE0c6LJbcULFq0aJdZV9VoVTdoAyw6\n6MwoaZnhokQ11rQNmgDLDFYBlh1izxBzxj7arEMosCWIdl4suaUzEAic0d4/DvcoJMKBexT6qTPG\ncI9C2yltxNRajVoabAminRdLbhHQOfvhmnQinPy5OxXuuA20U9j7FEszL1Z4Ie3VqR2CEAFBjHHQ\niij6/ZLbeIEgNECQ2HBGkKMfEz9jNTV11AnPbDtDnZK8xpqPP6POWOujzli1kzqjvScy8ZFrBwJB\ndDjwliNpSeleFBzoEqz588cDXYE1p/4w0BU4CQRxNRBkoIEgrgaCDDTOCPL5h9YxA01wQ89Al2DN\nh+UDXYE19VsGugIncUYQAC4SIAgAJkAQAExwRJDoq0Vcyp4y11cZfGt9wVmXF9m1aW1hs8trTAxH\nBIm+WsSV9K5dXOb6Kiu2s5rVLi/y0G52+B2X15gYjggSfbWIKwn1lpS5vsraBtZS6PIi68+zIztd\nXmNiOCKIfLWIi9lZlgRVni2odH2Rf3yh2fU1JoIjgshXi7gYLojbqwyVrPW5vcjOXla90eU1JoYj\ngkRfLeJSuCBur/L45l7XP5XvHWan17m8xsRw5lOsqKtFXAoXxO1Vblm+cuXrLi/y/IbCV2tdXmNi\n4HsQAEyAIACYAEEAMAGCaAmNzuxm7MgYVppjt0uK0U/nhRQpQYtE/FB9ifRiFemPzhg89J8qGcv0\nCFRFHr0eT+qtuwy7GdcITIAgWj4dPuov4VHbuN1uF8PBJ6RICVokUgiiFxudvue6Z0/Uzh8TYpkf\ntHB6Io/evS1frkj/xLAqCBIPEETL/PnP5kfOIOuystZnsfLJL157/U5WPulXmZP333zFk4zt/V7a\nzLNy+wzPyDbGin78iHdSBWN/Gj14lp8Ffzok8zkhBd+5O0dulTpwjuUuuUnKIgjy6vWXT6yQYmeu\nYmxpnrSTI6QXu3NOec5zS+5tYZmlYrnyo/cIXzzzfWFVPLbQTUwrHEusEcQIBNHQe93RY95OSZDP\nrjxYM4ULkv58+1O3sfJLX2/KufrLAx5/Y+bW5p9Nk9vFf52LBr3if2pcqNr7XlN+HivOPnk49YvI\nSyypVe7ABfHmH5ey8EOdvmxPQ/6jUuzK+xib9Ja0UyAlKHXnBL87c0dAWNEV5KNvCavisXk3Ka1w\nLJxB4gKCaNj7PcZu3CIJ8usFjG3hggwOsmPZrHwEYwvnMZZVtX42Yx1pPVK7JMhYxrqHVi6fy5j/\n6z3FNxwMNXRFBJFa5Q5ckMs6mZSFH6rjFGubnyfF1l7R4RvSIe0UYlOCUndho/OVf8yceYiLke71\neq/tewwLUpUqhIjH5t2ktMKxIEhcQBANj10+bFjqDyRB5r7MX2xxQUYzVp4t/MeeXcTYqKolGVlZ\nWUNqpXZJkPv5YvyuhTyApft6CscNXxKICCK1yh24IKMYk7LwQwX/7eY77pYFYbfuWDVX3inEpgSl\n7nzR1cEdWZd6iGVuOXny5Jcs8hh1BhGPzbtJaYVjQZC4gCBquq/c7fPtSm8XBXn6Kca2ZoXHtEKQ\nwll8GJaGpPa+M0jwyorl/A2Mf1CwspqdmrCi7wwitsodWPg8ImXhh9o0vokVRQT5/eN37ZB3CrH8\nDCJ25+uv/b3QMm2V2XsQ8di8m5Q2fM6CIPEAQdTs+BYfk6GszaIgn1xVWpurI4jvqu0N8yeziCAt\nvGeRp6Dh6bG9Xwwuaf7hQ2zZxLrqcYVhQVr4UmpVCiJl4YdacXugfsJ9Uiz724hru+WdAiktUndO\n/ZDfVla8klEpfXrVFXn07m059XL6p0KMeGzeTUorCtIyQM9oUgNB1Mz9pbB8crb0KVbBNaMLxmgF\nYe+OSZt+IjLe52QIn2Ld/f2MWz9n7M3RGQ/UswsPpA/9abeQYk6G8CmW2KoURMrCD9U6Y+ikt4dt\nlGLZjfMiOwV4erG7QOU9wzJu2yF/D1IUeRS+B5kofg8iHpt3qxHTho8VrhHECAQxp3w3H6jT7MUW\n5TlbCxgAIIg5+6+pD9y71F4sBLkIgSAWPH/tiJ/Y/C03BLkIgSAAmABBADABggBgAgQBwAQIAoAJ\nEAQAEyAIACZAEABM+P8JSrswk65udwAAAABJRU5ErkJggg==\n"
      }
     ],
     "prompt_number": 53
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}