{
 "metadata": {
  "name": "",
  "signature": "sha256:edf359eb395f9164955a79ce9a45638e9f6b193ad8e38462fa222e52ae3efa10"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "# Description:\n",
      "\n",
      "* This notebook goes through the analysis of blasting spacers vs a database containing potential protospacers.\n",
      "* Any blast datbase can be used.\n",
      "* A script is provided to easily blast all of the genomes in CLdb."
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### Before running this notebook:\n",
      "\n",
      "* __run the [Setup](./Setup.pynb) notebook__"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from IPython.display import FileLinks\n",
      "%load_ext rmagic\n",
      "%load_ext rpy2.ipython"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "The rmagic extension is already loaded. To reload it, use:\n",
        "  %reload_ext rmagic\n",
        "The rpy2.ipython extension is already loaded. To reload it, use:\n",
        "  %reload_ext rpy2.ipython\n"
       ]
      }
     ],
     "prompt_number": 54
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# This tutorial assumes that this IPython notebook is in the ./CLdb/examples/ecoli_ipynb/ directory.\n",
      "## these are the relative paths to everything we will be working with\n",
      "CLdbDir = \"../CLdb/\"\n",
      "scriptDir = \"../../../\""
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 60
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Checking on files in the CLdb directory\n",
      "FileLinks(CLdbDir)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "../CLdb/<br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/CLdb.sqlite' target='_blank'>CLdb.sqlite</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/GIs.txt' target='_blank'>GIs.txt</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/SELECT * from loci' target='_blank'>SELECT * from loci</a><br>\n",
        "../CLdb/array/<br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/array/Ecoli_K12_DH10B_a2.txt' target='_blank'>Ecoli_K12_DH10B_a2.txt</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/array/Ecoli_K12_W3110_a1.txt' target='_blank'>Ecoli_K12_W3110_a1.txt</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/array/Ecoli_K12_MG1655_a1.txt' target='_blank'>Ecoli_K12_MG1655_a1.txt</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/array/Ecoli_BL21_DE3_a2.txt' target='_blank'>Ecoli_BL21_DE3_a2.txt</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/array/Ecoli_K12_MG1655_a2.txt' target='_blank'>Ecoli_K12_MG1655_a2.txt</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/array/Ecoli_BL21_DE3_a1.txt' target='_blank'>Ecoli_BL21_DE3_a1.txt</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/array/Ecoli_K12_W3110_a2.txt' target='_blank'>Ecoli_K12_W3110_a2.txt</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/array/Ecoli_K12_DH10B_a1.txt' target='_blank'>Ecoli_K12_DH10B_a1.txt</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/array/Ecoli_0157_H7_a1.txt' target='_blank'>Ecoli_0157_H7_a1.txt</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/array/Ecoli_0157_H7_a2.txt' target='_blank'>Ecoli_0157_H7_a2.txt</a><br>\n",
        "../CLdb/clustering/<br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/clustering/spacers.cdhit.clstr' target='_blank'>spacers.cdhit.clstr</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/clustering/spacers.cdhit' target='_blank'>spacers.cdhit</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/clustering/DRs.cdhit' target='_blank'>DRs.cdhit</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/clustering/DRs.fna' target='_blank'>DRs.fna</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/clustering/DRs.cdhit.clstr' target='_blank'>DRs.cdhit.clstr</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/clustering/spacers.fna' target='_blank'>spacers.fna</a><br>\n",
        "../CLdb/fasta/<br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/fasta/Escherichia_coli_BL2_DE3.fasta' target='_blank'>Escherichia_coli_BL2_DE3.fasta</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/fasta/Escherichia_coli_K_2_W3.fasta' target='_blank'>Escherichia_coli_K_2_W3.fasta</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/fasta/Escherichia_coli_K_2_MG_655.fasta' target='_blank'>Escherichia_coli_K_2_MG_655.fasta</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/fasta/Escherichia_coli_K_2_DH_B.fasta' target='_blank'>Escherichia_coli_K_2_DH_B.fasta</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/fasta/Escherichia_coli_O_57_H7.fasta' target='_blank'>Escherichia_coli_O_57_H7.fasta</a><br>\n",
        "../CLdb/genbank/<br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/genbank/Escherichia_coli_BL2_DE3.gbk' target='_blank'>Escherichia_coli_BL2_DE3.gbk</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/genbank/Escherichia_coli_K_2_W3.gbk' target='_blank'>Escherichia_coli_K_2_W3.gbk</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/genbank/Escherichia_coli_K_2_MG_655.gbk' target='_blank'>Escherichia_coli_K_2_MG_655.gbk</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/genbank/Escherichia_coli_K_2_DH_B.gbk' target='_blank'>Escherichia_coli_K_2_DH_B.gbk</a><br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/genbank/Escherichia_coli_O_57_H7.gbk' target='_blank'>Escherichia_coli_O_57_H7.gbk</a><br>\n",
        "../CLdb/loci/<br>\n",
        "&nbsp;&nbsp;<a href='../CLdb/loci/140419_loci.txt' target='_blank'>140419_loci.txt</a><br>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 56,
       "text": [
        "../CLdb/\n",
        "  CLdb.sqlite\n",
        "  GIs.txt\n",
        "  SELECT * from loci\n",
        "../CLdb/array/\n",
        "  Ecoli_K12_DH10B_a2.txt\n",
        "  Ecoli_K12_W3110_a1.txt\n",
        "  Ecoli_K12_MG1655_a1.txt\n",
        "  Ecoli_BL21_DE3_a2.txt\n",
        "  Ecoli_K12_MG1655_a2.txt\n",
        "  Ecoli_BL21_DE3_a1.txt\n",
        "  Ecoli_K12_W3110_a2.txt\n",
        "  Ecoli_K12_DH10B_a1.txt\n",
        "  Ecoli_0157_H7_a1.txt\n",
        "  Ecoli_0157_H7_a2.txt\n",
        "../CLdb/clustering/\n",
        "  spacers.cdhit.clstr\n",
        "  spacers.cdhit\n",
        "  DRs.cdhit\n",
        "  DRs.fna\n",
        "  DRs.cdhit.clstr\n",
        "  spacers.fna\n",
        "../CLdb/fasta/\n",
        "  Escherichia_coli_BL2_DE3.fasta\n",
        "  Escherichia_coli_K_2_W3.fasta\n",
        "  Escherichia_coli_K_2_MG_655.fasta\n",
        "  Escherichia_coli_K_2_DH_B.fasta\n",
        "  Escherichia_coli_O_57_H7.fasta\n",
        "../CLdb/genbank/\n",
        "  Escherichia_coli_BL2_DE3.gbk\n",
        "  Escherichia_coli_K_2_W3.gbk\n",
        "  Escherichia_coli_K_2_MG_655.gbk\n",
        "  Escherichia_coli_K_2_DH_B.gbk\n",
        "  Escherichia_coli_O_57_H7.gbk\n",
        "../CLdb/loci/\n",
        "  140419_loci.txt"
       ]
      }
     ],
     "prompt_number": 56
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\"\n",
      "# making new directory\n",
      "\n",
      "cd $1\n",
      "\n",
      "BlastDir='arrayBlast'\n",
      "\n",
      "if [ -d $1$BlastDir ]; then\n",
      "    rm -r $1$BlastDir\n",
      "fi    \n",
      "\n",
      "mkdir $1$BlastDir"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 58
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### Selectin spacer sequences"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "# Getting 'unique' spacer sequences (OTUs at 100% sequence identity)\n",
      "\n",
      "cd $1\n",
      "\n",
      "## selecting spacer sequences and writing as a fasta file\n",
      "$2'CLdb' array2fasta -- -da CLdb.sqlite -clust -cut 1 > ./arrayBlast/spacers_cut1.fna\n",
      "\n",
      "## getting number of spacer sequences\n",
      "printf 'Number_unique_spacers:\\t'; grep -c \">\" ./arrayBlast/spacers_cut1.fna\n",
      "\n",
      "## looking at spacers\n",
      "head -n 6 ./arrayBlast/spacers_cut1.fna"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Number_unique_spacers:\t39\n",
        ">NA|spacer|NA|22|1\n",
        "TTTTACAGTGTCAACAATCGTTCCCTGATTGT\n",
        ">NA|spacer|NA|13|1\n",
        "GACAGAACGGCCTCAGTAGTCTCGTCAGGCTC\n",
        ">NA|spacer|NA|6|1\n",
        "GAAATGCTGGTGAGCGTTAATGCCGCAAACACA\n"
       ]
      }
     ],
     "prompt_number": 61
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### Blast\n",
      "\n",
      "* The array blast commands are sub-sub commands.\n",
      "* __SO__, that's why when you type: '`CLdb arrayBlast -- h`', you get then following:"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "cd $1\n",
      "\n",
      "$2'CLdb' arrayBlast -- -h\n",
      "\n",
      "# you need to use sub-sub commands!"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "Usage:\n",
        "    arrayBlast [options] subcommand -- [subcommand_options]\n",
        "\n",
        "  Options:\n",
        "    --list\n",
        "        List all subcommands.\n",
        "\n",
        "    --perldoc\n",
        "        Get perldoc of subcommand.\n",
        "\n",
        "    -v Verbose output\n",
        "    -h This help message\n",
        "\n",
        "  For more information:\n",
        "    perldoc arrayBlast\n",
        "\n"
       ]
      }
     ],
     "prompt_number": 72
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "# list arrayBlast subcommands\n",
      "\n",
      "cd $1\n",
      "\n",
      "$2'CLdb' arrayBlast -- --list"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "add_crRNA\n",
        "add_proto\n",
        "align_proto\n",
        "filter_DRs\n",
        "get_PAM\n",
        "get_SEEDstats\n",
        "get_align\n",
        "get_proto\n",
        "run\n"
       ]
      }
     ],
     "prompt_number": 73
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "#### Subsubcommand:\n",
      "\n",
      "`CLdb arrayBlast -- run -- -h`"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "# spacer blast\n",
      "## Really just a thin wrapper around blastn that will blast the spacers against all genomes in CLdb\n",
      "### If you want to blast some else, just run the blast yourself (use 'perldoc CLdb_arrayBlast.pl' for more info)\n",
      "\n",
      "cd $1\n",
      "\n",
      "$2'CLdb' arrayBlast -- run -- -da CLdb.sqlite -query ./arrayBlast/spacers_cut1.fna > ./arrayBlast/spacers_cut1_blastn.xml"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "Making BLAST databases...\n",
        "...making blast databases in /home/nyoungb2/perl/projects/CLdb/examples/ecoli/CLdb//arrayBlast/blast_db/\n",
        "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/perl/projects/CLdb/examples/ecoli/CLdb//arrayBlast/blast_db//Escherichia_coli_BL2_DE3.fasta\n",
        "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/perl/projects/CLdb/examples/ecoli/CLdb//arrayBlast/blast_db//Escherichia_coli_K_2_DH_B.fasta\n",
        "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/perl/projects/CLdb/examples/ecoli/CLdb//arrayBlast/blast_db//Escherichia_coli_K_2_MG_655.fasta\n",
        "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/perl/projects/CLdb/examples/ecoli/CLdb//arrayBlast/blast_db//Escherichia_coli_K_2_W3.fasta\n",
        "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/perl/projects/CLdb/examples/ecoli/CLdb//arrayBlast/blast_db//Escherichia_coli_O_57_H7.fasta\n",
        "\n",
        "Blasting sequences...\n",
        "Number of blast hits to Escherichia_coli_BL2_DE3.fasta:\t1290\n",
        "Number of blast hits to Escherichia_coli_K_2_DH_B.fasta:\t1317\n",
        "Number of blast hits to Escherichia_coli_K_2_MG_655.fasta:\t1317\n",
        "Number of blast hits to Escherichia_coli_K_2_W3.fasta:\t1317\n",
        "Number of blast hits to Escherichia_coli_O_57_H7.fasta:\t912\n"
       ]
      }
     ],
     "prompt_number": 74
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "# converting .xml to .srl\n",
      "## This retains all of the info in the xml, but produces a smaller file that is faster to load and write.\n",
      "## For large blast runs (xml files > 100 Mb), this will take a little while\n",
      "\n",
      "cd $1\n",
      "\n",
      "$2'CLdb' arrayBlast -- xml2srl -- ./arrayBlast/spacers_cut1_blastn.xml > ./arrayBlast/spacers_cut1_blastn.srl\n",
      "\n",
      "\n",
      "# checking output file\n",
      "ls -thlc ./arrayBlast/spacers_cut1_blastn.srl"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 62K Oct 26 13:52 ./arrayBlast/spacers_cut1_blastn.srl\n"
       ]
      }
     ],
     "prompt_number": 77
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### Blasting DRs against the same queries. \n",
      "\n",
      "* This is needed to filter out spacer blast hits to CRISPR arrays"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "# selecting DRs, so we can blast them.\n",
      "## This is needed blast filter out spacers that hit CRISPR arrays\n",
      "\n",
      "cd $1\n",
      "\n",
      "$2'CLdb' array2fasta -- -da CLdb.sqlite -clust -cut 1 -r > ./arrayBlast/DRs_cut1.fna\n",
      "\n",
      "\n",
      "# checking output file\n",
      "ls -thlc ./arrayBlast/DRs_cut1.fna"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 528 Oct 26 13:52 ./arrayBlast/DRs_cut1.fna\n"
       ]
      }
     ],
     "prompt_number": 78
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "# DR blast, just like spacer blast\n",
      "\n",
      "cd $1\n",
      "\n",
      "$2'CLdb' arrayBlast -- run -- -da CLdb.sqlite -query ./arrayBlast/DRs_cut1.fna > ./arrayBlast/DRs_cut1_blastn.xml\n",
      "\n",
      "\n",
      "# checking output file\n",
      "ls -thlc ./arrayBlast/DRs_cut1_blastn.xml"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 783K Oct 26 13:53 ./arrayBlast/DRs_cut1_blastn.xml\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "Making BLAST databases...\n",
        "...making blast databases in /home/nyoungb2/perl/projects/CLdb/examples/ecoli/CLdb//arrayBlast/blast_db/\n",
        "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/perl/projects/CLdb/examples/ecoli/CLdb//arrayBlast/blast_db//Escherichia_coli_BL2_DE3.fasta\n",
        "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/perl/projects/CLdb/examples/ecoli/CLdb//arrayBlast/blast_db//Escherichia_coli_K_2_DH_B.fasta\n",
        "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/perl/projects/CLdb/examples/ecoli/CLdb//arrayBlast/blast_db//Escherichia_coli_K_2_MG_655.fasta\n",
        "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/perl/projects/CLdb/examples/ecoli/CLdb//arrayBlast/blast_db//Escherichia_coli_K_2_W3.fasta\n",
        "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/perl/projects/CLdb/examples/ecoli/CLdb//arrayBlast/blast_db//Escherichia_coli_O_57_H7.fasta\n",
        "\n",
        "Blasting sequences...\n",
        "Number of blast hits to Escherichia_coli_BL2_DE3.fasta:\t4730\n",
        "Number of blast hits to Escherichia_coli_K_2_DH_B.fasta:\t4825\n",
        "Number of blast hits to Escherichia_coli_K_2_MG_655.fasta:\t4882\n",
        "Number of blast hits to Escherichia_coli_K_2_W3.fasta:\t4882\n",
        "Number of blast hits to Escherichia_coli_O_57_H7.fasta:\t1747\n"
       ]
      }
     ],
     "prompt_number": 80
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "# converting .xml to .srl\n",
      "## This retains all of the info in the xml, but produces a smaller file that is faster to load and write.\n",
      "## For large blast runs (xml files > 100 Mb), this will take a little while\n",
      "\n",
      "cd $1\n",
      "\n",
      "$2'CLdb' arrayBlast -- xml2srl -- ./arrayBlast/DRs_cut1_blastn.xml > ./arrayBlast/DRs_cut1_blastn.srl\n",
      "\n",
      "\n",
      "# checking output file\n",
      "ls -thlc ./arrayBlast/DRs_cut1_blastn.srl"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 247K Oct 26 13:55 ./arrayBlast/DRs_cut1_blastn.srl\n"
       ]
      }
     ],
     "prompt_number": 82
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "#### Filtering out spacer hits to arrays\n",
      "\n",
      "* Based on DR hits falling adjacent to spacer hits"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "# filtering spacer hits to arrays\n",
      "## This is based on the premise that DR hits adjacent to spacer hits signify that the spacer is hitting a CRISPR array\n",
      "\n",
      "cd $1\n",
      "\n",
      "$2'CLdb' arrayBlast -- filter_arrayHits -- ./arrayBlast/spacers_cut1_blastn.srl \\\n",
      "    ./arrayBlast/DRs_cut1_blastn.srl > ./arrayBlast/spacers_cut1_blastn_filt.srl\n",
      "    \n",
      "    \n",
      "# checking output file\n",
      "ls -thlc ./arrayBlast/spacers_cut1_blastn_filt.srl"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 43K Oct 26 13:58 ./arrayBlast/spacers_cut1_blastn_filt.srl\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "### Summary for filtering of 'bad' DR hits ###\n",
        "Total DR hits: 1014\n",
        "DR hits with E-values < 10: 0\n",
        "DR hits with hit lengths (fraction of total) < 0.66: 0\n",
        "DR hits used for filtering: 1014\n",
        "### DR filtering of spacer blast hits ###\n",
        "Total spacer blast hits: 79\n",
        "Spacer blast hits hitting an array (filtered out): 79\n",
        "\tDeleting the blast hits to arrays\n",
        "Spacer blast hits hitting a protospacer: 0\n"
       ]
      }
     ],
     "prompt_number": 84
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### No spacers hit any protospacers on the Ecoli genomes in the CLdb!\n",
      "\n",
      "* This indicates that there are no interegrated mobile genetic elements in the E.coli genomes that can be detected using the CRISPR spacers.\n",
      "* Maybe we will get some hits when we blast against NCBI's nt database."
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Array Blast vs NCBI's nt database\n",
      "\n",
      "* NOTE: you will need the BLAST nt database & the blast+ toolkit"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "# The arrayBlast wrapper only works with the genomes in CLdb.\n",
      "# We will need to do this part 'manually'\n",
      "## The blast output just needs to be in .xml format.\n",
      "## NOTE: you will need the BLAST nt database \n",
      "\n",
      "num_threads=20\n",
      "\n",
      "cd $1'arrayBlast'\n",
      "\n",
      "## spacer blast with blastn\n",
      "blastn -task blastn-short -db nt -query spacers_cut1.fna \\\n",
      "    -outfmt 5 -evalue 1e-5 -num_threads $num_threads > spacers_cut1_vNT.xml\n",
      "    \n",
      "\n",
      "## checking output\n",
      "ls -thlc spacers_cut1_vNT.xml"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 1.6M Oct 26 14:04 spacers_cut1_vNT.xml\n"
       ]
      }
     ],
     "prompt_number": 87
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "# Now, blasting with DR sequences (needed to filter out hits to CRISPR arrays)\n",
      "\n",
      "num_threads=20\n",
      "\n",
      "cd $1'arrayBlast'\n",
      "\n",
      "# DR blast\n",
      "blastn -task blastn-short -db nt -query DRs_cut1.fna \\\n",
      "    -outfmt 5 -evalue 1e-5 -num_threads $num_threads > DRs_cut1_vNT.xml\n",
      "    \n",
      "    \n",
      " ## checking output\n",
      "ls -thlc DRs_cut1_vNT.xml"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 15M Oct 26 14:03 DRs_cut1_vNT.xml\n"
       ]
      }
     ],
     "prompt_number": 86
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# checking files in arrayBlast directory\n",
      "!ls -thlc $CLdbDir'arrayBlast'"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "total 18M\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 1.6M Oct 26 14:04 spacers_cut1_vNT.xml\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2  15M Oct 26 14:03 DRs_cut1_vNT.xml\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2  43K Oct 26 13:58 spacers_cut1_blastn_filt.srl\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 247K Oct 26 13:55 DRs_cut1_blastn.srl\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 783K Oct 26 13:53 DRs_cut1_blastn.xml\r\n",
        "drwxrwxr-x 2 nyoungb2 nyoungb2 4.0K Oct 26 13:53 blast_db\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2  528 Oct 26 13:52 DRs_cut1.fna\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2  62K Oct 26 13:52 spacers_cut1_blastn.srl\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 227K Oct 26 13:47 spacers_cut1_blastn.xml\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 2.0K Oct 26 13:39 spacers_cut1.fna\r\n"
       ]
      }
     ],
     "prompt_number": 88
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "# converting *.xml files to *.srl format\n",
      "\n",
      "cd $1\n",
      "\n",
      "## spacers\n",
      "$2'CLdb' arrayBlast -- xml2srl -- ./arrayBlast/spacers_cut1_vNT.xml > ./arrayBlast/spacers_cut1_vNT.srl\n",
      "\n",
      "## DRs\n",
      "$2'CLdb' arrayBlast -- xml2srl -- ./arrayBlast/DRs_cut1_vNT.xml > ./arrayBlast/DRs_cut1_vNT.srl\n",
      "\n",
      "\n",
      "## checking output\n",
      "ls -thlc ./arrayBlast/*_vNT.srl"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 4.8M Oct 26 14:07 ./arrayBlast/DRs_cut1_vNT.srl\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 584K Oct 26 14:06 ./arrayBlast/spacers_cut1_vNT.srl\n"
       ]
      }
     ],
     "prompt_number": 92
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "# filtering spacer hits to arrays\n",
      "## This is based on the premise that DR hits adjacent to spacer hits signify that the spacer is hitting a CRISPR array\n",
      "\n",
      "cd $1\n",
      "\n",
      "$2'CLdb' arrayBlast -- filter_arrayHits -- ./arrayBlast/spacers_cut1_vNT.srl ./arrayBlast/DRs_cut1_vNT.srl \\\n",
      "    > ./arrayBlast/spacers_cut1_blastn_filt.srl\n",
      "    \n",
      "    \n",
      "## checking output file\n",
      "ls -thlc ./arrayBlast/spacers_cut1_blastn_filt.srl"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 385K Oct 26 14:07 ./arrayBlast/spacers_cut1_blastn_filt.srl\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "### Summary for filtering of 'bad' DR hits ###\n",
        "Total DR hits: 20159\n",
        "DR hits with E-values < 10: 0\n",
        "DR hits with hit lengths (fraction of total) < 0.66: 0\n",
        "DR hits used for filtering: 20159\n",
        "### DR filtering of spacer blast hits ###\n",
        "Total spacer blast hits: 1573\n",
        "Spacer blast hits hitting an array (filtered out): 863\n",
        "\tDeleting the blast hits to arrays\n",
        "Spacer blast hits hitting a protospacer: 710\n"
       ]
      }
     ],
     "prompt_number": 94
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## OK. It's looks like we have some hits to some putative protospacers!\n",
      "\n",
      "__Let the investigation begin...__"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### File conversion:\n",
      "\n",
      "* __FYI__, the .srl files can be converted to json or csv format. For example:"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "# The srl file can be converted to a json or csv file if you would like to use tools external to CLdb.\n",
      "\n",
      "cd $1\n",
      "\n",
      "$2'CLdb' arrayBlast -- srl2json -- < ./arrayBlast/spacers_cut1_blastn_filt.srl > ./arrayBlast/spacers_cut1_blastn_filt.json\n",
      "\n",
      "\n",
      "## viewing some of the json file\n",
      "cat ./arrayBlast/spacers_cut1_blastn_filt.json | python -mjson.tool 2>/dev/null | head -n 40"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "{\n",
        "    \"0\": {\n",
        "        \"BlastOutput_db\": \"nt\", \n",
        "        \"BlastOutput_iterations\": {\n",
        "            \"Iteration\": [\n",
        "                {\n",
        "                    \"Iteration_hits\": {\n",
        "                        \"Hit\": [\n",
        "                            {\n",
        "                                \"Hit_accession\": \"KC765378\", \n",
        "                                \"Hit_def\": \"Escherichia coli strain 90-0009 CRISPR2a repeat region\", \n",
        "                                \"Hit_hsps\": {\n",
        "                                    \"Hsp\": {\n",
        "                                        \"2NhOQxJMOvPdb\": {\n",
        "                                            \"CLdb_array-hit\": 0, \n",
        "                                            \"Hsp_align-len\": \"32\", \n",
        "                                            \"Hsp_bit-score\": \"63.9282\", \n",
        "                                            \"Hsp_evalue\": \"3.97536e-08\", \n",
        "                                            \"Hsp_gaps\": \"0\", \n",
        "                                            \"Hsp_hit-frame\": \"-1\", \n",
        "                                            \"Hsp_hit-from\": \"286\", \n",
        "                                            \"Hsp_hit-to\": \"255\", \n",
        "                                            \"Hsp_hseq\": \"TTTTACAGTGTCAACAATCGTTCCCTGATTGT\", \n",
        "                                            \"Hsp_identity\": \"32\", \n",
        "                                            \"Hsp_midline\": \"||||||||||||||||||||||||||||||||\", \n",
        "                                            \"Hsp_num\": \"1\", \n",
        "                                            \"Hsp_positive\": \"32\", \n",
        "                                            \"Hsp_qseq\": \"TTTTACAGTGTCAACAATCGTTCCCTGATTGT\", \n",
        "                                            \"Hsp_query-frame\": \"1\", \n",
        "                                            \"Hsp_query-from\": \"1\", \n",
        "                                            \"Hsp_query-to\": \"32\", \n",
        "                                            \"Hsp_score\": \"32\"\n",
        "                                        }\n",
        "                                    }\n",
        "                                }, \n",
        "                                \"Hit_id\": \"gi|528889517|gb|KC765378.1|\", \n",
        "                                \"Hit_len\": \"922\", \n",
        "                                \"Hit_num\": \"1\"\n",
        "                            }, \n",
        "                            {\n"
       ]
      }
     ],
     "prompt_number": 95
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# status updated on while files are in the arrayBlast directory\n",
      "!ls -thlc $CLdbDir'arrayBlast' | sort -k 9"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "total 24M\r\n",
        "drwxrwxr-x 2 nyoungb2 nyoungb2 4.0K Oct 26 13:53 blast_db\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 247K Oct 26 13:55 DRs_cut1_blastn.srl\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 783K Oct 26 13:53 DRs_cut1_blastn.xml\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2  528 Oct 26 13:52 DRs_cut1.fna\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 4.8M Oct 26 14:07 DRs_cut1_vNT.srl\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2  15M Oct 26 14:03 DRs_cut1_vNT.xml\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 638K Oct 26 14:09 spacers_cut1_blastn_filt.json\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 385K Oct 26 14:07 spacers_cut1_blastn_filt.srl\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2  62K Oct 26 13:52 spacers_cut1_blastn.srl\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 227K Oct 26 13:47 spacers_cut1_blastn.xml\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 2.0K Oct 26 13:39 spacers_cut1.fna\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 584K Oct 26 14:06 spacers_cut1_vNT.srl\r\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 1.6M Oct 26 14:04 spacers_cut1_vNT.xml\r\n"
       ]
      }
     ],
     "prompt_number": 96
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "# adding crRNA region to .srl (actually the DNA template, refering to it as 'crDNA')\n",
      "## How much of the adjacent DR sequences are included in the crDNA is determined by the user (default: 10bp on either side)\n",
      "\n",
      "cd $1\n",
      "\n",
      "$2'CLdb' arrayBlast -- add_crRNA -- -da CLdb.sqlite \\\n",
      "    < ./arrayBlast/spacers_cut1_blastn_filt.srl \\\n",
      "    > ./arrayBlast/spacers_cut1_blastn_filt_crDNA.srl\n",
      "   \n",
      "   \n",
      "## checking output\n",
      "ls -thcl ./arrayBlast/spacers_cut1_blastn_filt_crDNA.srl"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 403K Oct 26 14:14 ./arrayBlast/spacers_cut1_blastn_filt_crDNA.srl\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "Decoding .srl file...\n",
        "Extracting all queries with blast hits from blast hit file...\n",
        "...Number of total unique query IDs:\t39\n",
        "Getting info from CLdb on each blast query spacer...\n",
        "Getting spacer regions from each genome...\n",
        " Processing genome: Escherichia_coli_K_2_DH_B.fasta...\n",
        " Processing genome: Escherichia_coli_K_2_MG_655.fasta...\n",
        " Processing genome: Escherichia_coli_O_57_H7.fasta...\n",
        " Processing genome: Escherichia_coli_BL2_DE3.fasta...\n",
        " Processing genome: Escherichia_coli_K_2_W3.fasta...\n"
       ]
      }
     ],
     "prompt_number": 103
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "# adding protospacer region to .srl\n",
      "\n",
      "cd $1\n",
      "\n",
      "\n",
      "$2'CLdb' arrayBlast -- add_proto -- -workers 24 \\\n",
      "    < ./arrayBlast/spacers_cut1_blastn_filt_crDNA.srl \\\n",
      "    > ./arrayBlast/spacers_cut1_blastn_filt_crDNA_proto.srl   \n",
      "    \n",
      "    \n",
      "## checking output\n",
      "ls -thlc ./arrayBlast/spacers_cut1_blastn_filt_crDNA_proto.srl   "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 514K Oct 26 14:17 ./arrayBlast/spacers_cut1_blastn_filt_crDNA_proto.srl\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "Retrieving proto info from blast db: 'nt'\n",
        "Retrieving proto sequences from blast db: 'nt'\n"
       ]
      }
     ],
     "prompt_number": 106
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "# aligning crRNA and protospacer\n",
      "## alignment added to the .srl file.\n",
      "## These are individual alignments between the spacer and protospacer (using clustalw).\n",
      "\n",
      "cd $1\n",
      "\n",
      "$2'CLdb' arrayBlast -- align_proto -- \\\n",
      "    < ./arrayBlast/spacers_cut1_blastn_filt_crDNA_proto.srl \\\n",
      "    > ./arrayBlast/spacers_cut1_blastn_filt_crDNA_proto_aln.srl\n",
      "    \n",
      "## checking output\n",
      "ls -thlc ./arrayBlast/spacers_cut1_blastn_filt_crDNA_proto_aln.srl"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 707K Oct 26 14:17 ./arrayBlast/spacers_cut1_blastn_filt_crDNA_proto_aln.srl\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "Aligning protospacers from blast db: 'nt'\n"
       ]
      }
     ],
     "prompt_number": 107
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "# getting alignment of protospacer and crDNA and writing as a fasta\n",
      "## You can parse out the alignments you want (parse by subtype, taxon_name, etc.)\n",
      "##   OR you can parse them after writing (add subtype, taxon_name, e-value, etc. to the sequence name)\n",
      "## The alignments will be oriented to the crDNA (5'-3' for the crRNA)\n",
      "\n",
      "cd $1\n",
      "\n",
      "$2'CLdb' arrayBlast -- get_align -- \\\n",
      "    -da CLdb.sqlite -outfmt taxon_name,subtype,evalue \\\n",
      "    < ./arrayBlast/spacers_cut1_blastn_filt_crDNA_proto_aln.srl \\\n",
      "    > ./arrayBlast/all_crDNA-proto_aln.fna\n",
      "    \n",
      "\n",
      "## checking output\n",
      "ls -thlc ./arrayBlast/all_crDNA-proto_aln.fna"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 384K Oct 26 14:18 ./arrayBlast/all_crDNA-proto_aln.fna\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "Decoding .srl file...\n",
        "Getting metadata info from CLdb...\n",
        "Getting alignments...\n"
       ]
      }
     ],
     "prompt_number": 110
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# let's take a look at the sequences \n",
      "!head -n 20 $CLdbDir'/arrayBlast/all_crDNA-proto_aln.fna'"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        ">crDNA|J|10|2NhOQxJMOvPdb|Escherichia_coli_BL21_DE3|I-E|3.97536e-08\r\n",
        "gggataaaccGACAATCAGGGAACGATTGTTGACACTGTAAAAgtgttccccg\r\n",
        ">proto|J|10|2NhOQxJMOvPdb|Escherichia_coli_BL21_DE3|I-E|3.97536e-08\r\n",
        ".ggataaaccgACAATCAGGGAACGATTGTTGACACTGTAAAAgtgttccccg\r\n",
        ">crDNA|J|10|2RjzImOT39gS9|Escherichia_coli_BL21_DE3|I-E|3.97536e-08\r\n",
        "gggataaaccGACAATCAGGGAACGATTGTTGACACTGTAAAAgtgttccccg\r\n",
        ">proto|J|10|2RjzImOT39gS9|Escherichia_coli_BL21_DE3|I-E|3.97536e-08\r\n",
        ".ggataaaccgACAATCAGGGAACGATTGTTGACACTGTAAAAgtgttccccg\r\n",
        ">crDNA|J|10|2RIDXN3JOzkVv|Escherichia_coli_BL21_DE3|I-E|3.97536e-08\r\n",
        "gggataaaccGACAATCAGGGAACGATTGTTGACACTGTAAAAgtgttccccg\r\n",
        ">proto|J|10|2RIDXN3JOzkVv|Escherichia_coli_BL21_DE3|I-E|3.97536e-08\r\n",
        ".ggataaaccgACAATCAGGGAACGATTGTTGACACTGTAAAAgtgttccccg\r\n",
        ">crDNA|J|10|2S7ULOBVdK1yB|Escherichia_coli_BL21_DE3|I-E|3.97536e-08\r\n",
        "gggataaaccGACAATCAGGGAACGATTGTTGACACTGTAAAAgtgttccccg\r\n",
        ">proto|J|10|2S7ULOBVdK1yB|Escherichia_coli_BL21_DE3|I-E|3.97536e-08\r\n",
        ".ggataaaccgACAATCAGGGAACGATTGTTGACACTGTAAAAgtgttccccg\r\n",
        ">crDNA|J|10|2TkZohEMYZehT|Escherichia_coli_BL21_DE3|I-E|3.97536e-08\r\n",
        "gggataaaccGACAATCAGGGAACGATTGTTGACACTGTAAAAgtgttccccg\r\n",
        ">proto|J|10|2TkZohEMYZehT|Escherichia_coli_BL21_DE3|I-E|3.97536e-08\r\n",
        ".ggataaaccgACAATCAGGGAACGATTGTTGACACTGTAAAAgtgttccccg\r\n"
       ]
      }
     ],
     "prompt_number": 111
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "__The alignment is really just for each crDNA-protospacer pair (every 4 lines):__\n",
      "\n",
      "~~~\n",
      ">crDNA\n",
      "ATAGACA\n",
      ">proto\n",
      "ATAGACA\n",
      "~~~\n",
      "\n",
      "__If you wanted each pair individually, you could use something like the following:__"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# WARNING: this will produce a lot of files!\n",
      "!split -l 4 -d -a 4 $CLdbDir'arrayBlast/all_crDNA-proto_aln.fna' $CLdbDir'arrayBlast/crDNA-proto_aln_split'\n",
      "\n",
      "## Let look at one of them\n",
      "!find $CLdbDir'arrayBlast/' -name 'crDNA-proto_aln_split[0-9]*' | head -n 1 | xargs -I % cat %"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        ">crDNA|J|6|9k4U1yAElmYJ|Escherichia_coli_BL21_DE3|I-E|3.97536e-08\r\n",
        "gggataaaccGGCGACCACGCAGAGCGGCGACAACCTGACGTTgtgttccccg\r\n",
        ">proto|J|6|9k4U1yAElmYJ|Escherichia_coli_BL21_DE3|I-E|3.97536e-08\r\n",
        ".ggataaaccgGCGACCACGCAGAGCGGCGACAACCTGACGTTgtgttccccg\r\n",
        "find: `standard output': Broken pipe\r\n",
        "find: write error\r\n"
       ]
      }
     ],
     "prompt_number": 114
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Let's remove those files\n",
      "!find $CLdbDir'arrayBlast' -name 'crDNA-proto_aln_split[0-9]*' | xargs -I % rm %"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 115
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "# Looking at the alignments can be very useful, but it can help to extract what we specifically want and summarize.\n",
      "## Let's pull out the potential PAM regions from the alignment \n",
      "## We are going to assume that the PAM is:\n",
      "### 4bp (really 3bp) long \n",
      "### adjacent to the 5' end of the protospacer (right side of the alignment)\n",
      "\n",
      "cd $1\n",
      "\n",
      "$2'CLdb' arrayBlast -- get_PAM -- -PAM 1 4 -f - \\\n",
      "    < ./arrayBlast/all_crDNA-proto_aln.fna \\\n",
      "    > ./arrayBlast/all_crDNA-proto_aln_PAM.fna\n",
      "\n",
      "\n",
      "## checking output\n",
      "ls -thlc ./arrayBlast/all_crDNA-proto_aln_PAM.fna\n",
      "head ./arrayBlast/all_crDNA-proto_aln_PAM.fna"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 115K Oct 26 14:20 ./arrayBlast/all_crDNA-proto_aln_PAM.fna\n",
        ">proto|E|9|31etS2GOyPEEd|Escherichia_coli_K-12_W3110|I-E|1.08991e-08\n",
        "gagt\n",
        ">proto|B|5|1FQXZpz9NF153|Escherichia_coli_K-12_MG1655|I-E|1.08991e-08\n",
        "gtgt\n",
        ">proto|E|5|19t75kH7llSgB|Escherichia_coli_K-12_W3110|I-E|3.97536e-08\n",
        "gagt\n",
        ">proto|D|5|1sC3vTY2xPMjj|Escherichia_coli_K-12_DH10B|I-E|1.08991e-08\n",
        "gtgt\n",
        ">proto|J|5|2IuamidfMPyMR|Escherichia_coli_BL21_DE3|I-E|3.97536e-08\n",
        "gtgt\n"
       ]
      }
     ],
     "prompt_number": 117
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#summarizing each PAM sequence (4 nt most proximal to spacer-protospacer alignment)\n",
      "!egrep -v \"^>\" $CLdbDir'./arrayBlast/all_crDNA-proto_aln_PAM.fna' | sort | uniq -c | sort -k 1 -n -r "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "   1201 gtgt\r\n",
        "    371 gagt\r\n",
        "      6 \r\n",
        "      4 gagc\r\n",
        "      3 aatt\r\n",
        "      3 aagt\r\n",
        "      2 cgtg\r\n",
        "      2 cggt\r\n",
        "      2 acgg\r\n"
       ]
      }
     ],
     "prompt_number": 119
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# and if we reverse the sequence...\n",
      "!egrep -v \"^>\" $CLdbDir'./arrayBlast/all_crDNA-proto_aln_PAM.fna' | rev | sort | uniq -c | sort -k 1 -n -r "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "   1201 tgtg\r\n",
        "    371 tgag\r\n",
        "      6 \r\n",
        "      4 cgag\r\n",
        "      3 ttaa\r\n",
        "      3 tgaa\r\n",
        "      2 tggc\r\n",
        "      2 gtgc\r\n",
        "      2 ggca\r\n"
       ]
      }
     ],
     "prompt_number": 120
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "__We get the (less prevalent) PAM region ('TGT') for known I-E CRISPR systems as shown in [Rorek et al., 2013](http://www.annualreviews.org/doi/abs/10.1146/annurev-biochem-072911-172315)!__"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Spacer (crDNA) - protospacer mismatch distribution\n",
      "\n",
      "Let's look at potential spacer targeting by assessing mismatches of the crDNAs and protospacers."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "# This script will make 2 summary files of mismatches for each crDNA-protospacer alignment\n",
      "# Given that spacers can vary in length, the alignment is relative to a user-defined 'SEED' region.\n",
      "# I'm going to use the default SEED region: '-SEED -8 -1'\n",
      "## This SEED region covers the last 8 bp of the alignment\n",
      "\n",
      "cd $1\n",
      "\n",
      "$2'CLdb' arrayBlast -- get_SEEDstats -- --SEED -8 -1 \\\n",
      "    < ./arrayBlast/all_crDNA-proto_aln.fna \\\n",
      "    -prefix ./arrayBlast/all_crDNA-proto_aln\n",
      "\n",
      "\n",
      "## checking output files\n",
      "ls -thlc ./arrayBlast/all_crDNA-proto_aln_SEEDstats-summary.txt\n",
      "ls -thlc ./arrayBlast/all_crDNA-proto_aln_SEEDstats-byPos.txt"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 725K Oct 26 14:23 ./arrayBlast/all_crDNA-proto_aln_SEEDstats-summary.txt\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 2.3K Oct 26 14:23 ./arrayBlast/all_crDNA-proto_aln_SEEDstats-byPos.txt\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "SEED summary file written: './arrayBlast/all_crDNA-proto_aln_SEEDstats-summary.txt'\n",
        "SEED by-position table written: './arrayBlast/all_crDNA-proto_aln_SEEDstats-byPos.txt'\n"
       ]
      }
     ],
     "prompt_number": 123
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%R -i CLdbDir -w 800\n",
      "# plotting mismatches across the protospacer alignments\n",
      "library(ggplot2)\n",
      "library(gridExtra)\n",
      "\n",
      "# loading table\n",
      "filename = './arrayBlast/all_crDNA-proto_aln_SEEDstats-byPos.txt'\n",
      "fullname = paste(c(CLdbDir, filename), collapse='')\n",
      "\n",
      "tbl = read.delim( fullname )\n",
      "tbl = tbl[tbl$region != 'protospacer', ]\n",
      "\n",
      "# mismatches and counts, & normalized mismatches\n",
      "p.mis = ggplot(tbl, aes(pos_rel_SEED, mismatches, group=region, fill=region)) +\n",
      "    geom_bar(stat='identity', position='dodge') +\n",
      "    labs(y='Number\\nof mismatches', x='Alignment position relative to SEED start')\n",
      "p.count = ggplot(tbl, aes(pos_rel_SEED, count, group=region, fill=region)) +  \n",
      "    geom_bar(stat='identity', position='dodge') +\n",
      "    labs(y='Number of alignments\\nwith nucleotide\\nat position', x='Alignment position relative to SEED start')\n",
      "p.norm = ggplot(tbl, aes(pos_rel_SEED, mismatch_norm, group=region, fill=region)) +\n",
      "  geom_bar(stat='identity', position='dodge') +\n",
      "    labs(y='Mismatchs /\\nalignments', x='Alignment position relative to SEED start')\n",
      "\n",
      "grid.arrange(p.count, p.mis, p.norm, ncol=1)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "display_data",
       "text": [
        "NULL\n"
       ]
      },
      {
       "metadata": {},
       "output_type": "display_data",
       "png": "iVBORw0KGgoAAAANSUhEUgAAAyAAAAHgCAMAAABq2fnHAAAC9FBMVEUAAAAAv8QBAQECAgIDAwME\nBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUWFhYX\nFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMlJSUmJiYnJycoKCgpKSkqKior\nKyssLCwtLS0uLi4vLy8wMDAyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6Ojo7Ozs8PDw9PT0+Pj4/\nPz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tMTExNTU1OTk5PT09QUFBRUVFS\nUlJTU1NUVFRVVVVWVlZXV1dZWVlaWlpbW1tcXFxdXV1eXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVm\nZmZnZ2doaGhpaWlra2tsbGxtbW1ubm5vb29wcHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6\nenp7e3t8fHx9fX1+fn5/f3+AgICBgYGCgoKDg4OGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Pj4+Q\nkJCRkZGSkpKTk5OUlJSVlZWWlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKj\no6OkpKSlpaWmpqanp6eoqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2\ntra3t7e4uLi5ubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJ\nycnKysrLy8vMzMzNzc3Ozs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc\n3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u7u7v\n7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4dm34+Pj5+fn6+vr7trH7+/v8/Pz9/f3+/v7///8q\n/YX8AAAgAElEQVR4nO2dfXwU1b3wFw01NjFsSkRQLLFS0kcqlCggAQNVyvXl+gKV3Hotefq0UGv1\n1vKipbe5mGqR4n2wEiQQIIo22PqAvFQJ8moiBEGEkDQxQV6SzW7eQ7Ixyd3Nnn+eMzszm3mf2d3f\nZGezv++Hz+ycM+f85sdyvszM7s4ZG0EQRBVbpBNAECuDgiCIBigIgmiAgiCIBigIgmiAgiCIBigI\ngmggEaS6171uU19kUkEQ6yEWZFV840tTJz0doVwQxHKIBRl52jfmYk1KhHJBEMshFsR+4fhk4kiK\nUC4IYjnEgixJu31T7bT5EcoFQSyHWJC2oiJv1ZqGCOWCIJZDKIjHM85DaQ35FKtOzlWXQmU4OIHj\n1bnrgQNCZ9jQCRwQPMPGduCA9Q0GGgGMfgMIBYmLs8UxZIUaTOFv0dsE/N5Bx6vzOYADQmfo9AIH\nBM+wpRs4oMPIf6sAo98A4lOsueEFU/hboCBhg4IoE95YNQroN+kKfwsUJGxQEGUgR646YkGK70lj\nCDWYwt8CBQkbFESZsMe+IcSCjFtxtoISajCFv8WAIMNE6JWH6ZaDDFhX97UY3XJdcGXwgEMx42gX\n5GZ3WMFQEMiAMZFxlAny6ivecIKhIJABYyLjKBMkI3HEBP1rkMPlxPPy+vXHfLu3vNXpX3AbUBDI\ngDGRcZQJUsGi2aN/86py4vo7XastIqf2+xfcJhQEMmBMZBxlghDirffp9PD1F5eTyjffKWo7VEJc\nm/0LWv3Z0aNHK9tleLv4NfF41SsP0y0HGbC9XfKPo1tuD64MHjAmMlamo1NtywBNJhkhQSzIldkJ\nyWUza3X6HCgnl86Rs9v3nie9r/sXtPLjXbt2ne2S0d/Dr4nHq155mG45yIBdXZJ/HN1yV3Bl8IAx\nkbEK3apbArSY5oQIyTfpz3Snepbep9OHCkLpe+1jevDY5F9wG/AUCzJgTGQcZadY32wlqaQxgS+q\nnG9RQY6eILXba3aQz/f7F9wGFAQyYExkHGWCTPqACrL7+2xB9XyLCuJ+Z/O2Zt/et7Z3+RfcBhQE\nMmBMZBxlghxKWZA0f9SHbMHY+ZYQFAQyYExkHGWCkKaC3HwHty453zIACgIZMCYyjjZBXNUM7Lro\nfMsQKAhkwJjIOMoEedaWOp7CFkTnW4ZAQSADxkTGUSZIYqmwJDzfMgQKAhkwJjKOMkHSL4YVDAWB\nDBgTGUeZICU3LcmhMKt2luSxQQRDQSADxkTGUSbIjPQVKynMamvrn39Y4ii5f3MQwVAQyIAxkXGU\nCZIi/IHLrUwKV24NIhgKAhkwJjKOMkGe3y4ojD1CF4e/HUQwFAQyYExkHGWCzLh27MCkDf/9rRc2\nvvCtdUEEQ0EgA8ZExlEmiPiGqSPPZf3mk2CCOZtkeNr5NfF41SsP0y0HGbCpSfKPo1tuCq4MHjAm\nMlamuUVtywBOKAW0MTovlqEJgfAIAhkwJjKOsiOIQIP0Peks7BZDEwKhIJABYyLjKBNEoEFZ06Ey\nhkPsFkMTAqEgkAFjIuMoE0SggWSqd0MTAqEgkAFjIuMoE0SggWSqd0MTAqEgkAFjIuMoE0SkgWiq\ndyMTAqEgoAFjImOzBCmMC94FRTTnxfI6PAPbRAVlUBDIgDGRsVmCdFQHa4IKGh/zuhYOTxy+0KVQ\nUAMFgQwYExmDCbIx4dkXnQ8k3VvVtWBEtu00PYJ0ZyePySUbr1+Y+Kjuf+1GBUn1M3H2q8w0DI9l\nu4jrp9wTPUUFFCS00WCxHVghYzhBbMsrfj3Z+eTUNTcc+41fkFcTSgtsRzba3iuwfQYlSMGsXad3\n3Zv3j4cW00IS88vFphHsFlEBBQltNFhsB1bIGE6QeC+ZnZyWOubfZpN/+gV5KpP0X/+XjQnksu0o\nlCDj6plhPoG4mbtAvstMd7Wfu2IXFVCQ0EaDxXZghYzhBEkkZPGs5oL/WmUvXeYXZHVC2VZ6BEkE\nFaSMLk7cTJyMIO/ZF+Ussv+N3SIsHC4nAxO74+zuZgWMyYzDEqRuTtKcoy1zE+f7BXEvso/+I4EV\npGDUsteXjco7dPOfmFLN2hVra/hNgYJ/dveBid1xdnezAsZkxiELEqB4lWev7XLoRmgKQr743aIX\nTpKKErakOPWof3b3gYnd+dndGVAQyIAxmXH4gtTOum3saig9ND/mFUw9Kv7l4oFyMjCxO87ublbA\nmMzY0rO7S37AK5h6VPzLRSrIwMTu/Ozu+HwQCzxtI8IBAXZg6eeDlDWVsbBFwdSj4l8uUkEGJnbH\n2d3NChiTGYd8ivW1FBMEkSCYelT8y0UqyMDE7ji7u1kBYzJjSwsykYctiqd6NxIMBYEMGJMZW1qQ\n0zxcGaceDaZsxeE2yAGHuiA822Q1kst3NVAQyIAxmbHFBalYnJWV9choZlV0viW5fEdBFMpWHG6D\nHNDagvh+P2X6A/XVIzIyMpYFllNn6YxpsSB3Zy99sHDqbmZVer5V3etet6kPBVEtW3G4DXJAawvy\nySwf+evT1Zn+0RxYnv+u9mwLYkHimzrvJw0z2UJP/mFSuLaXLayKb3xp6qSnURDVshWH2yAHtJgg\nO372VMbO9vn3zznvX6u6rdjjbZMKQp78KAhBUo+Q9Bb3SLawJP1zUnL3z9jCyNO+MRdrUlAQ1bIV\nh9sgB7SaIHN8x5/4Yw458LB/jZQ8NTGrqtqemZm5jlsygqzYEoQgG667+Psp9/6Ic4L5bWKNnS3Y\nLxyfTBxJKIhq2YrDbZADWk2QHHJhQXYx6fmef+1sLfG9N012BPn3/bJxrC4IudTteXtdO7s+/lO6\nKOGex7Yk7fZNtdPwjkL1shWH2yAHtJoguVQLegT5+CH/2u6sflI9RSpIZVow1yAi3k359Zr/SOE+\n8/UUFXmr1lxFQVTLVhxugxzQioK0Pz6XXoMwa54/ZNw965T/86sFgeW0mSe1x7TW3LyVuUtyzgRK\nOKuJdtmKw22QA1pMEBCEgjiJ+qwlOKtJNA63QQ441AVJdnz7qh+2OI2FLRia1QQffwAZMCYzDvnx\nB19LMUGQpcnXsE/uZIulpaUlRZk72ALOahKN/x8PcsChfgShxwnZ9sY72Vec1SQah9sgB4ykIP8j\nJQwphEgv0qWX4qe4rz5EU5yogYJABozJjEMWZJiU8LwIIBZEdCnOXICkD3+O2ySa4kQFFAQyYExm\nbHFBRJfipQwV/LwmvqqDlf06wVAQyIAxmbHFBVG/FK+YnDLlxkn4+APQ0RDhHVgxY4sLIroUFz23\nc/qKPtK3YoZ2MBQEMmBMZmxxQUSX4qLndtob6cKFH/OCjoYI78CKGUMKEtodUpqCiC7FRc/tfDKP\nLt54AgWBHA0R3oEVM4YUJLQ7pLQFESJ6bufPh096dJLtwaysLBQEbDREeAdWzBhSkNDukDIuiOiB\nhYUBUBCw0RDhHVgxY0hBQrtDyrggWs/t9Ly8fv0xfPyBqQFjMmNIQUK7Q0pTkDvPGezm+jtd4OMP\nTA0YkxlDChLaHVKaguRm96i06xQsKZVvvlPUho8/MDVgTGYMKUhod0hpCpJpv2584LJDgHjyasql\nc+TsdtHjD/66YcOGk30yfB5+TZy+XnmYbjnIgH19kn8c3XJfcGXwgDGZMYdHPpKktOsKAoJYkLMs\nslbiyatZ+l4TPf7gUg0F7wcBDBiTGXP3g7QFfT/I4AgieqhU8UzBTNaSyauPniC12/HxB6YGjMmM\nIU+xTBBE8FApQsbllks/xQrM2ut+Z/O2Znz8gakBYzJjiwsieKgUITd1C7YIZu1VBwWBDBiTGVtc\nEMFDpQh5SfhNumDWXnVQEMiAMZlxyIKYhVgQwUOl0tImXCP4Jl00a68aKAhkwJjM2OKCCB4qVVFx\n0v9FOvepsWjWXjVQEMiAMZmxxQURPFRK8tWHaNZeNVAQyIAxmbHVBfFWHPyS/ZxX+tWHcNZeNVAQ\nyIAxmbHFBTnzvZQpKXdxN4QYem6nCBQEMmBMZmxxQe5mbqxdmhlqMBQEMmBMZmxxQZIGbqw1+NxO\nESgIZMCYzNjigizcSBcbHmdWDT63UwQKAhkwJjO2tCBPPjnfNmXBlGsXs8WtQaeAgkAGjMmMLS2I\n5K7aOYl3PLtL53MrMfVy+pr5NfF41SsP0y0HGbC+XvKPo1uuD64MHjAmM2ZpcCkMJSmAFmgg+ZjX\nVc3AFfrK/nuh/Z4ggjW4ZHja+DXxeNUrD9MtBxnQ5ZL84+iWXcGVwQPGZMYsjc3ykSSlAUoBbcSC\nPGtLHU9hC93HVj8+5n/9IohgeIoFGTAmM7b0KRYhiaWCwjdueGZnU1DBUBDIgDGZscUFSb8oKKx+\nfFz6rwqriXFQEMiAMZmxxQUpuWlJDoUv+qqeTdR6yqcUFAQyYExmbHFBZqSvWElhC3tenJU0K+dw\nEMFQEMiAMZmxxQVJaREU7lr+j04SFCgIZMCYzNjigjy/PaxgKAhkwJjM2OKCzLh2bJrSvFgGQUEg\nA8ZkxhYXRGs6XgOgIJABYzJjiwsSJigIZMCYzNjigkxjCTUYCgIZMCYztrggpaWlJUWZO4IJgI8/\nMCtgTGZscUH8NN4ZTAB8/IFZAWMy42gQ5FSSvE4dfPyBWQFjMmOLC8JcgKQPfy6YAPzjD9bl5OQc\n02ooHq965WG65SADEiL5x9Etk+DK4AEx46+1xlOQX2KHiuwapLS0wqfSVhH+8Qc9brf7skNGX4u8\nLiyageM5iBM4IHSGjV7ggOAZtvUAB3Q26reJzA1TwWP48QdAQMer8zmAA0Jn6PQCBwTPsKUbOKBF\nT7EyeYIJYPjxB0CgIOGDghhHKMgHfnK/OTnUYAp/CxQkbFAQZQBGvwGkp1hXf2v/syfUYAp/i0OV\nwO9dA3C8uj0XgQNCZ1jzEXBA8AzPHQMOWG/kP60wR75BJILs/PZjl0Djv/klaDgTeGmQPg4JGeef\nI52BHp+H9yNwSyMS5PLj43YBx0dBwgYFiSRCQdZ+axn4YEFBwgYFiSRCQWzXxLMAxt8zSKeKoVPY\nrd8morQG9du4SFBzINIZmIdQkMB3MBHLBkEsBuj9IAgy1EBBEEQDcwUR3ixiTQ6XWzpJz/tb8q5Y\nOUHSu31zfoulMwwPcwUR3ixiRfo3ryq3dJKVe0jdRisnSE4cIqf2WjrD8DBXEOHNIlbE119cbukk\n6xtJa76VEyTODnL6gKUzDA9zBeFvFrEuB8otnuSVvCprJ0jefaXF4hmGg7mC8DeLWBcqiJWT9BVv\ndlg6QdLTT2q3WTrD8DBXEOHNItaECmLlJM/t6Lf4u/jRKXKpwNIZhofJn2IJbhaxJlQQKye5c+36\n9W9bOUHSsTX/zXpLZxge+D0IgmiAgiCIBigIgmiAgijjm5DSR8jpiaQs3WiXOLU7MZkQcR6dQHRX\nA4GU2orCn5k7YuS/VhGSYmOoDrzabbb46QfVs1LNEVEBBVHmszHj/+EftU17jHZRHXxMiDiPTiCR\nIEptheG9t66sqV860UdSPm6leAOv9iOtX61LPKmaFQoSLCiIMkuXrswOHEEKUlO3pJKKjFdvue0A\nqZjx25SMkrtueI6QIz9ImHeFr59rG9dJSOHPn7LPqCTkbxNGzHcRzy+TU15iQtCNh9L5Wq4D5Wxm\n7p1cFEaQN2+7flol13beBkJWZ3EbKUx4tjvloq2DWvJwK0kpY9PlX+2n6eLFHzOr7L6ZbmxYZl9s\njkgQoCCK9N965qy9hxPkixuP182kgiS+3LX8HlJx7dvN6Td9VWpzNaXsavnVbL6e/d+5cPgbruWT\nfbX2j5qzs0hR2oVT8V8GTrG4Wr4DFcSefY6LQnd16brDjdmLubbrHyFkxvvcRoY4D9ed4vn+vH1u\nZkVRkE++w6yy+6bduLDMvvAIEjQoiCJHfkDIHTs5QX63jJCdVJARHnI2jVSMJWTFEkJSq7csIKQ7\nwcvVc4JMIqRvZNXaRYS4vuEtuv24r7E3IAhXy3egglzXQ7godFfdF0nn0iyubf0N3Y7kbm4j0zbO\nw3VnCj1v/EvKvBNUjES73X7LwKtfkGr/LaHsvmk3LiyzLxQkaFAQRZ6+fvTo+J9wgix6nZ5sUUEm\nEFKRxvwhK3MIGV+dm5Samppcz9VzgjxKF1MOrmAepZ3o8OZPHpPrDgjC1fIdqCDjCeGi0F15/vOu\n+x7kBSHT921YxG9k2sZ5uO500dtNHSmIP0FSdl64cOErEngVHEHYfdNuXFhmXyhI0KAgSvTdeMjh\nOJjYxQrywnJCdqX6x7RIkPz5dBiW+bj6gSOI58bKtfQCxjXcU1VLLk5dN3AEYWv5DsR/HOGi0F1t\nn9JMCgOC/OmZB/bxG5m29AjCdqfrb/2QqZm9QesahN037caF9R+zUJBgQUGU2PcdOiZ9qTtYQU6O\nKqvPVBDEMWpP49IMEhCklfYstOU1vjCp/8sRxS0/fYKsmdZQOznfL0grXXK1YkG4KHRX6+51O6c+\nwrUl/xx7Sx+/kSGuletOcSb/oaryjaQq7tOr3sCr/UjrxdcTP2PasPum3biwrCCtEXpHoxYURIlF\nzzPL5xZwn2Ll3Twhb6JcEPLhxIQ5NYHxvjCJ+RTrwR8nTT9PyHsTkh5zkquPJY78ZR8TYmES8ykW\nWysWhItCd9U2d+SMD0Zv49qSO5YENjLQ8Gx3hqqHRifds4//HqQw8Mp8DzKN/R6E3TftVseG9e/L\nnyMSBCiIPhWH6ECdbaxtYZa5uSCDDAqiT8nNTvfDq421RUGGGCiIAV6+ZewvDP6aGwUZYqAgCKIB\nCoIgGqAgCKIBCoIgGqAgCKIBCoIgGqAgCKIBCoIgGqAgCKIBCoIgGqAgCKIBCoIgGqAgCKIBCoIg\nGqAgCKIBCoIgGqAgCKIBCoIgGqAgCKIBCoIgGqAgCKIBCoIgGoQmSJ2cene9Qm2YOE0IeRU+Zr0J\neTa3wcd0OOBjtjfBx3QZagVsggpwghATBDHhzW/sg49Z3wgfs70LPqazAT6muxU+ZqOhoQRsggoo\nCAAoCCwoiFFQEFBQkOBBQQBAQWBBQYyCgoASjCBfi1Fth4IooZAuCgILCqIHsAkqoCAAoCCwoCBG\nQUFAQUGCBwUBAAWBBQUxCgoCCgoSPIYE6d2+Ob/Ft3vLW/wTIBXSRUFgQUH0MNEKAYYEOXGInNpb\nW0RO7ecqFNJFQWBBQfQw0QoBhgRxdpDTBw6VENdmWqg5d+5cVYuMViKvC5sOE0J64WOakae7Bz5m\nW5vxthJBVNv1dgEkJsHQ2+ky2QwOg9cg777Ssvc86X2dru7eunVrWY8colAXLn0mhPTBxzQjT48X\nPmZvr/G2EkFU2/V7ABKTYOjtbDPViwBaglT3utdt6qMrPf2kdtvH9AiyiduicMDDUyxY8BRLj0Gw\ng2gKsiq+8aWpk56max+dIpcKanaQz/EaRBEUBJYoEWTkad+YizUpdK1ja/6b9b69b23nHxaukC4K\nAgsKosdg6KEpiP3C8cnEkaS0SSFdFAQWFEQPs5QQoyHIkrTbN9VOm6+0SSFdFAQWFEQPs5QQoyGI\np6jIW7XmqtImhXRREFhQED3MUkKM5se83nqf8gaFdFEQWFAQPczQQY6GIFdmJySXzaxV2qSQLgoC\nCwqih1lKiNEQZO4z3amepfcpbVJIFwWBBQXRwywlxGgI8s1WkkoaE5Q2KaSLgsCCguhhlhJiNASZ\n9AEVZPf3lTYppIuCwIKC6GGWEmI0BDmUsiBp/qgPlTYppIuCwIKC6GGWEmK0PsVqKsjNdyhuUUgX\nBYEFBdHDHCGkhHZHYb0cB3Eo1IZJkwkh++BjOhrhY7Z3wcd0Oo23lQii2q67DSAxCU2GhhKwCSpo\nCFI8cyKD0qYGl4xG0iivDJcW+JCtHviYjc3wMTu74WM2NRlvKxFEtV1PB0BiEgz9szeYpYQYDUHG\n5ZZXUJQ2KRzw8BQLFjzF0sMsJcRoCHJTt+omhXRREFhQED1MsEEBsSB3nhMUXnrFq9ZLIV0UBBYU\nRA/NcV0YF5YWA4gFyc3u4VfT0iZcM2JCWlqaUi+FdFEQWFAQPTTHdUc1hB1EKkim/brxnBMVPCiI\nLigILKEIsjHh2RedDyTdW9W1YES27TQ9gnRnJ4/JJRuvX5j4qAdKkLMsbKEn/zApXNuLguiCgsAS\nkiC25RW/nux8cuqaG479xi/IqwmlBbYjG23vFdg+gxJE9Av3Jemfk5K7f4aC6IKCwBKSIPFeMjs5\nLXXMv80m//QL8lQm6b/+LxsTyGXbUShBRL9wH1lDFzV2FEQXFASWkARJJGTxrOaC/1plL13mF2R1\nQtlWegRJhBRE9Av38Z/SRcl4FEQXFASWUAWpm5M052jL3MT5fkHci+yj/0hgBRH9wv3dlF+v+Y+U\nbSiILigILGF9ilW8yrPXdjl0IzQFEf3CvbMyd0nOmU6FTiiIGBQElrAEqZ1129jVYH5IBBH8wt3j\nGeehtDLT/nje35J3BWd3VwUFgcW636QLfuEeF2eLY8ii65V7SN1GnN1dFRQEFgsL4q04+CX/Oe/c\nQG19I2nN52d3/3T//v1fXJVDFOrCxW1CyH74mGbk2dsHH7Or03hbiSCq7Tw9AIlJMPR2NotH7tdS\nTBHkzPdSpqTcVSOsYi/Sr+RV8bO7oyAKQeFDoiA6RESQu1f0kb6lmWyhYnFWVtYjo+mar3izg+Ds\n7qrgKRYsoZxiDY4gSY104RrB2ZK99MHCqbvp2rkd/YTg7O6qoCCwWFeQhRvpYsPjbCG+qfN+0jCT\nru1cu3792zi7uyooCCxWFeTJJ+fbpiyYcu1itph6hKS3uEcq9VJIFwWBBQXRI1hBfL+fMv2B+uoR\nGRkZywLLqbPKjAtSyMMWN1x38fdT7v0RCqILCgKLOYJ8MstH/vp0tf8Ke2B5/rtuw4JQXNUMXOFS\nt+ftde0oiC4oCCxAguz42VMZO9vn3z/nvH+t6rZij7dNKgh58qMgBHnWljqeomQLCqIOCgILlCBz\nfMef+GMOOfCwf42UPDUxq6ranpmZuY5bMoKs2BKEIImlqragIOqgILBACZJDLizILiY93/Ovna0l\nvvemyY4g/75fNrzVBUm/qGYLCqIBCgILlCC5VAt6BPn4If/a7qx+Uj1FKkhlWjDXICU3LcmhKNiC\ngmiAgsACKUj743PpNQiz5vlDxt2zTvk/v1oQWE6beVLTD4kgM9JXrKQo2IKCaICCwGLV70EISWlR\nswUF0QAFgcW6gjy/Xc0WFEQDFAQW6woy49qxaYG54kS2iMHJq0Xg5NWwhDJ59eAIIporTmSLGIXJ\n6PHxB7Dg4w/0EA/J/5FiiiDqtohROODhKRYseIqlh3hIDpNiiiDTWLhSda973aY+FEQXFAQW6wpS\nWlpaUpS5gy2sim98aeqkp1EQXVAQWKwriJ/GO9nXkad9Yy7WpKAguqAgsFhckFNJ7Kv9wvHJxJGE\nguiCgsBiXUGYC5D04c+xhSVpt2+qnTYfBdEl+gTRH/5DQJDQ7pDSFKSUoYKb98dTVOStWnMVBdEF\nBYEFSJDQ7pDSFMQoCumiILCgIHroChLaHVIagmTyMIX0PeksKIguKIgxjMaEugYJ6Q4pDUE+8JP7\nzclMoaypjAUF0QUFMcYgCxLaHVIagjBc/a39z/wT3QS33B4uJzh5tSooiDEGWZDQ7pDSFmTntx+7\nxK8P3HLbv3lVOcHJq1WJCUFUulhYkNDukNIS5PLj43YNlAZuufX1F5cTfvJqFEQGCmKMQRYEBKEg\na7+1TPi4HOEttwfKCT959e6tW7eW9cghCnXh0mdCSB98TDPy9HjhY/b28muSoarQVtZCpUu/x/j+\n9ffKYujtbBt8QWzXxLOwReEtt1QQfvLqmnPnzlW1yGgl8rqw6TAhpBc+phl5unvgY7a18WuSoarQ\nVtZCpUtvl/H96++VxdDb6Rp8QRw8bFF4yy0VBCevVgVPsYyhv1cWq55iSRDecksFwcmrVUFBjDHE\nBNG45VYhXRQEliEgiNEQMiz8CDYhGrfcKqSLgsCCguhhlhJi8JZbAFAQRYa6IBoopIuCwIKC6AFs\nggooCAAoiCJDWxAncaltQkEkoCCKDG1Bkh3fZh+3q7RRIV0UBBYURA/zpBCiKsjS5GvsfpQ2KqSL\ngsCCguhhnhRCNK5BHlPfpJAuCgILCqIHvAxKaF6kex0e5Q0K6WoJIvmO08BfPvgeRkFBGFAQo2gI\n4lo4PHH4QsVLdYV0URBYUBA9zFJCjNYpVraLuH6qOO2Ps0lGM2mWV3JIhrtqu3B6GKXdAxiMo7kF\nPmZXN3zMlsA/kWSoKrSVtVDp0tupujujIWS0GvnLOM1SQoyGIEnMjxWbRihtUvBZcASR/fePR5Dg\nwSOIHiYZIUFDkO8yP27fH8JvsVAQAFAQPUwyQoKGIO/ZF+Ussv9NaZNCuigILCiIHmYpIUbrU6ya\ntSvW1ihuUUg3LEEAlDLKUBTE6MBDQYLHjN9ioSAAoCB6AJugAgoCAApiLC0UhEV3uOv7ELwgRnvE\nhCAqI9Fqgnwta8GCgqAguqAgegCboAIKAgAKElpaKAgKEjIaguiPXRQEDBQEABQktDyHliDGZ3dH\nQQBAQfQwRwgpxgUxPrv7YAgSslIoiGKIEHbCgoLwGJ/dHQUBAAXRwxwhpBgXhJ/dPS8nJ+egZkvJ\nWJVX6DaQVwS9jyGMZFgZqdANEcJOws9TKkhQ70KnfhMIjAvCz+7e43a7LztkNJAGeWW4NJsQ0gMf\ns6EJPmZHF3xMlws+Znc7fMwmI0Op3jwphBgXBGd3VyXS34MYZeAUCw6rPOXWLIL4FAtnd1cDBYEl\nOgURopAuCgILCqIHsAkqgAlyefdl4LeJYsI/aNXH8DHrHPAhzxyHj+kwIc9PzsLHNPbPDmyCCqEJ\nokBfjsoUQRbjwvpIZ2CMY7v021iBouAeGht1oCAWBQWxBiiIRUFBrAGYIJ6tXqhQptIQJUuqjxsA\nABe0SURBVAPvXEmkMzDGoapIZ2AuYIIgyFAEBUEQDVAQBNEAShDh3SIW5nB5NGTqeX9L3pUoyLN3\n++b8lijIMxygBBHeLWJZ+jevKo+GTCv3kLqNUZDniUPk1N4oyDMcoAQR3i1iWXz9xeXRkGl9I2nN\nj4I8nR3k9IEoyDMcoATh7xaxOAfKoyPTK3lVUZHnu6+0REWeoQMlCH+3iMWhgkRBpr7izY5oyLOn\nn9Rui4I8wwFKEOHdIhaGChIFmZ7b0R8V7+hHp8ilgijIMxzAPsUS3C1iYaggUZDpzrXr178dBXl2\nbM1/sz4K8gwH/B4EQTRAQRBEAxQEQTRAQQT4JqT0EXJ6IilLN9olTu1H/kyIOI9OILqrgUBKbUXh\nz8wdMfJfqwhJsTFUB17tNlv89IOq3dRzRPRBQQR8Nmb8P/yjtmmP0S6qg48JEefRCSQSRKmtMLz3\n1pU19Usn+kjKx60Ub+DVfqT1q3WJJ1W6oSBhgYIIWLp0ZXbgCFKQmrollVRkvHrLbQdIxYzfpmSU\n3HXDc4Qc+UHCvCt8/VzbuE5CCn/+lH1GJSF/mzBivot4fpmc8hITgm48lM7Xch0oZzNz7+SiMIK8\nedv10yq5tvM2ELI6i9tIYcKz3SkXbR3UkodbSUoZmy7/aj9NFy/+mFll9810Y8My+2JzREIDBRmg\n/9YzZ+09nCBf3Hi8biYVJPHlruX3kIpr325Ov+mrUpurKWVXy69m8/Xs/86Fw99wLZ/sq7V/1Jyd\nRYrSLpyK/zJwisXV8h2oIPbsc1wUuqtL1x1uzF7MtV3/CCEz3uc2MsR5uO4Uz/fn7XMzK4qCfPId\nZpXdN+3GhWX2hUeQcEBBBjjyA0Lu2MkJ8rtlhOykgozwkLNppGIsISuWEJJavWUBId0JXq6eE2QS\nIX0jq9YuIsT1DW/R7cd9jb0BQbhavgMV5LoewkWhu+q+SDqXZnFt62/odiR3cxuZtnEerjtT6Hnj\nX1LmnaBiJNrt9lsGXv2CVMczTdh9025cWGZfKEg4oCADPH396NHxP+EEWfQ6PdmigkwgpCKN+UNW\n5hAyvjo3KTU1Nbmeq+cEeZQuphxcQRuQRIc3f/KYXHdAEK6W70AFGU8IF4XuyvOfd933IC8Imb5v\nwyJ+I9M2zsN1p4vebupIQfwJkrLzwoULX5HAq+AIwu6bduPCMvtCQcIBBQnQd+Mhh+NgYhcryAvL\nCdmV6h/TIkHy59NhWObj6geOIJ4bK9fSCxjXcE9VLbk4dd3AEYSt5TsQ/3GEi0J3tX1KMykMCPKn\nZx7Yx29k2tIjCNudrr/1Q6Zm9gataxB237QbF9Z/zEJBwgAFCbDvO3RM+lJ3sIKcHFVWn6kgiGPU\nnsalGSQgSCvtWWjLa3xhUv+XI4pbfvoEWTOtoXZyvl+QVrrkasWCcFHortbd63ZOfYRrS/459pY+\nfiNDXCvXneJM/kNV5RtJVdynV72BV/uR1ouvJ37GtGH3TbtxYVlBWiP0jg4FUJAAi55nls8t4D7F\nyrt5Qt5EuSDkw4kJc2oC431hEvMp1oM/Tpp+npD3JiQ95iRXH0sc+cs+JsTCJOZTLLZWLAgXhe6q\nbe7IGR+M3sa1JXcsCWxkoOHZ7gxVD41Oumcf/z1IYeCV+R5kGvs9CLtv2q2ODevflz9HJDRQEBUq\nDtGBOttY28Isc3NBIgcKokLJzU73w6uNtUVBhi4oiBov3zL2FwZ/xo2CDF1QEATRAAVBEA1QEATR\nAAVBEA1QEATRAAVBEA1QEATRAAVBEA1QEATRAAVBEA1QEATRAAVBEA1QEATRAAVBEA1QEATRAAVB\nEA1QEATRAAVBEA1QEATRAAVBEA1QEATRAAVBEA0ABKmT09ymUBkeTuiA9e566JDgOda1NUFHdDRA\nR3RehY5o7I0Mf+QawRxB2rqA37G6OvCRUk/ABQHPsa4L/H8aB7jFjR7oiMbeyPBHrhFQEEBQECBQ\nkOBBQWBAQYIDBQEEBQEi2gTx7d7yVif/6nl/S94VvoZBIXcUBIiICvK1GLVmKAipLSKn9vOvlXtI\n3Ua+hkEhdxQECBREHfOkEGJIkEMlxLWZf61vJK35fA2DQu4oCBAoiDrmSSHEkCB7z5Pe1wder+RV\n8TUFr7zyypF+GT6fvC5M4CMS8Igm/K0j+UZKBFFtF5k38qq5YvAYEuRjerzYxL/6ijc7AjXtLS0t\nXzXI6HDL68KkFTqgkzihQ4Ln2NDdAR3R1WS4qUQQtWbNHpDEhBh5Ix3misFjSJCaHeTz/fzruR39\nAzUMCkc/PMUCAk+x1DFPCiHGPsXa+9b2Lmce+7pz7fr1b/vXuK0KuaMgQKAg6phohQD8HgQQFAQI\nFCR4UBAYUJDgQEEAQUGAQEGCBwWBAQUJDhQEEBQECBQkeFAQGFCQ4EBBAEFBgEBBggcFgQEFCQ4A\nQZxNMjp75HVh0godsJk0Q4cEz7GppxM6YnOL4aYSQdSatXlBEhNi5I10hj9yjSAU5KAnpBAKcuMR\nBAg8gqgDI4AeQkHe+MmSXe7gQyjkjoIAgYKoA+aAJuJTrJbCn2a/3RpkCIXcURAgUBB14CTQQnYN\n4v5/v/gJt1rd6163qU83hELuKAgQKIg60Cooo3SRzl2KrIpvfGnqpKd1QyjkjoIAgYKoAyuCGhqf\nYo087RtzsSZFN4RC7igIECiIOoAWaKAhiP3C8cnEkaQbQiF3FAQIFEQdQAs0kAgivOxYknb7ptpp\n83VDKOSOggCBgqhjgg0KiAURXXa0FRV5q9Y0EMm8WIQcLieel9evP4aCSEBBgLCsIILLDo9nnIfS\nypxiiebF6t+8qpy4/j7QSSF3FAQIFESdSAgiuOyIi7PFMWQRybxYvv7iclL55jtFbbTGRROtdcm4\n2i2vC5M26ICNpBE6JHiOru6r0BGbmg03lQii1qzVA5KYECNvZEMkBBFddswNVEvmxTpQTi6dI2e3\n05rtr732WolXRr9PXmc5SKQTMICvP4I7lwii1ixC/9gdmuO6MM4UQTz+yw7ZlFyiebH8glD6XuO2\nKhz98BQLCDzFUkdzXHdUw/ih9TFv8T1pDEQyL5ZfkKMnSO12FEQCCgJECIJsTHj2RecDSfdWdS0Y\nkW07TY8g3dnJY3LJxusXJj4a2q9wFQQpnjmRgS2MW3G2gkIk82L5BXG/s3lbMwoiAQUBIhRBbMsr\nfj3Z+eTUNTcc+41fkFcTSgtsRzba3iuwfQYlyLjccs4Jys0Gf9mrkDsKAgQKoo5EkHgvmZ2cljrm\n32aTf/oFeSqT9F//l40J5LLtKJQgN3ULCq++4kVBggIFASIUQRIJWTyrueC/VtlLl/kFWZ1QtpUe\nQRIhBXlJ6ERG4ogJ7DUICmIMFASIEAWpm5M052jL3MT5fkHci+yj/0ggBUlLm3CNwIkKFhTEMCgI\nEOF8ilW8yrPXdjl0IyQIBamokDjhrfcZCKGQOwoCBAqijspwrJ1129jVYTkhQnyK1ZN/mBSu7WUL\nV2YnJJfNrNUNoZA7CgIECqIOnARaSL5JT/+clNz9M7Yw95nuVM/S+3RDKOSOggCBgqhjhg5yJD9W\nrKGLGjtb+GYrSSWNCboh6uW0uxUqw6MJOqCDOKBDgudY726HjtjgMtxUIohasyYPSGKikAbaSAT5\nWoopgoz/lC5KxrOFSR9QQXZ/XzeEs1FGZ7e8LkxaoQM2kSbokOA5NvZ0QkdsajbcVDLi1Jq1eUES\nE2LkjZT8WHFwBHk35ddr/iNlG1s4lLIgaf6oD3VDKBz98BQLCDzFUicSgnRW5i7JOdPJlZoKcvMN\nPCpRIXcUBAgURB3xKBwMQYT3SDG4qhlQEMOgIEBYVBDhPVKUZ22p4ykoiGFQECBMEcT3+ynTH6iv\nHpGRkbEssJw6q8y4IMJ7pCiJpbpuoCAiUBAgTBHkk1k+8tenqzOZ9YHl+e9q/yZX6X4Q7iI9/SIK\nEhwoCBAwguz42VMZO9vn3z/nvH+t6rZij7dNKgh58qMgBKlYnJWV9chotlBy05IcCgpiGBQECCBB\n5viOP/HHHHLgYf8aKXlqYlZVtT0zM3Mdt2QEWbElCEHuzl76YOHU3WxhRvqKlRQUxDAoCBBAguSQ\nCwuyi0nP9/xrZ2uJ771psiPIv+8PQpD4ps77ScNMtpDSwlfL58Xia1AQISgIEECC5FIt6BHk44f8\na7uz+kn1FKkglWnBXIOkHiHpLe6RbOF5/qZzhXmx+BoURAgKAgSgIO2Pz6XXIMya5w8Zd8865f/8\nakFgOW3mSU0/JIJsuO7i76fc+yO2MOPasdykDfJ5sfia00ePHi1vl9HdJ68Lky7wiAQ8InyOfd3Q\nETuuGm4qGXFqzTr7QRITYuSNbNIVBATJp1iXuj1vr2tn1wduDpHPi8XXHN61a9eZLhm9HnldmHSD\nRyTgEeFz9PSCh3QbbikZcWrNuvtB8hKFNNCmRTxyB0kQ5S/P5fNi8TUMCkc/PMUCAk+x1ImEIKIv\nz7XmxeJrUBAhKAgQlhVE9OW51rxY3BoKIgIFASIEQf5HiimCiL48x3mxggUFASIEQYZJCUuLAcSC\niL48x3mxggUFAcKygoi+PMd5sYIFBQHCsoKkCD87w3mxggUFAcKyggx8eR4ECrmjIECgIOqIR+Eg\nnWIFvjynpPqZOPvVLoV+KIgC0SWI7viPdkFCu0NKUxDRWVXBrF2nd92b94+HFqMgxkBBgIARJLQ7\npDQFET0Gelw9M/onEPdY7RAN8ufHmfCMwhbogCY8oxA8R1OfUSgZ//Kmug1YTHhGoZE3UjKdiIIg\nod0hpSmI6DHQ45hj0YmbiVNHEIecDrdCZXg0QwdsIA3QIcFzdLjboSM6G/k1yfiXN9VtwNLsMbxz\ngxENvZH1uoKEdoeUpiCCx0DTU6xRy15fNirv0M1/0g6hcPTDUywghtgplsGIQKdYod0hpSmI4DHQ\nlC9+t+iFk6SiRCeEQu4oCBAoiDq6goR2h5SmIKLHQBtFIXcUBAhLCaLSwbKChHaHlKYggsdAp+9J\nZ0FBDIOC6DC4goCg+hjosqYyFt0QCrmjIECgIOqIR+FgCGL8oCFCIXcUBAgURB3xKBwMQUQHjYk8\nKIhhUBAdolyQCzxM4TQPkUz74194Xl6//hgKIgEF0SHKBbHFp7AIGzDzkIqm/fEvXH8faKGQOwoC\nRJQLYjCCDIs+gu3pW+9bL/gCf2AeUtG0P/5F5ZvvFLWhIBJQkNAiyLCoIKT/6G9uv/f/XuFKA/OQ\niqb98S8unSNnmZ/G7966dWtZjwyPV14XJn3gEQl4RPgcvR7oiL29/JpktMqbShuodOjzqe7MYAQZ\nRt7INulQHgRBKL6TL6bdw64OzEMqmvaHL/S9Rhc1586dq2qR0dUrrwuTDuiArQQ6InyOLb1d0BFb\n2/g1yWiVN5U2UOnQ4VXdmcEIMoy8ka7ICNL34S/G3M+uDsxDKpr2x784eoLU8jdXKRz98BQLCDzF\nUicCgvTu+z9j5uU3cqWBeUhF0/74F+53Nm9rRkEkoCChRZBhUUH+95iHCpoFZeE8pBoo5I6CAIGC\nqAPugiKij3mHJ7IEF0IhdxQECBREHVAPVBEKErgXJbgQCrmjIECgIOqAeqCK6o8VjaOQu4Yg4q87\nDbwPw4LuYQwUBAUxAAoCCAoSWgQZKAgKYgwURJ3wR64RzBZENrxRkGBAQdQJf+QaAQUBBAUJNoJK\nCkNLEGeTjM4efk0yvKUV8q4yZBGAaCbNYLE4WqEDNvV0QkdsbuHXJKNV3lTaQKVDm1d1Z7oRVFIw\n8kY6wx+5RrDcEUSnAx5B1FD5/10KHkGCAwUBBAUJNgIKgoKEBQqiTvgj1wgRF0TaIHxBDHYY6oKo\nDGcUJDhQEEBQkGAjoCAoSFigIOqEP3KNgIIAMpiC6Iw1FAQIFAQQFCTYCENEEPm8WHwNg0LuKAgQ\nKIg65kkhxJAg8nmx+BoGhdwHVRDdCMqgICiIAQwJIp8Xi69hUMgdBQECBVHHPCmEGBJEPi8WX5OX\nk5NzUKurZLRKK+QNdDoozCipG2FIIhlr8gpZA90Ieg10OwQfIeiIA3TqN4HAkCDyebH4mh63231Z\n/vw4fEYhECY8o9AFHTGIZxQaDmmgTb3eqIXBkCDyebH4GgaFox/ekw6EmT93h8Iaj4E2C2OfYsnm\nxfIvuK0KuaMgQKAg6phohYBBv+U2VFAQGFCQ4DBHkDOfAr9jdXUN0AEv774MHRI8x7pPv4COWO+A\njlh9ADqisTcy/JFrBABBFCh935SwoPTleCKdgj5//zTSGehz8S+RzsBMUBBLg4JEGhTE0qAgkcYc\nQc4f028TaTxbvZFOQZ9jFZHOQB/nzkhnYCbmCIIgQwQUBEE0QEEQRANTBBHeLWJRDpdbPkvP+1vy\nrlg8yd7tm/NbLJ5jeJgiiPBuEUvSv3lVueWzrNxD6jZaPMkTh8ipvRbPMTxMEUR4t4gl8fUXl1s+\ny/pG0ppv8SSdHeT0AYvnGB6mCMLfLWJhDpRHQZZX8qosn+S7r7RYPsdwMEUQ/m4RC0MFsXqWvuLN\nDqsn2dNPardZPMfwMEUQ4d0iFoUKYvUsz+3ot/xb+dEpcqnA4jmGhzmfYgnuFrEoVBCrZ7lz7fr1\nb1s8yY6t+W/WWzzH8MDvQRBEAxQEQTRAQRBEAxREjm9CSh8hpyeSsnSjXeLUfjrPhIjz6ASiuxoI\npNRWFP7M3BEj/7WKkBQbQ3Xg1W6zxU8/qNpNPUdEAxREzmdjxv/DP2qb9hjtojr4mBBxHp1AIkGU\n2grDe29dWVO/dKKPpHzcSvEGXu1HWr9al3hSNSsUJBRQEDlLl67MDhxBClJTt6SSioxXb7ntAKmY\n8duUjJK7bniOkCM/SJh3ha+faxvXSUjhz5+yz6gk5G8TRsx3Ec8vk1NeYkLQjYfS+VquA+VsZu6d\nXBRGkDdvu35aJdd23gZCVmdxGylMeLY75aKtg1rycCtJKWPT5V/tp+nixR8zq+y+mW5sWGZfbI5I\nkKAgMvpvPXPW3sMJ8sWNx+tmUkESX+5afg+puPbt5vSbviq1uZpSdrX8ajZfz/7vXDj8Ddfyyb5a\n+0fN2VmkKO3CqfgvA6dYXC3fgQpizz7HRaG7unTd4cbsxVzb9Y8QMuN9biNDnIfrTvF8f94+N7Oi\nKMgn32FW2X3TblxYZl94BAkJFETGkR8QcsdOTpDfLSNkJxVkhIecTSMVYwlZsYSQ1OotCwjpTvBy\n9ZwgkwjpG1m1dhEhrm94i24/7mvsDQjC1fIdqCDX9RAuCt1V90XSuTSLa1t/Q7cjuZvbyLSN83Dd\nmULPG/+SMu8EFSPRbrffMvDqF6Q6nmnC7pt248Iy+0JBQgIFkfH09aNHx/+EE2TR6/RkiwoygZCK\nNOYPWZlDyPjq3KTU1NTkeq6eE+RRuphycAVtQBId3vzJY3LdAUG4Wr4DFWQ8IVwUuivPf95134O8\nIGT6vg2L+I1M2zgP150uerupIwXxJ0jKzgsXLnxFAq+CIwi7b9qNC8vsCwUJCRRESt+NhxyOg4ld\nrCAvLCdkV6p/TIsEyZ9Ph2GZj6sfOIJ4bqxcSy9gXMM9VbXk4tR1A0cQtpbvQPzHES4K3dX2Kc2k\nMCDIn555YB+/kWlLjyBsd7r+1g+ZmtkbtK5B2H3TblxY/zELBQkFFETKvu/QMelL3cEKcnJUWX2m\ngiCOUXsal2aQgCCttGehLa/xhUn9X44obvnpE2TNtIbayfl+QVrpkqsVC8JFobtad6/bOfURri35\n59hb+viNDHGtXHeKM/kPVZVvJFVxn171Bl7tR1ovvp74GdOG3TftxoVlBWmN0Dsa1aAgUhY9zyyf\nW8B9ipV384S8iXJByIcTE+bUBMb7wiTmU6wHf5w0/Twh701IesxJrj6WOPKXfUyIhUnMp1hsrVgQ\nLgrdVdvckTM+GL2Na0vuWBLYyEDDs90Zqh4anXTPPv57kMLAK/M9yDT2exB237RbHRvWvy9/jkiQ\noCDaVByiA3W2sbaFWebmgkQAFESbkpud7odXG2uLggxBUBAdXr5l7C8M/pYbBRmCoCAIogEKgiAa\noCAIogEKgiAaoCAIogEKgiAaoCAIogEKgiAa/H/hYhSIv+g1bAAAAABJRU5ErkJggg==\n"
      }
     ],
     "prompt_number": 128
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "__Notes on plots:__\n",
      "\n",
      "* The top plot shows that some spacer-protospacer are not quiet as long as the rest, which is why the 'Number of alignment with nucleotide at position' drops at the (positions 33-35).\n",
      "* There are no mismatches in the middle of the spacer-protospacer alignments."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}