{
 "metadata": {
  "name": "",
  "signature": "sha256:2412d81cc466ee483b43b8f5c7de3d145954d93f124b3f2c4d1e5ddecfda6e48"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "__This notebook describes the setup of CLdb with a set of E. coli genomes.__\n"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# This tutorial assumes that this IPython notebook is in the ./CLdb/examples/ecoli_ipynb/ directory.\n",
      "## these are the relative paths to everything we will be working with\n",
      "rawDir = \"../ecoli_raw/\"\n",
      "CLdbDir = \"../ecoli_CLdb/\"\n",
      "scriptDir = \"../../bin/\""
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 138
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from IPython.display import FileLinks\n",
      "import pandas as pd"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 139
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "__The required files are in '../ecoli_raw/':__\n",
      "\n",
      "* a loci table\n",
      "* array files\n",
      "* genome nucleotide sequences\n",
      " * genbank (preferred) or fasta format\n",
      "\n",
      "__Let's look at the provided files for this example:__"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "FileLinks(rawDir)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "../ecoli_raw/<br>\n",
        "&nbsp;&nbsp;<a href='../ecoli_raw/GIs.txt' target='_blank'>GIs.txt</a><br>\n",
        "../ecoli_raw/array/<br>\n",
        "&nbsp;&nbsp;<a href='../ecoli_raw/array/Ecoli_K12_DH10B_a2.txt' target='_blank'>Ecoli_K12_DH10B_a2.txt</a><br>\n",
        "&nbsp;&nbsp;<a href='../ecoli_raw/array/Ecoli_K12_W3110_a1.txt' target='_blank'>Ecoli_K12_W3110_a1.txt</a><br>\n",
        "&nbsp;&nbsp;<a href='../ecoli_raw/array/Ecoli_K12_MG1655_a1.txt' target='_blank'>Ecoli_K12_MG1655_a1.txt</a><br>\n",
        "&nbsp;&nbsp;<a href='../ecoli_raw/array/Ecoli_BL21_DE3_a2.txt' target='_blank'>Ecoli_BL21_DE3_a2.txt</a><br>\n",
        "&nbsp;&nbsp;<a href='../ecoli_raw/array/Ecoli_K12_MG1655_a2.txt' target='_blank'>Ecoli_K12_MG1655_a2.txt</a><br>\n",
        "&nbsp;&nbsp;<a href='../ecoli_raw/array/Ecoli_BL21_DE3_a1.txt' target='_blank'>Ecoli_BL21_DE3_a1.txt</a><br>\n",
        "&nbsp;&nbsp;<a href='../ecoli_raw/array/Ecoli_K12_W3110_a2.txt' target='_blank'>Ecoli_K12_W3110_a2.txt</a><br>\n",
        "&nbsp;&nbsp;<a href='../ecoli_raw/array/Ecoli_K12_DH10B_a1.txt' target='_blank'>Ecoli_K12_DH10B_a1.txt</a><br>\n",
        "&nbsp;&nbsp;<a href='../ecoli_raw/array/Ecoli_0157_H7_a1.txt' target='_blank'>Ecoli_0157_H7_a1.txt</a><br>\n",
        "&nbsp;&nbsp;<a href='../ecoli_raw/array/Ecoli_0157_H7_a2.txt' target='_blank'>Ecoli_0157_H7_a2.txt</a><br>\n",
        "../ecoli_raw/loci/<br>\n",
        "&nbsp;&nbsp;<a href='../ecoli_raw/loci/140419_loci.txt' target='_blank'>140419_loci.txt</a><br>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 140,
       "text": [
        "../ecoli_raw/\n",
        "  GIs.txt\n",
        "../ecoli_raw/array/\n",
        "  Ecoli_K12_DH10B_a2.txt\n",
        "  Ecoli_K12_W3110_a1.txt\n",
        "  Ecoli_K12_MG1655_a1.txt\n",
        "  Ecoli_BL21_DE3_a2.txt\n",
        "  Ecoli_K12_MG1655_a2.txt\n",
        "  Ecoli_BL21_DE3_a1.txt\n",
        "  Ecoli_K12_W3110_a2.txt\n",
        "  Ecoli_K12_DH10B_a1.txt\n",
        "  Ecoli_0157_H7_a1.txt\n",
        "  Ecoli_0157_H7_a2.txt\n",
        "../ecoli_raw/loci/\n",
        "  140419_loci.txt"
       ]
      }
     ],
     "prompt_number": 140
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Setting up the CLdb directory"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$rawDir\" \"$CLdbDir\"\n",
      "\n",
      "# new ecoli_CLdb directory (removing old versions)\n",
      "if [ -d $2 ]; then\n",
      "    rm -rf $2\n",
      "    mkdir $2\n",
      "fi\n",
      "\n",
      "# copying the files in the raw folder over to the newly made working folder\n",
      "find $1 -type d | egrep \"array|loci\" | xargs -I % cp -r % $2\n",
      "find $1 -name \"GIs.txt\" | xargs -I % cp % $2\n",
      "\n",
      "# checking files\n",
      "ls -thlc $2"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "total 12K\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2  222 Jun  7 18:44 GIs.txt\n",
        "drwxrwxr-x 2 nyoungb2 nyoungb2 4.0K Jun  7 18:44 loci\n",
        "drwxrwxr-x 2 nyoungb2 nyoungb2 4.0K Jun  7 18:44 array\n"
       ]
      }
     ],
     "prompt_number": 146
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "__Downloading the genome genbank files. Using the 'GIs.txt' file__"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "cd $1\n",
      "\n",
      "# making directory\n",
      "genbankDir='genbank/'\n",
      "if [ ! -d $genbankDir ]; then\n",
      "    mkdir $genbankDir\n",
      "fi\n",
      "\n",
      "\n",
      "# loading genbanks (uses BioPerl)\n",
      "perl $2'accession-GI2fastaGenome.pl' -format genbank -fork 5 < 'GIs.txt' -prefix $genbankDir\n",
      "\n",
      "# checking files; all file sizes should be > 10 Mb\n",
      "ls -thlc genbank/*gbk"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 13M Jun  7 18:53 genbank/Escherichia_coli_K_2_MG_655.gbk\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 11M Jun  7 18:53 genbank/Escherichia_coli_K_2_W3.gbk\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 10M Jun  7 18:53 genbank/Escherichia_coli_BL2_DE3.gbk\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 13M Jun  7 18:53 genbank/Escherichia_coli_O_57_H7.gbk\n",
        "-rw-rw-r-- 1 nyoungb2 nyoungb2 11M Jun  7 18:53 genbank/Escherichia_coli_K_2_DH_B.gbk\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "Writing files to '/home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb/genbank'\n",
        "Attempting to stream: Escherichia_coli_K_2_DH_B (accession/GI = 170079663)\n",
        "Attempting to stream: Escherichia_coli_K_2_MG_655 (accession/GI = 49175990)\n",
        "Attempting to stream: Escherichia_coli_BL2_DE3 (accession/GI = 387825439)\n",
        "Attempting to stream: Escherichia_coli_K_2_W3 (accession/GI = 388476123)\n",
        "Attempting to stream: Escherichia_coli_O_57_H7 (accession/GI = 16445223)\n",
        "\n",
        "--------------------- WARNING ---------------------\n",
        "MSG: Unrecognized DBSOURCE data: BioProject: PRJNA57831\n",
        "\n",
        "---------------------------------------------------\n",
        "WARNING: nothing streamed for 387825439.Retrying\n",
        "Starting trial: 2\n",
        "WARNING: nothing streamed for 388476123.Retrying\n",
        "Starting trial: 2\n",
        "WARNING: nothing streamed for 49175990.Retrying\n",
        "Starting trial: 2\n"
       ]
      }
     ],
     "prompt_number": 150
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Loading CLdb"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "# making the sqlite database\n",
      "cd $1\n",
      "if [ -e 'CLdb.sqlite' ]; then \n",
      "    rm 'CLdb.sqlite'\n",
      "fi\n",
      "\n",
      "perl $2'CLdb_makeDB.pl'"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "...sqlite3 database tables created\n"
       ]
      }
     ],
     "prompt_number": 151
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "* The next step is loading the loci table.\n",
      " * This table contains the user-provided info on each CRISPR-CAS system in the genomes.\n",
      " * Let's look at the table before loading it in CLdb"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "lociFile = 'loci/140419_loci.txt'\n",
      "pd.set_option('display.max_columns', 30)\n",
      "tbl = pd.read_table(CLdbDir + lociFile)\n",
      "tbl"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>Locus_ID</th>\n",
        "      <th>Taxon_ID</th>\n",
        "      <th>Taxon_Name</th>\n",
        "      <th>Subtype</th>\n",
        "      <th>Scaffold</th>\n",
        "      <th>Locus_Start</th>\n",
        "      <th>Locus_End</th>\n",
        "      <th>CAS_Start</th>\n",
        "      <th>CAS_End</th>\n",
        "      <th>Array_Start</th>\n",
        "      <th>Array_End</th>\n",
        "      <th>CAS_Status</th>\n",
        "      <th>Array_Status</th>\n",
        "      <th>Genbank_File</th>\n",
        "      <th>Fasta_File</th>\n",
        "      <th>Array_File</th>\n",
        "      <th>Scaffold_count</th>\n",
        "      <th>File_Creation_Date</th>\n",
        "      <th>Author</th>\n",
        "      <th>Leader_Start</th>\n",
        "      <th>Leader_End</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>0</th>\n",
        "      <td> A</td>\n",
        "      <td>  49175990</td>\n",
        "      <td> Escherichia_coli_K-12_MG1655</td>\n",
        "      <td> I-E</td>\n",
        "      <td> NC_000913</td>\n",
        "      <td> 2875723</td>\n",
        "      <td> 2876485</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 2875723</td>\n",
        "      <td> 2876485</td>\n",
        "      <td>NaN</td>\n",
        "      <td> intact</td>\n",
        "      <td> Escherichia_coli_K_2_MG_655.gbk</td>\n",
        "      <td>NaN</td>\n",
        "      <td> Ecoli_K12_MG1655_a1.txt</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 4/19/14</td>\n",
        "      <td> Nick</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>1</th>\n",
        "      <td> B</td>\n",
        "      <td>  49175990</td>\n",
        "      <td> Escherichia_coli_K-12_MG1655</td>\n",
        "      <td> I-E</td>\n",
        "      <td> NC_000913</td>\n",
        "      <td> 2902036</td>\n",
        "      <td> 2902429</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 2902036</td>\n",
        "      <td> 2902429</td>\n",
        "      <td>NaN</td>\n",
        "      <td> intact</td>\n",
        "      <td> Escherichia_coli_K_2_MG_655.gbk</td>\n",
        "      <td>NaN</td>\n",
        "      <td> Ecoli_K12_MG1655_a2.txt</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 4/19/14</td>\n",
        "      <td> Nick</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>2</th>\n",
        "      <td> C</td>\n",
        "      <td> 170079663</td>\n",
        "      <td>  Escherichia_coli_K-12_DH10B</td>\n",
        "      <td> I-E</td>\n",
        "      <td> NC_010473</td>\n",
        "      <td> 2968265</td>\n",
        "      <td> 2969027</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 2968265</td>\n",
        "      <td> 2969027</td>\n",
        "      <td>NaN</td>\n",
        "      <td> intact</td>\n",
        "      <td>   Escherichia_coli_K_2_DH_B.gbk</td>\n",
        "      <td>NaN</td>\n",
        "      <td>  Ecoli_K12_DH10B_a1.txt</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 4/19/14</td>\n",
        "      <td> Nick</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>3</th>\n",
        "      <td> D</td>\n",
        "      <td> 170079663</td>\n",
        "      <td>  Escherichia_coli_K-12_DH10B</td>\n",
        "      <td> I-E</td>\n",
        "      <td> NC_010473</td>\n",
        "      <td> 2994578</td>\n",
        "      <td> 2994971</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 2994578</td>\n",
        "      <td> 2994971</td>\n",
        "      <td>NaN</td>\n",
        "      <td> intact</td>\n",
        "      <td>   Escherichia_coli_K_2_DH_B.gbk</td>\n",
        "      <td>NaN</td>\n",
        "      <td>  Ecoli_K12_DH10B_a2.txt</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 4/19/14</td>\n",
        "      <td> Nick</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>4</th>\n",
        "      <td> E</td>\n",
        "      <td> 388476123</td>\n",
        "      <td>  Escherichia_coli_K-12_W3110</td>\n",
        "      <td> I-E</td>\n",
        "      <td> NC_007779</td>\n",
        "      <td> 2876357</td>\n",
        "      <td> 2877119</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 2876357</td>\n",
        "      <td> 2877119</td>\n",
        "      <td>NaN</td>\n",
        "      <td> intact</td>\n",
        "      <td>     Escherichia_coli_K_2_W3.gbk</td>\n",
        "      <td>NaN</td>\n",
        "      <td>  Ecoli_K12_W3110_a1.txt</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 4/19/14</td>\n",
        "      <td> Nick</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>5</th>\n",
        "      <td> F</td>\n",
        "      <td> 388476123</td>\n",
        "      <td>  Escherichia_coli_K-12_W3110</td>\n",
        "      <td> I-E</td>\n",
        "      <td> NC_007779</td>\n",
        "      <td> 2902670</td>\n",
        "      <td> 2903063</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 2902670</td>\n",
        "      <td> 2903063</td>\n",
        "      <td>NaN</td>\n",
        "      <td> intact</td>\n",
        "      <td>     Escherichia_coli_K_2_W3.gbk</td>\n",
        "      <td>NaN</td>\n",
        "      <td>  Ecoli_K12_W3110_a2.txt</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 4/19/14</td>\n",
        "      <td> Nick</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>6</th>\n",
        "      <td> G</td>\n",
        "      <td>  16445223</td>\n",
        "      <td>     Escherichia_coli_O157_H7</td>\n",
        "      <td> I-E</td>\n",
        "      <td> NC_002655</td>\n",
        "      <td> 3665521</td>\n",
        "      <td> 3665732</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 3665521</td>\n",
        "      <td> 3665732</td>\n",
        "      <td>NaN</td>\n",
        "      <td> intact</td>\n",
        "      <td>    Escherichia_coli_O_57_H7.gbk</td>\n",
        "      <td>NaN</td>\n",
        "      <td>    Ecoli_0157_H7_a1.txt</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 4/19/14</td>\n",
        "      <td> Nick</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>7</th>\n",
        "      <td> H</td>\n",
        "      <td>  16445223</td>\n",
        "      <td>     Escherichia_coli_O157_H7</td>\n",
        "      <td> I-E</td>\n",
        "      <td> NC_002655</td>\n",
        "      <td> 3691424</td>\n",
        "      <td> 3691513</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 3691424</td>\n",
        "      <td> 3691513</td>\n",
        "      <td>NaN</td>\n",
        "      <td> intact</td>\n",
        "      <td>    Escherichia_coli_O_57_H7.gbk</td>\n",
        "      <td>NaN</td>\n",
        "      <td>    Ecoli_0157_H7_a2.txt</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 4/19/14</td>\n",
        "      <td> Nick</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>8</th>\n",
        "      <td> I</td>\n",
        "      <td> 387825439</td>\n",
        "      <td>    Escherichia_coli_BL21_DE3</td>\n",
        "      <td> I-E</td>\n",
        "      <td> NC_012971</td>\n",
        "      <td> 2717668</td>\n",
        "      <td> 2717940</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 2717668</td>\n",
        "      <td> 2717940</td>\n",
        "      <td>NaN</td>\n",
        "      <td> intact</td>\n",
        "      <td>    Escherichia_coli_BL2_DE3.gbk</td>\n",
        "      <td>NaN</td>\n",
        "      <td>   Ecoli_BL21_DE3_a1.txt</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 4/19/14</td>\n",
        "      <td> Nick</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>9</th>\n",
        "      <td> J</td>\n",
        "      <td> 387825439</td>\n",
        "      <td>    Escherichia_coli_BL21_DE3</td>\n",
        "      <td> I-E</td>\n",
        "      <td> NC_012971</td>\n",
        "      <td> 2736174</td>\n",
        "      <td> 2736996</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 2736174</td>\n",
        "      <td> 2736996</td>\n",
        "      <td>NaN</td>\n",
        "      <td> intact</td>\n",
        "      <td>    Escherichia_coli_BL2_DE3.gbk</td>\n",
        "      <td>NaN</td>\n",
        "      <td>   Ecoli_BL21_DE3_a2.txt</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 4/19/14</td>\n",
        "      <td> Nick</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 152,
       "text": [
        "  Locus_ID   Taxon_ID                    Taxon_Name Subtype   Scaffold  \\\n",
        "0        A   49175990  Escherichia_coli_K-12_MG1655     I-E  NC_000913   \n",
        "1        B   49175990  Escherichia_coli_K-12_MG1655     I-E  NC_000913   \n",
        "2        C  170079663   Escherichia_coli_K-12_DH10B     I-E  NC_010473   \n",
        "3        D  170079663   Escherichia_coli_K-12_DH10B     I-E  NC_010473   \n",
        "4        E  388476123   Escherichia_coli_K-12_W3110     I-E  NC_007779   \n",
        "5        F  388476123   Escherichia_coli_K-12_W3110     I-E  NC_007779   \n",
        "6        G   16445223      Escherichia_coli_O157_H7     I-E  NC_002655   \n",
        "7        H   16445223      Escherichia_coli_O157_H7     I-E  NC_002655   \n",
        "8        I  387825439     Escherichia_coli_BL21_DE3     I-E  NC_012971   \n",
        "9        J  387825439     Escherichia_coli_BL21_DE3     I-E  NC_012971   \n",
        "\n",
        "   Locus_Start  Locus_End  CAS_Start  CAS_End  Array_Start  Array_End  \\\n",
        "0      2875723    2876485        NaN      NaN      2875723    2876485   \n",
        "1      2902036    2902429        NaN      NaN      2902036    2902429   \n",
        "2      2968265    2969027        NaN      NaN      2968265    2969027   \n",
        "3      2994578    2994971        NaN      NaN      2994578    2994971   \n",
        "4      2876357    2877119        NaN      NaN      2876357    2877119   \n",
        "5      2902670    2903063        NaN      NaN      2902670    2903063   \n",
        "6      3665521    3665732        NaN      NaN      3665521    3665732   \n",
        "7      3691424    3691513        NaN      NaN      3691424    3691513   \n",
        "8      2717668    2717940        NaN      NaN      2717668    2717940   \n",
        "9      2736174    2736996        NaN      NaN      2736174    2736996   \n",
        "\n",
        "   CAS_Status Array_Status                     Genbank_File  Fasta_File  \\\n",
        "0         NaN       intact  Escherichia_coli_K_2_MG_655.gbk         NaN   \n",
        "1         NaN       intact  Escherichia_coli_K_2_MG_655.gbk         NaN   \n",
        "2         NaN       intact    Escherichia_coli_K_2_DH_B.gbk         NaN   \n",
        "3         NaN       intact    Escherichia_coli_K_2_DH_B.gbk         NaN   \n",
        "4         NaN       intact      Escherichia_coli_K_2_W3.gbk         NaN   \n",
        "5         NaN       intact      Escherichia_coli_K_2_W3.gbk         NaN   \n",
        "6         NaN       intact     Escherichia_coli_O_57_H7.gbk         NaN   \n",
        "7         NaN       intact     Escherichia_coli_O_57_H7.gbk         NaN   \n",
        "8         NaN       intact     Escherichia_coli_BL2_DE3.gbk         NaN   \n",
        "9         NaN       intact     Escherichia_coli_BL2_DE3.gbk         NaN   \n",
        "\n",
        "                Array_File  Scaffold_count File_Creation_Date Author  \\\n",
        "0  Ecoli_K12_MG1655_a1.txt             NaN            4/19/14   Nick   \n",
        "1  Ecoli_K12_MG1655_a2.txt             NaN            4/19/14   Nick   \n",
        "2   Ecoli_K12_DH10B_a1.txt             NaN            4/19/14   Nick   \n",
        "3   Ecoli_K12_DH10B_a2.txt             NaN            4/19/14   Nick   \n",
        "4   Ecoli_K12_W3110_a1.txt             NaN            4/19/14   Nick   \n",
        "5   Ecoli_K12_W3110_a2.txt             NaN            4/19/14   Nick   \n",
        "6     Ecoli_0157_H7_a1.txt             NaN            4/19/14   Nick   \n",
        "7     Ecoli_0157_H7_a2.txt             NaN            4/19/14   Nick   \n",
        "8    Ecoli_BL21_DE3_a1.txt             NaN            4/19/14   Nick   \n",
        "9    Ecoli_BL21_DE3_a2.txt             NaN            4/19/14   Nick   \n",
        "\n",
        "   Leader_Start  Leader_End  \n",
        "0           NaN         NaN  \n",
        "1           NaN         NaN  \n",
        "2           NaN         NaN  \n",
        "3           NaN         NaN  \n",
        "4           NaN         NaN  \n",
        "5           NaN         NaN  \n",
        "6           NaN         NaN  \n",
        "7           NaN         NaN  \n",
        "8           NaN         NaN  \n",
        "9           NaN         NaN  "
       ]
      }
     ],
     "prompt_number": 152
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "* __As you can see, not all of the fields have values. Some are not required (e.g., 'fasta_file')__\n",
      "* __You will get an error if you try to load a table with missing values in required fields.__"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "# Loading the loci table\n",
      "## This will involve a lot of checking files and loading data\n",
      "## The genbanks will be checked for the genome nucleotide sequence.\n",
      "### This is because the fasta files \n",
      "### If found fasta files of each genome will be created in the ./fasta/ directory\n",
      "\n",
      "cd $1\n",
      "perl $2'CLdb_loadLoci.pl' -da CLdb.sqlite < $1'loci/140419_loci.txt'"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "### checking line breaks for all external files (converting to unix) ###\n",
        " processing: /home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb/genbank/Escherichia_coli_K_2_W3.gbk\n",
        " processing: /home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb/array/Ecoli_K12_W3110_a2.txt\n",
        " processing: /home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb/genbank/Escherichia_coli_K_2_MG_655.gbk\n",
        " processing: /home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb/array/Ecoli_K12_MG1655_a1.txt\n",
        " processing: /home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb/genbank/Escherichia_coli_BL2_DE3.gbk\n",
        " processing: /home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb/array/Ecoli_BL21_DE3_a2.txt\n",
        " processing: /home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb/genbank/Escherichia_coli_K_2_W3.gbk\n",
        " processing: /home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb/array/Ecoli_K12_W3110_a1.txt\n",
        " processing: /home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb/genbank/Escherichia_coli_K_2_MG_655.gbk\n",
        " processing: /home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb/array/Ecoli_K12_MG1655_a2.txt\n",
        " processing: /home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb/genbank/Escherichia_coli_O_57_H7.gbk\n",
        " processing: /home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb/array/Ecoli_0157_H7_a2.txt\n",
        " processing: /home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb/genbank/Escherichia_coli_K_2_DH_B.gbk\n",
        " processing: /home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb/array/Ecoli_K12_DH10B_a1.txt\n",
        " processing: /home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb/genbank/Escherichia_coli_K_2_DH_B.gbk\n",
        " processing: /home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb/array/Ecoli_K12_DH10B_a2.txt\n",
        " processing: /home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb/genbank/Escherichia_coli_BL2_DE3.gbk\n",
        " processing: /home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb/array/Ecoli_BL21_DE3_a1.txt\n",
        " processing: /home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb/genbank/Escherichia_coli_O_57_H7.gbk\n",
        " processing: /home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb/array/Ecoli_0157_H7_a1.txt\n",
        "### checking locus_ID values ###\n",
        "...locus_ID values are OK\n",
        "### Checking for existence of genome fasta ###\n",
        "### Processing locus: \"F\" ###\n",
        "  No genome fasta for locus_id: \"F\"! Trying to extract sequence from genbank...\n",
        "   Success! fasta file extracted from Escherichia_coli_K_2_W3.gbk.\n",
        "   File written to: /home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb//fasta//Escherichia_coli_K_2_W3.fasta\n",
        "### Processing locus: \"A\" ###\n",
        "  No genome fasta for locus_id: \"A\"! Trying to extract sequence from genbank...\n",
        "   Success! fasta file extracted from Escherichia_coli_K_2_MG_655.gbk.\n",
        "   File written to: /home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb//fasta//Escherichia_coli_K_2_MG_655.fasta\n",
        "### Processing locus: \"J\" ###\n",
        "  No genome fasta for locus_id: \"J\"! Trying to extract sequence from genbank...\n",
        "   Success! fasta file extracted from Escherichia_coli_BL2_DE3.gbk.\n",
        "   File written to: /home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb//fasta//Escherichia_coli_BL2_DE3.fasta\n",
        "### Processing locus: \"E\" ###\n",
        "  No genome fasta for locus_id: \"E\"! Trying to extract sequence from genbank...\n",
        "  Success! Found fasta in '/home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb//fasta//Escherichia_coli_K_2_W3.fasta'\n",
        "  Adding fasta to loci table\n",
        "### Processing locus: \"B\" ###\n",
        "  No genome fasta for locus_id: \"B\"! Trying to extract sequence from genbank...\n",
        "  Success! Found fasta in '/home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb//fasta//Escherichia_coli_K_2_MG_655.fasta'\n",
        "  Adding fasta to loci table\n",
        "### Processing locus: \"H\" ###\n",
        "  No genome fasta for locus_id: \"H\"! Trying to extract sequence from genbank...\n",
        "   Success! fasta file extracted from Escherichia_coli_O_57_H7.gbk.\n",
        "   File written to: /home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb//fasta//Escherichia_coli_O_57_H7.fasta\n",
        "### Processing locus: \"C\" ###\n",
        "  No genome fasta for locus_id: \"C\"! Trying to extract sequence from genbank...\n",
        "   Success! fasta file extracted from Escherichia_coli_K_2_DH_B.gbk.\n",
        "   File written to: /home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb//fasta//Escherichia_coli_K_2_DH_B.fasta\n",
        "### Processing locus: \"D\" ###\n",
        "  No genome fasta for locus_id: \"D\"! Trying to extract sequence from genbank...\n",
        "  Success! Found fasta in '/home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb//fasta//Escherichia_coli_K_2_DH_B.fasta'\n",
        "  Adding fasta to loci table\n",
        "### Processing locus: \"I\" ###\n",
        "  No genome fasta for locus_id: \"I\"! Trying to extract sequence from genbank...\n",
        "  Success! Found fasta in '/home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb//fasta//Escherichia_coli_BL2_DE3.fasta'\n",
        "  Adding fasta to loci table\n",
        "### Processing locus: \"G\" ###\n",
        "  No genome fasta for locus_id: \"G\"! Trying to extract sequence from genbank...\n",
        "  Success! Found fasta in '/home/nyoungb2/perl/projects/CLdb/examples/ecoli_CLdb//fasta//Escherichia_coli_O_57_H7.fasta'\n",
        "  Adding fasta to loci table\n",
        "### checking for genome fasta for scaffold names ###\n",
        "\n",
        "### loading entries into CLdb ###\n",
        "...Number of entries added/updated to 'loci' table: 10\n",
        "### Leader start-end provided. Loading values into leader table ###\n",
        "...Number of entries added/updated to 'leaders' table: 0\n"
       ]
      }
     ],
     "prompt_number": 153
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "! sqlite3 $CLdbDir'CLdb.sqlite' \"select distinct(fasta_file) from loci\""
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Escherichia_coli_K_2_W3.fasta\r\n",
        "Escherichia_coli_K_2_MG_655.fasta\r\n",
        "Escherichia_coli_BL2_DE3.fasta\r\n",
        "Escherichia_coli_O_57_H7.fasta\r\n",
        "Escherichia_coli_K_2_DH_B.fasta\r\n"
       ]
      }
     ],
     "prompt_number": 162
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "* __The next step is to load the array tables.__\n",
      "* __These are tables in 'CRISPRFinder format' that have CRISPR array info.__\n",
      "* __Let's take a look at one array file.__"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "arrayFile = 'array/Ecoli_0157_H7_a1.txt'\n",
      "pd.set_option('display.max_columns', 30)\n",
      "tbl = pd.read_table(CLdbDir + arrayFile)\n",
      "tbl"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>3665521</th>\n",
        "      <th>CGGTTTATCCCCGCTGATGCGGGGAACAC</th>\n",
        "      <th>AGCGGCACGCTGGATTGAACAAATCCCTGGGC </th>\n",
        "      <th>3665581</th>\n",
        "      <th>Unnamed: 4</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>0</th>\n",
        "      <td> 3665582</td>\n",
        "      <td> CGGTTTATCCCCGCTGGCGCGGGGAACAC</td>\n",
        "      <td> AAACCGAAACACACGATCAATCCGAATATGAG </td>\n",
        "      <td> 3665642</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>1</th>\n",
        "      <td> 3665643</td>\n",
        "      <td> CGGTTTATCCCCGCTGGCGCGGGGAACAC</td>\n",
        "      <td> TTTGGTGACAGTTTTTGTCACTGTTTTGGTGA </td>\n",
        "      <td> 3665703</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>2</th>\n",
        "      <td> 3665704</td>\n",
        "      <td> CGGTTTATCCCCGCTGGCGCGGGGAACAC</td>\n",
        "      <td>                               NaN</td>\n",
        "      <td> 3665732</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 157,
       "text": [
        "   3665521  CGGTTTATCCCCGCTGATGCGGGGAACAC  AGCGGCACGCTGGATTGAACAAATCCCTGGGC   \\\n",
        "0  3665582  CGGTTTATCCCCGCTGGCGCGGGGAACAC  AAACCGAAACACACGATCAATCCGAATATGAG    \n",
        "1  3665643  CGGTTTATCCCCGCTGGCGCGGGGAACAC  TTTGGTGACAGTTTTTGTCACTGTTTTGGTGA    \n",
        "2  3665704  CGGTTTATCCCCGCTGGCGCGGGGAACAC                                NaN   \n",
        "\n",
        "   3665581  Unnamed: 4  \n",
        "0  3665642         NaN  \n",
        "1  3665703         NaN  \n",
        "2  3665732         NaN  "
       ]
      }
     ],
     "prompt_number": 157
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "* __The array file consists of 4 columns:__\n",
      " * spacer start\n",
      " * spacer sequence\n",
      " * direct-repeat sequence\n",
      " * direct-repeat stop\n",
      " * all extra columns ignored"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "# now, load the array tables in to CLdb\n",
      "\n",
      "cd $1\n",
      "perl $2'CLdb_loadArrays.pl' -da CLdb.sqlite "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "...Number of entries added/updated to 'DRs' table: 85\n",
        "...Number of entries added/updated to 'spacers' table: 75\n"
       ]
      }
     ],
     "prompt_number": 158
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "# Next, loading the CAS genes\n",
      "## Technically, all coding seuqences in the region specified in the loci table (CAS_start, CAS_end) will be loaded.\n",
      "\n",
      "#--- skipping for now, CAS genes not identified ---#\n",
      "\n",
      "#cd $1\n",
      "#perl $1'CLdb_getGenesInLoci.pl' -da CLdb.sqlite |\\\n",
      "#perl $1'CLdb_loadGenes.pl' -da CLdb.sqlite"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 159
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### Setting array sense strand\n",
      "\n",
      "* __The strand that is transcribed needs to be defined in order to have the correct sequence for downstream analyses (e.g., blasting spacers and getting PAM regions)__\n",
      "* __The sense (reading) strand is defined by:__\n",
      " * The leader region (if defined; in this case, no).\n",
      " * Array_start * Array_end in the loci table\n",
      "  * The genome negative strand will be used if array_start > array_end"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "# clustering is performed with cd-hit-est\n",
      "## I'm looking into other methods such as usearch\n",
      "\n",
      "cd $1\n",
      "perl $2'CLdb_setSenseStrand.pl' -da CLdb.sqlite"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "Setting sense strand for 10 loci...\n",
        "...sense strand set based on array_start-array_end for:\t10 loci\n",
        "...sense strand set base on leader region for:\t0 loci\n"
       ]
      }
     ],
     "prompt_number": 160
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Spacer and DR clustering\n",
      "\n",
      "* __Clustering of spacer and/or DR sequences accomplishes:__\n",
      " * A method of comparing within and between CRISPRs\n",
      " * A reducing redundancy for spacer and DR blasting"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%bash -s \"$CLdbDir\" \"$scriptDir\"\n",
      "\n",
      "# clustering is performed with cd-hit-est\n",
      "## I'm looking into other methods such as usearch\n",
      "\n",
      "cd $1\n",
      "perl $2'CLdb_clusterArrayElements.pl' -da CLdb.sqlite -s -r"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "Deleted all entries in 'spacer_clusters' table\n",
        "Deleted all entries in 'DR_clusters' table\n",
        "Getting array element sequences. Orienting sequences by array_sense_strand\n",
        "\n",
        "Clustering spacer sequences...\n",
        "...Clustering spacers at cutoff: 0.80\n",
        "\tNumber of clusters produced: 39\n",
        "...Clustering spacers at cutoff: 0.81\n",
        "\tNumber of clusters produced: 39\n",
        "...Clustering spacers at cutoff: 0.82\n",
        "\tNumber of clusters produced: 39\n",
        "...Clustering spacers at cutoff: 0.83\n",
        "\tNumber of clusters produced: 39\n",
        "...Clustering spacers at cutoff: 0.84\n",
        "\tNumber of clusters produced: 39\n",
        "...Clustering spacers at cutoff: 0.85\n",
        "\tNumber of clusters produced: 39\n",
        "...Clustering spacers at cutoff: 0.86\n",
        "\tNumber of clusters produced: 39\n",
        "...Clustering spacers at cutoff: 0.87\n",
        "\tNumber of clusters produced: 39\n",
        "...Clustering spacers at cutoff: 0.88\n",
        "\tNumber of clusters produced: 39\n",
        "...Clustering spacers at cutoff: 0.89\n",
        "\tNumber of clusters produced: 39\n",
        "...Clustering spacers at cutoff: 0.90\n",
        "\tNumber of clusters produced: 39\n",
        "...Clustering spacers at cutoff: 0.91\n",
        "\tNumber of clusters produced: 39\n",
        "...Clustering spacers at cutoff: 0.92\n",
        "\tNumber of clusters produced: 39\n",
        "...Clustering spacers at cutoff: 0.93\n",
        "\tNumber of clusters produced: 39\n",
        "...Clustering spacers at cutoff: 0.94\n",
        "\tNumber of clusters produced: 39\n",
        "...Clustering spacers at cutoff: 0.95\n",
        "\tNumber of clusters produced: 39\n",
        "...Clustering spacers at cutoff: 0.96\n",
        "\tNumber of clusters produced: 39\n",
        "...Clustering spacers at cutoff: 0.97\n",
        "\tNumber of clusters produced: 39\n",
        "...Clustering spacers at cutoff: 0.98\n",
        "\tNumber of clusters produced: 39\n",
        "...Clustering spacers at cutoff: 0.99\n",
        "\tNumber of clusters produced: 39\n",
        "...Clustering spacers at cutoff: 1.00\n",
        "\tNumber of clusters produced: 39\n",
        "\n",
        "Clustering DR sequences...\n",
        "...Clustering DRs at cutoff: 0.80\n",
        "\tNumber of clusters produced: 2\n",
        "...Clustering DRs at cutoff: 0.81\n",
        "\tNumber of clusters produced: 2\n",
        "...Clustering DRs at cutoff: 0.82\n",
        "\tNumber of clusters produced: 2\n",
        "...Clustering DRs at cutoff: 0.83\n",
        "\tNumber of clusters produced: 2\n",
        "...Clustering DRs at cutoff: 0.84\n",
        "\tNumber of clusters produced: 2\n",
        "...Clustering DRs at cutoff: 0.85\n",
        "\tNumber of clusters produced: 2\n",
        "...Clustering DRs at cutoff: 0.86\n",
        "\tNumber of clusters produced: 2\n",
        "...Clustering DRs at cutoff: 0.87\n",
        "\tNumber of clusters produced: 2\n",
        "...Clustering DRs at cutoff: 0.88\n",
        "\tNumber of clusters produced: 2\n",
        "...Clustering DRs at cutoff: 0.89\n",
        "\tNumber of clusters produced: 2\n",
        "...Clustering DRs at cutoff: 0.90\n",
        "\tNumber of clusters produced: 4\n",
        "...Clustering DRs at cutoff: 0.91\n",
        "\tNumber of clusters produced: 4\n",
        "...Clustering DRs at cutoff: 0.92\n",
        "\tNumber of clusters produced: 4\n",
        "...Clustering DRs at cutoff: 0.93\n",
        "\tNumber of clusters produced: 4\n",
        "...Clustering DRs at cutoff: 0.94\n",
        "\tNumber of clusters produced: 8\n",
        "...Clustering DRs at cutoff: 0.95\n",
        "\tNumber of clusters produced: 8\n",
        "...Clustering DRs at cutoff: 0.96\n",
        "\tNumber of clusters produced: 8\n",
        "...Clustering DRs at cutoff: 0.97\n",
        "\tNumber of clusters produced: 12\n",
        "...Clustering DRs at cutoff: 0.98\n",
        "\tNumber of clusters produced: 12\n",
        "...Clustering DRs at cutoff: 0.99\n",
        "\tNumber of clusters produced: 12\n",
        "...Clustering DRs at cutoff: 1.00\n",
        "\tNumber of clusters produced: 12\n",
        "\n",
        "Inserting/updating entries in CLdb...\n",
        "...1575 spacer cluster entries added/updated\n",
        "...1785 DR cluster entries added/updated\n",
        "\n",
        "Creating indices for cluster ids...\n"
       ]
      }
     ],
     "prompt_number": 161
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "# Next Steps\n",
      "\n",
      "### [arrayBlast](./arrayBlast.ipynb)\n",
      "\n",
      "Blast spacers (& DRs), get protospacers, PAM regions, mismatches to the protospacer & SEED sequence"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 106
    }
   ],
   "metadata": {}
  }
 ]
}