{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Description:\n",
    "\n",
    "* This notebook goes through the analysis of blasting spacers vs a database containing potential protospacers.\n",
    "* Any blast datbase can be used.\n",
    "* A script is provided to easily blast all of the genomes in CLdb."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Before running this notebook:\n",
    "\n",
    "* __run the [Setup](./Setup.pynb) notebook__"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# User-defined variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# directory where you want the spacer blasting to be done\n",
    "## CHANGE THIS!\n",
    "workDir = \"/home/nyoungb2/t/CLdb_Ecoli/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Init"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The rpy2.ipython extension is already loaded. To reload it, use:\n",
      "  %reload_ext rpy2.ipython\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "from IPython.display import FileLinks\n",
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "blastDir = os.path.join(workDir, 'arrayBlast')\n",
    "if not os.path.isdir(blastDir):\n",
    "    os.makedirs(blastDir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Blast vs all E. coli genomes in the CLdb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Spacer blast"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Selecting spacer sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique spacers:\t39\n",
      "\n",
      ">NA|spacer|NA|10|1\n",
      "ATAGACCCCGAACAACAATACGCGCAAACCGA\n",
      ">NA|spacer|NA|1|1\n",
      "TGGCTCTGCAACAGCAGCACCCATGACCACGTC\n",
      ">NA|spacer|NA|29|1\n",
      "GACAGAACGGCCTCAGTAGTCTCGTCAGGCTC\n"
     ]
    }
   ],
   "source": [
    "!cd $blastDir; \\\n",
    "    CLdb -- array2fasta -cluster -cut 1 > spacers_cut1.fna\n",
    "\n",
    "## getting number of spacer sequences\n",
    "!printf 'Number of unique spacers:\\t'\n",
    "!cd $blastDir; \\\n",
    "    grep -c \">\" spacers_cut1.fna; \\\n",
    "    echo; \\\n",
    "    head -n 6 spacers_cut1.fna"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Blast\n",
    "\n",
    "* The array blast commands are sub-sub commands.\n",
    "* __SO__, that's why when you type: '`CLdb -- arrayBlast -h`', you get then following:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage:\r\n",
      "    arrayBlast [options] subcommand -- [subcommand_options]\r\n",
      "\r\n",
      "  Options:\r\n",
      "    --list\r\n",
      "        List all subcommands.\r\n",
      "\r\n",
      "    --perldoc\r\n",
      "        Get perldoc of subcommand.\r\n",
      "\r\n",
      "    -v Verbose output\r\n",
      "    -h This help message\r\n",
      "\r\n",
      "  For more information:\r\n",
      "    CLdb_perldoc arrayBlast\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!CLdb -- arrayBlast -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "add_crRNA\r\n",
      "add_proto\r\n",
      "align_proto\r\n",
      "filter_arrayHits\r\n",
      "get_PAM\r\n",
      "get_SEEDstats\r\n",
      "get_align\r\n",
      "get_proto\r\n",
      "run\r\n",
      "srl2csv\r\n",
      "srl2json\r\n",
      "xml2json\r\n",
      "xml2srl\r\n"
     ]
    }
   ],
   "source": [
    "# listing all arrayBlast subcommands\n",
    "!CLdb -- arrayBlast --list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Sub-sub-command help:\n",
    "\n",
    "`CLdb arrayBlast -- run -- -h`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Making BLAST databases...\n",
      "...making blast databases in /home/nyoungb2/t/CLdb_Ecoli//arrayBlast/blast_db/\n",
      "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/t/CLdb_Ecoli//arrayBlast/blast_db//Escherichia_coli_BL21_DE3.fasta\n",
      "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/t/CLdb_Ecoli//arrayBlast/blast_db//Escherichia_coli_K-12_DH10B.fasta\n",
      "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/t/CLdb_Ecoli//arrayBlast/blast_db//Escherichia_coli_K-12_MG1655.fasta\n",
      "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/t/CLdb_Ecoli//arrayBlast/blast_db//Escherichia_coli_K-12_W3110.fasta\n",
      "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/t/CLdb_Ecoli//arrayBlast/blast_db//Escherichia_coli_O157_H7.fasta\n",
      "\n",
      "Blasting sequences...\n",
      "Number of blast hits to Escherichia_coli_BL21_DE3.fasta:\t1290\n",
      "Number of blast hits to Escherichia_coli_K-12_DH10B.fasta:\t1317\n",
      "Number of blast hits to Escherichia_coli_K-12_MG1655.fasta:\t1317\n",
      "Number of blast hits to Escherichia_coli_K-12_W3110.fasta:\t1317\n",
      "Number of blast hits to Escherichia_coli_O157_H7.fasta:\t912\n"
     ]
    }
   ],
   "source": [
    "# spacer blast\n",
    "## Really just a thin wrapper around blastn that will blast the spacers against all genomes in CLdb\n",
    "### If you want to blast some else, just run the blast yourself (use 'perldoc CLdb_arrayBlast.pl' for more info)\n",
    "\n",
    "!cd $blastDir; \\\n",
    "    CLdb -- arrayBlast -- run \\\n",
    "    -query spacers_cut1.fna \\\n",
    "    > spacers_cut1_blastn.xml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note:**\n",
    "\n",
    "* Using spacer clustering cutoff of 1, so just 'unique' spacer sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-rw-r-- 1 nyoungb2 nyoungb2  62K Dec 31 06:13 spacers_cut1_blastn.srl\r\n",
      "-rw-rw-r-- 1 nyoungb2 nyoungb2 227K Dec 31 06:13 spacers_cut1_blastn.xml\r\n"
     ]
    }
   ],
   "source": [
    "# converting .xml to .srl\n",
    "## This retains all of the info in the xml, but produces a smaller file that is faster to load and write.\n",
    "## For large blast runs (xml files > 100 Mb), this will take a little while\n",
    "\n",
    "!cd $blastDir; \\\n",
    "    CLdb -- arrayBlast -- xml2srl \\\n",
    "    spacers_cut1_blastn.xml \\\n",
    "    > spacers_cut1_blastn.srl\n",
    "\n",
    "# checking output file\n",
    "!cd $blastDir; \\\n",
    "    ls -thlc spacers_cut1_blastn.*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note:**\n",
    "\n",
    "* You could pipe the output from `CLdb -- arrayBlast -- run` directly into `CLdb -- arrayBlast -- xml2srl`\n",
    "  * Example: `CLdb -- arrayBlast -- run -query spacers_cut1.fna | CLdb -- arrayBlast -- xml2srl`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DR Blast\n",
    "\n",
    "* BLASTing DR sequences against the same subjects.\n",
    "* This is needed to filter out spacer blast hits to CRISPR arrays."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique DRs:\t12\n",
      "\n",
      ">NA|DR|NA|9|1\n",
      "TGGTTTATCCCCGCTGGCGCGGGGAACTC\n",
      ">NA|DR|NA|6|1\n",
      "CGGTTTATCCTCGCTGGCGCGGGGAACTC\n",
      ">NA|DR|NA|10|1\n",
      "GGTTTATCCCCGCTGGCGCGGGGAACAC\n"
     ]
    }
   ],
   "source": [
    "# selecting DRs, so we can blast them.\n",
    "## This is needed blast filter out spacers that hit CRISPR arrays\n",
    "\n",
    "!cd $blastDir; \\\n",
    "    CLdb -- array2fasta \\\n",
    "    -cluster -cut 1 -r \\\n",
    "    > DRs_cut1.fna\n",
    "\n",
    "## getting number of spacer sequences\n",
    "!printf 'Number of unique DRs:\\t'\n",
    "!cd $blastDir; \\\n",
    "    grep -c \">\" DRs_cut1.fna; \\\n",
    "    echo; \\\n",
    "    head -n 6 DRs_cut1.fna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Making BLAST databases...\n",
      "...making blast databases in /home/nyoungb2/t/CLdb_Ecoli//arrayBlast/blast_db/\n",
      "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/t/CLdb_Ecoli//arrayBlast/blast_db//Escherichia_coli_BL21_DE3.fasta\n",
      "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/t/CLdb_Ecoli//arrayBlast/blast_db//Escherichia_coli_K-12_DH10B.fasta\n",
      "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/t/CLdb_Ecoli//arrayBlast/blast_db//Escherichia_coli_K-12_MG1655.fasta\n",
      "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/t/CLdb_Ecoli//arrayBlast/blast_db//Escherichia_coli_K-12_W3110.fasta\n",
      "makeblastdb -dbtype nucl -parse_seqids -in /home/nyoungb2/t/CLdb_Ecoli//arrayBlast/blast_db//Escherichia_coli_O157_H7.fasta\n",
      "\n",
      "Blasting sequences...\n",
      "Number of blast hits to Escherichia_coli_BL21_DE3.fasta:\t4730\n",
      "Number of blast hits to Escherichia_coli_K-12_DH10B.fasta:\t4825\n",
      "Number of blast hits to Escherichia_coli_K-12_MG1655.fasta:\t4882\n",
      "Number of blast hits to Escherichia_coli_K-12_W3110.fasta:\t4882\n",
      "Number of blast hits to Escherichia_coli_O157_H7.fasta:\t1747\n"
     ]
    }
   ],
   "source": [
    "# DR blast, just like spacer blast\n",
    "\n",
    "!cd $blastDir; \\\n",
    "    CLdb -- arrayBlast -- run \\\n",
    "    -query DRs_cut1.fna \\\n",
    "    > DRs_cut1_blastn.xml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-rw-r-- 1 nyoungb2 nyoungb2 247K Dec 31 06:14 DRs_cut1_blastn.srl\r\n",
      "-rw-rw-r-- 1 nyoungb2 nyoungb2 783K Dec 31 06:14 DRs_cut1_blastn.xml\r\n"
     ]
    }
   ],
   "source": [
    "# converting .xml to .srl\n",
    "## This retains all of the info in the xml, but produces a smaller file that is faster to load and write.\n",
    "## For large blast runs (xml files > 100 Mb), this will take a little while\n",
    "\n",
    "!cd $blastDir; \\\n",
    "    CLdb -- arrayBlast -- xml2srl \\\n",
    "    DRs_cut1_blastn.xml \\\n",
    "    > DRs_cut1_blastn.srl\n",
    "\n",
    "\n",
    "# checking output file\n",
    "!cd $blastDir; \\\n",
    "    ls -thlc DRs_cut1_blastn.*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filtering out spacer hits to CRISPRs\n",
    "\n",
    "* Based on DR hits falling adjacent to spacer hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "### Summary for filtering of 'bad' DR hits ###\n",
      "Total DR hits: 1014\n",
      "DR hits with E-values < 10: 0\n",
      "DR hits with hit lengths (fraction of total) < 0.66: 0\n",
      "DR hits used for filtering: 1014\n",
      "### DR filtering of spacer blast hits ###\n",
      "Total spacer blast hits: 79\n",
      " NOTE: Deleting the blast hits to arrays\n",
      "Spacer blast hits hitting an array defined just by DR-blast hits: 79\n",
      " NOTE: Deleting the blast hits to arrays\n",
      "Spacer blast hits hitting a mobile genetic element: 0\n",
      "\n",
      "-rw-rw-r-- 1 nyoungb2 nyoungb2 43K Dec 31 06:14 spacers_cut1_blastn_filt.srl\n"
     ]
    }
   ],
   "source": [
    "# filtering spacer hits to arrays\n",
    "## This is based on the premise that DR hits adjacent to spacer hits signify that the spacer is hitting a CRISPR array\n",
    "\n",
    "!cd $blastDir; \\\n",
    "    CLdb -- arrayBlast -- filter_arrayHits \\\n",
    "    spacers_cut1_blastn.srl \\\n",
    "    DRs_cut1_blastn.srl \\\n",
    "    > spacers_cut1_blastn_filt.srl\n",
    "    \n",
    "    \n",
    "# checking output file\n",
    "!cd $blastDir; \\\n",
    "    echo; \\\n",
    "    ls -thlc spacers_cut1_blastn_filt.srl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Converting blast .srl to .csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Decoding the srl file...\n",
      "qseqid\tsseqid\tpident\tlength\tmismatch\tgapopen\tqstart\tqend\tsstart\tsend\tevalue\tbitscore\t\n"
     ]
    }
   ],
   "source": [
    "!cd $blastDir; \\\n",
    "    CLdb -- arrayBlast -- srl2csv \\\n",
    "    spacers_cut1_blastn_filt.srl \\\n",
    "    > spacers_cut1_blastn_filt.txt\n",
    "    \n",
    "# assessing table of blast hits    \n",
    "!cd $blastDir; \\\n",
    "    head spacers_cut1_blastn_filt.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Conclusions\n",
    "\n",
    "* **No spacers hit any protospacers on the Ecoli genomes in the CLdb!**\n",
    "  * This indicates that there are no interegrated mobile genetic elements in the E.coli genomes that can be detected using the CRISPR spacers.\n",
    "  * Maybe we will get some hits when we blast against NCBI's nt database."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Array Blast vs NCBI's nt database\n",
    "\n",
    "* NOTE: you will need the BLAST nt database & the blast+ toolkit\n",
    "  * The arrayBlast wrapper only works with the genomes in CLdb.\n",
    "  * We will need to do this part 'manually'\n",
    "    * The blast output just needs to be in .xml format.\n",
    "    * You will need the BLAST nt database \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#set the number of threads to use for your machine \n",
    "num_threads=20"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Spacer blast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-rw-r-- 1 nyoungb2 nyoungb2 1.6M Dec 31 06:15 spacers_cut1_vNT.xml\r\n"
     ]
    }
   ],
   "source": [
    "## spacer blast with blastn\n",
    "!cd $blastDir; \\\n",
    "    blastn -task blastn-short -db nt -query spacers_cut1.fna \\\n",
    "    -outfmt 5 -evalue 1e-5 -num_threads $num_threads > spacers_cut1_vNT.xml\n",
    "    \n",
    "\n",
    "## checking output\n",
    "!cd $blastDir; \\\n",
    "    ls -thlc spacers_cut1_vNT.xml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DR blast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-rw-r-- 1 nyoungb2 nyoungb2 15M Dec 31 06:16 DRs_cut1_vNT.xml\r\n"
     ]
    }
   ],
   "source": [
    "## spacer blast with blastn\n",
    "!cd $blastDir; \\\n",
    "    blastn -task blastn-short -db nt -query DRs_cut1.fna \\\n",
    "    -outfmt 5 -evalue 1e-5 -num_threads $num_threads > DRs_cut1_vNT.xml\n",
    "    \n",
    "\n",
    "## checking output\n",
    "!cd $blastDir; \\\n",
    "    ls -thlc DRs_cut1_vNT.xml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Converting to .srl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-rw-r-- 1 nyoungb2 nyoungb2 4.8M Dec 31 06:17 DRs_cut1_vNT.srl\r\n",
      "-rw-rw-r-- 1 nyoungb2 nyoungb2 584K Dec 31 06:16 spacers_cut1_vNT.srl\r\n"
     ]
    }
   ],
   "source": [
    "## spacers\n",
    "!cd $blastDir; \\\n",
    "    CLdb -- arrayBlast -- xml2srl \\\n",
    "    spacers_cut1_vNT.xml \\\n",
    "    > spacers_cut1_vNT.srl\n",
    "\n",
    "## DRs\n",
    "!cd $blastDir; \\\n",
    "    CLdb -- arrayBlast -- xml2srl \\\n",
    "    DRs_cut1_vNT.xml \\\n",
    "    > DRs_cut1_vNT.srl\n",
    "\n",
    "\n",
    "## checking output\n",
    "!cd $blastDir; \\\n",
    "    ls -thlc *_vNT.srl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filtering spacer hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "### Summary for filtering of 'bad' DR hits ###\n",
      "Total DR hits: 20159\n",
      "DR hits with E-values < 10: 0\n",
      "DR hits with hit lengths (fraction of total) < 0.66: 0\n",
      "DR hits used for filtering: 20159\n",
      "### DR filtering of spacer blast hits ###\n",
      "Total spacer blast hits: 1573\n",
      " NOTE: Deleting the blast hits to arrays\n",
      "Spacer blast hits hitting an array defined just by DR-blast hits: 863\n",
      " NOTE: Deleting the blast hits to arrays\n",
      "Spacer blast hits hitting a mobile genetic element: 710\n",
      "\n",
      "-rw-rw-r-- 1 nyoungb2 nyoungb2 382K Dec 31 06:17 spacers_cut1_vNT_blastn_filt.srl\n"
     ]
    }
   ],
   "source": [
    "# filtering spacer hits to arrays\n",
    "## This is based on the premise that DR hits adjacent to spacer hits signify that the spacer is hitting a CRISPR array\n",
    "\n",
    "!cd $blastDir; \\\n",
    "    CLdb -- arrayBlast -- filter_arrayHits \\\n",
    "    spacers_cut1_vNT.srl \\\n",
    "    DRs_cut1_vNT.srl \\\n",
    "    > spacers_cut1_vNT_blastn_filt.srl\n",
    "    \n",
    "    \n",
    "## checking output file\n",
    "!cd $blastDir; echo; \\\n",
    "    ls -thlc spacers_cut1_vNT_blastn_filt.srl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Great! It's looks like we have some hits to some putative protospacers!__"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## File conversion\n",
    "\n",
    "* __FYI__, the .srl files can be converted to json or csv format. For example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\r\n",
      "    \"0\": {\r\n",
      "        \"BlastOutput_db\": \"nt\",\r\n",
      "        \"BlastOutput_iterations\": {\r\n",
      "            \"Iteration\": [\r\n",
      "                {\r\n",
      "                    \"Iteration_hits\": {\r\n",
      "                        \"Hit\": [\r\n",
      "                            {\r\n",
      "                                \"Hit_accession\": \"KC765378\",\r\n",
      "                                \"Hit_def\": \"Escherichia coli strain 90-0009 CRISPR2a repeat region\",\r\n",
      "                                \"Hit_hsps\": {\r\n",
      "                                    \"Hsp\": {\r\n",
      "                                        \"2lOuFxXa0uuEg\": {\r\n",
      "                                            \"CLdb_array-hit\": 0,\r\n",
      "                                            \"Hsp_align-len\": \"32\",\r\n",
      "                                            \"Hsp_bit-score\": \"63.9282\",\r\n",
      "                                            \"Hsp_evalue\": \"3.97536e-08\",\r\n",
      "                                            \"Hsp_gaps\": \"0\",\r\n",
      "                                            \"Hsp_hit-frame\": \"-1\",\r\n",
      "                                            \"Hsp_hit-from\": \"652\",\r\n",
      "                                            \"Hsp_hit-to\": \"621\",\r\n",
      "                                            \"Hsp_hseq\": \"ATAGACCCCGAACAACAATACGCGCAAACCGA\",\r\n",
      "                                            \"Hsp_identity\": \"32\",\r\n",
      "                                            \"Hsp_midline\": \"||||||||||||||||||||||||||||||||\",\r\n",
      "                                            \"Hsp_num\": \"1\",\r\n",
      "                                            \"Hsp_positive\": \"32\",\r\n",
      "                                            \"Hsp_qseq\": \"ATAGACCCCGAACAACAATACGCGCAAACCGA\",\r\n",
      "                                            \"Hsp_query-frame\": \"1\",\r\n",
      "                                            \"Hsp_query-from\": \"1\",\r\n",
      "                                            \"Hsp_query-to\": \"32\",\r\n",
      "                                            \"Hsp_score\": \"32\"\r\n",
      "                                        }\r\n",
      "                                    }\r\n",
      "                                },\r\n",
      "                                \"Hit_id\": \"gi|528889517|gb|KC765378.1|\",\r\n",
      "                                \"Hit_len\": \"922\",\r\n",
      "                                \"Hit_num\": \"1\"\r\n",
      "                            },\r\n",
      "                            {\r\n"
     ]
    }
   ],
   "source": [
    "# The srl file can be converted to a json or csv file if you would like to use tools external to CLdb.\n",
    "\n",
    "!cd $blastDir; \\\n",
    "    CLdb -- arrayBlast -- srl2json \\\n",
    "    < spacers_cut1_vNT_blastn_filt.srl \\\n",
    "    > spacers_cut1_vNT_blastn_filt.json\n",
    "\n",
    "\n",
    "## viewing some of the json file\n",
    "!cd $blastDir; \\\n",
    "    cat spacers_cut1_vNT_blastn_filt.json | \\\n",
    "    python -mjson.tool 2>/dev/null | head -n 40"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## adding crRNA info\n",
    "\n",
    "* Adding crRNA region to .srl (actually the DNA template, refering to it as 'crDNA')\n",
    "  * How much of the adjacent DR sequences are included in the crDNA is determined by the user (default: 10bp on either side)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Decoding .srl file...\n",
      "Extracting all queries with blast hits from blast hit file...\n",
      "...Number of total unique query IDs:\t39\n",
      "Getting info from CLdb on each blast query spacer...\n",
      "Getting spacer regions from each genome...\n",
      " Processing genome: Escherichia_coli_K-12_DH10B.fasta...\n",
      " Processing genome: Escherichia_coli_K-12_MG1655.fasta...\n",
      " Processing genome: Escherichia_coli_BL21_DE3.fasta...\n",
      " Processing genome: Escherichia_coli_K-12_W3110.fasta...\n",
      " Processing genome: Escherichia_coli_O157_H7.fasta...\n",
      "\n",
      "-rw-rw-r-- 1 nyoungb2 nyoungb2 406K Dec 31 06:18 spacers_cut1_vNT_blastn_filt_crDNA.srl\n"
     ]
    }
   ],
   "source": [
    "!cd $blastDir; \\\n",
    "    CLdb -- arrayBlast -- add_crRNA \\\n",
    "    < spacers_cut1_vNT_blastn_filt.srl \\\n",
    "    > spacers_cut1_vNT_blastn_filt_crDNA.srl\n",
    "   \n",
    "\n",
    "## checking output\n",
    "!cd $blastDir; echo; \\\n",
    "    ls -thlc spacers_cut1_vNT_blastn_filt_crDNA.srl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Adding protospacer info\n",
    "\n",
    "* Similar to adding crRNA info, but getting information from the blast database(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieving proto info from blast db: 'nt'\n",
      "Retrieving proto sequences from blast db: 'nt'\n",
      "\n",
      "-rw-rw-r-- 1 nyoungb2 nyoungb2 517K Dec 31 06:18 spacers_cut1_vNT_blastn_filt_crDNA_proto.srl\n"
     ]
    }
   ],
   "source": [
    "!cd $blastDir; \\\n",
    "    CLdb -- arrayBlast -- add_proto -workers 24 \\\n",
    "    < spacers_cut1_vNT_blastn_filt_crDNA.srl \\\n",
    "    > spacers_cut1_vNT_blastn_filt_crDNA_proto.srl   \n",
    "    \n",
    "## checking output\n",
    "!cd $blastDir; echo; \\\n",
    "    ls -thlc spacers_cut1_vNT_blastn_filt_crDNA_proto.srl   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Aligning crRNA & protospacer\n",
    "\n",
    "* Alignment added to the .srl file.\n",
    "* These are individual alignments between the spacer and protospacer (using clustalw)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Aligning protospacers from blast db: 'nt'\n",
      "\n",
      "-rw-rw-r-- 1 nyoungb2 nyoungb2 710K Dec 31 06:21 spacers_cut1_vNT_blastn_filt_crDNA_proto_aln.srl\n"
     ]
    }
   ],
   "source": [
    "# aligning crRNA and protospacer\n",
    "\n",
    "!cd $blastDir; \\\n",
    "    CLdb -- arrayBlast -- align_proto \\\n",
    "    < spacers_cut1_vNT_blastn_filt_crDNA_proto.srl \\\n",
    "    > spacers_cut1_vNT_blastn_filt_crDNA_proto_aln.srl\n",
    "    \n",
    "## checking output\n",
    "!cd $blastDir; echo; \\\n",
    "    ls -thlc spacers_cut1_vNT_blastn_filt_crDNA_proto_aln.srl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting alignments (fasta)\n",
    "\n",
    "* You can parse out the alignments you want (parse by subtype, taxon_name, etc.)\n",
    "  * **OR** you can parse them after writing (add subtype, taxon_name, e-value, etc. to the sequence name)\n",
    "* The alignments will be oriented to the crDNA (5'-3' for the crRNA)\n",
    "* See `CLdb -- arrayBlast --perldoc -- get_align` for info on what `-outfmt` does"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Decoding .srl file...\n",
      "Getting alignments...\n",
      "\n",
      "-rw-rw-r-- 1 nyoungb2 nyoungb2 285K Dec 31 06:23 vNT_crDNA-proto_aln.fna\n"
     ]
    }
   ],
   "source": [
    "# getting alignment of protospacer and crDNA and writing as a fasta\n",
    "\n",
    "!cd $blastDir; \\\n",
    "    CLdb -- arrayBlast -- get_align \\\n",
    "    -outfmt taxon_name,subtype,evalue \\\n",
    "    < spacers_cut1_vNT_blastn_filt_crDNA_proto_aln.srl \\\n",
    "    > vNT_crDNA-proto_aln.fna\n",
    "    \n",
    "\n",
    "## checking output\n",
    "!cd $blastDir; echo; \\\n",
    "    ls -thlc vNT_crDNA-proto_aln.fna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">crDNA|J|3|2lOuFxXa0uuEg|3.97536e-08\r\n",
      "gggataaaccGTCGGTTTGCGCGTATTGTTGTTCGGGGTCTATgtgttccccg\r\n",
      ">proto|J|3|2lOuFxXa0uuEg|3.97536e-08\r\n",
      ".ggataaaccgTCGGTTTGCGCGTATTGTTGTTCGGGGTCTATgtgttccccg\r\n",
      ">crDNA|J|3|2pFfHSafceYTm|3.97536e-08\r\n",
      "gggataaaccGTCGGTTTGCGCGTATTGTTGTTCGGGGTCTATgtgttccccg\r\n",
      ">proto|J|3|2pFfHSafceYTm|3.97536e-08\r\n",
      ".ggataaaccgTCGGTTTGCGCGTATTGTTGTTCGGGGTCTATgtgttccccg\r\n",
      ">crDNA|J|3|2q2X0hEcddMla|3.97536e-08\r\n",
      "gggataaaccGTCGGTTTGCGCGTATTGTTGTTCGGGGTCTATgtgttccccg\r\n"
     ]
    }
   ],
   "source": [
    "# let's take a look at the sequences \n",
    "!cd $blastDir; \\\n",
    "    head -n 10 vNT_crDNA-proto_aln.fna"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* __The alignment is really just for each crDNA-protospacer pair (every 4 lines)__\n",
    "* First, the crRNA (sequence name starts with \"crRNA\")\n",
    "* Second, the protospacer (sequence name start with \"proto\")\n",
    "* Example: \n",
    "  * ~~~\n",
    ">crDNA\n",
    "ATAGACA\n",
    ">proto\n",
    "ATAGACA\n",
    "~~~\n",
    "\n",
    "* **Note:** the lower case letters are sequence adjacent to the actual protospacer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**If you wanted each crRNA-protospacer pair individually, you could use the `split` bash command like this:**\n",
    "\n",
    ">WARNING: this produces a lot of files, so we will make them in a new directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of files produced: 1594\n"
     ]
    }
   ],
   "source": [
    "# file/directory\n",
    "inFile = os.path.join(blastDir, 'vNT_crDNA-proto_aln.fna')\n",
    "outDir = os.path.join(blastDir, 'vNT_crDNA-proto_aln_split')\n",
    "if not os.path.isdir(outDir):\n",
    "    os.makedirs(outDir)\n",
    "outPre = os.path.join(outDir, 'split')    \n",
    "\n",
    "# spliting alignment\n",
    "!split -l 4 -d -a 4 $inFile $outPre\n",
    "\n",
    "# checking out files\n",
    "!printf 'Number of files produced: '\n",
    "!find $outDir -name \"split*\" | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">crDNA|C|9|2BRoy0pg1sN3q|1.08991e-08\r\n",
      "gggataaaccGTTTGGATCGGGTCTGGAATTTCTGAGCGGTCGCgagttccccg\r\n",
      ">proto|C|9|2BRoy0pg1sN3q|1.08991e-08\r\n",
      "...ataaaccgTTTGGATCGGGTCTGGAATTTCTGAGCGGTCGCgagttccccg\r\n"
     ]
    }
   ],
   "source": [
    "## Let look at one of them\n",
    "!find $outDir -name \"split*\" 2>/dev/null | head -n 1 | xargs -I % cat % 2>/dev/null"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting PAMs\n",
    "\n",
    "* **Looking at the alignments can be very useful, but it can help to extract what we specifically want and summarize.**\n",
    "  * Let's pull out the potential PAM regions from the alignment \n",
    "  * We are going to assume that the PAM is:\n",
    "    * 3bp long \n",
    "    * 1bp away from the 5' end of the protospacer ('right' side of the alignment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\r\n",
      ">proto|C|9|2GoDPvFP7iQVG|2.65694e-06\r\n",
      "agt\r\n",
      ">proto|J|13|1tPvM8ZwJiJQY|3.97536e-08\r\n",
      "tgt\r\n",
      ">proto|F|3|J5IiK0FqYBYc|1.08991e-08\r\n",
      "tgt\r\n",
      ">proto|E|7|2NfzYigGApLo4|3.97536e-08\r\n",
      "agt\r\n",
      ">proto|F|5|35jIrFszg79Lq|1.08991e-08\r\n",
      "tgt\r\n"
     ]
    }
   ],
   "source": [
    "!cd $blastDir; \\\n",
    "    CLdb -- arrayBlast -- get_PAM -PAM 2 4 -f - \\\n",
    "    < vNT_crDNA-proto_aln.fna \\\n",
    "    > vNT_crDNA-proto_aln_PAM.fna\n",
    "\n",
    "\n",
    "## checking output\n",
    "!cd $blastDir; echo; \\\n",
    "    head vNT_crDNA-proto_aln_PAM.fna"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Summary table of each unique PAM sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   1201 tgt\r\n",
      "    374 agt\r\n",
      "      6 \r\n",
      "      4 agc\r\n",
      "      3 att\r\n",
      "      2 gtg\r\n",
      "      2 ggt\r\n",
      "      2 cgg\r\n"
     ]
    }
   ],
   "source": [
    "!cd $blastDir; \\\n",
    "    egrep -v \"^>\" vNT_crDNA-proto_aln_PAM.fna | \\\n",
    "    sort | uniq -c | sort -k 1 -n -r "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**And if we reverse the sequence...**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   1201 tgt\r\n",
      "    374 tga\r\n",
      "      6 \r\n",
      "      4 cga\r\n",
      "      3 tta\r\n",
      "      2 tgg\r\n",
      "      2 gtg\r\n",
      "      2 ggc\r\n"
     ]
    }
   ],
   "source": [
    "!cd $blastDir; \\\n",
    "    egrep -v \"^>\" vNT_crDNA-proto_aln_PAM.fna | \\\n",
    "    rev | sort | uniq -c | sort -k 1 -n -r "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**We get the (less prevalent) PAM region ('TGT') for known I-E CRISPR systems as shown in [Rorek et al., 2013](http://www.annualreviews.org/doi/abs/10.1146/annurev-biochem-072911-172315)!**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Spacer (crDNA) - protospacer mismatch distribution\n",
    "\n",
    "* Let's look at potential spacer targeting by assessing mismatches of the crDNAs and protospacers.\n",
    "* This script will make 2 summary files of mismatches for each crDNA-protospacer alignment\n",
    "* Given that spacers can vary in length, the alignment is relative to a user-defined 'SEED' region.\n",
    "* I'm going to use the default SEED region: '-SEED -8 -1'\n",
    "  * This SEED region covers the last 8 bp of the alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SEED summary file written: 'vNT_crDNA-proto_aln_SEEDstats-summary.txt'\n",
      "SEED by-position table written: 'vNT_crDNA-proto_aln_SEEDstats-byPos.txt'\n",
      "\n",
      "-rw-rw-r-- 1 nyoungb2 nyoungb2 2.3K Dec 31 06:59 vNT_crDNA-proto_aln_SEEDstats-byPos.txt\n",
      "-rw-rw-r-- 1 nyoungb2 nyoungb2 457K Dec 31 06:59 vNT_crDNA-proto_aln_SEEDstats-summary.txt\n"
     ]
    }
   ],
   "source": [
    "!cd $blastDir; \\\n",
    "    CLdb -- arrayBlast -- get_SEEDstats \\\n",
    "    --SEED -8 -1 \\\n",
    "    -prefix vNT_crDNA-proto_aln \\\n",
    "    < vNT_crDNA-proto_aln.fna \n",
    "    \n",
    "\n",
    "\n",
    "## checking output files\n",
    "!cd $blastDir; echo; \\\n",
    "    ls -thlc vNT_crDNA-proto_aln_SEEDstats*.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "region\tpos_rel_SEED\tpos_rel_align\tcount\tmismatches\tmismatch_norm\n",
      "SEED\t1\t11\t1594\t12\t0.00752823086574655\n",
      "SEED\t2\t12\t1594\t7\t0.00439146800501882\n",
      "SEED\t3\t13\t1594\t3\t0.00188205771643664\n",
      "SEED\t4\t14\t1594\t3\t0.00188205771643664\n",
      "SEED\t5\t15\t1594\t3\t0.00188205771643664\n",
      "--------------------------------------------------\n",
      "alignment\tcrDNA\tprotospacer\tregion\tcount\tmismatches\tmismatch_norm\n",
      "910\tcrDNA|I|4|1TElyNyRCfHEA|3.97536e-08\tproto|I|4|1TElyNyRCfHEA|3.97536e-08\tSEED\t10\t1\t0.1\n",
      "910\tcrDNA|I|4|1TElyNyRCfHEA|3.97536e-08\tproto|I|4|1TElyNyRCfHEA|3.97536e-08\tnonSEED\t27\t0\t0\n",
      "910\tcrDNA|I|4|1TElyNyRCfHEA|3.97536e-08\tproto|I|4|1TElyNyRCfHEA|3.97536e-08\tprotospacer\t35\t1\t0.0285714285714286\n",
      "969\tcrDNA|J|9|2bqVUCgqCkNao|3.97536e-08\tproto|J|9|2bqVUCgqCkNao|3.97536e-08\tSEED\t10\t1\t0.1\n",
      "969\tcrDNA|J|9|2bqVUCgqCkNao|3.97536e-08\tproto|J|9|2bqVUCgqCkNao|3.97536e-08\tnonSEED\t27\t0\t0\n"
     ]
    }
   ],
   "source": [
    "# checking files\n",
    "!cd $blastDir; \\\n",
    "    head -n 6 vNT_crDNA-proto_aln_SEEDstats-byPos.txt\n",
    "    \n",
    "print '-' * 50\n",
    "\n",
    "\n",
    "!cd $blastDir; \\\n",
    "    head -n 6 vNT_crDNA-proto_aln_SEEDstats-summary.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plotting mismatches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Need help? Try the ggplot2 mailing list: http://groups.google.com/group/ggplot2.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "library(ggplot2)\n",
    "library(gridExtra)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAyAAAAHgCAMAAABq2fnHAAAC9FBMVEUAAAAAv8QBAQECAgIDAwME\nBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUWFhYX\nFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMlJSUmJiYnJycoKCgpKSkqKior\nKyssLCwtLS0uLi4vLy8wMDAyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6Ojo7Ozs8PDw9PT0+Pj4/\nPz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tMTExNTU1OTk5PT09QUFBRUVFS\nUlJTU1NUVFRVVVVWVlZXV1dZWVlaWlpbW1tcXFxdXV1eXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVm\nZmZnZ2doaGhpaWlra2tsbGxtbW1ubm5vb29wcHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6\nenp7e3t8fHx9fX1+fn5/f3+AgICBgYGCgoKDg4OGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Pj4+Q\nkJCRkZGSkpKTk5OUlJSVlZWWlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKj\no6OkpKSlpaWmpqanp6eoqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2\ntra3t7e4uLi5ubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJ\nycnKysrLy8vMzMzNzc3Ozs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc\n3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u7u7v\n7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4dm34+Pj5+fn6+vr7trH7+/v8/Pz9/f3+/v7///8q\n/YX8AAAgAElEQVR4nO2de3wU5b3/Fw01NjFsyoqgWGKlpD+pUKKABAxUKcfL8QKVtNaSX38t1Fo9\ntVy09DQHUy1SPAcrt0CQKNqgtSAXL0ECaCIEQQohMTHBAEnI5h6SjUnOJvv95/fMzszuzn1m99ns\n7M73/eprdp65fObb8Lyd2cs8YwMEQRSxRboABDEzKAiCqICCIIgKKAiCqICCIIgKKAiCqICCIIgK\nIkGq+lzrtvRHphQEMR9CQVbFNz03ddLjEaoFQUyHUJCRJz1jaqsdEaoFQUyHUBD7uaOToSEpQrUg\niOkQCrIk9eYtNdPmR6gWBDEdQkHaCwoGKtdcilAtCGI6AgVxu8e5CW1BX2J1t0ro6ZAuC4k2ynmt\nfbQTaee191IOpF5hp4tyoK4KKfR+HQQKEhdni2PIDDasvU5CX7N0WUjQzqvzNFAOpF1h4wDlQOoV\ntvZQDmxw6tiIQu/XgfASa25oYSgIAwoSMqYVJERQEAYUJGRMK0jhHakMwYahIAwoSMiYVpBxK06X\nE4INUxVkmACt9jDNtsHAurqvhWi264y1qQfGYsXRLsj1rpDCUBCagZaoOMoEefGFgVDCUBCagZao\nOMoESU8cMUH7PcihMnA/v379x549217r8k64FSgIzUBLVBxlgpSzqO4xuHVVGTj/QeZqCuDEh94J\ntwoFoRloiYqjTBCAgXqPxh6ewcIyqNj0RkF7UTE4t3onZPFnR44cqemQMNDNzwn7q1Z7mGbbYGBH\nh+gfR7PdYaxNPdASFcvT2aW0xk9zmIwQIRTk4uyE5NKZNRr7HCiD82fg9I59Z6HvZe+ELPxo9+7d\nFd0SBnv5OWF/1WoP02wbDOzuFv3jaLa7jbWpB1qiYgV6FNf4iMBPTQDmPtGT4l56l8Y+RBBC/0sf\nkZPHFu+EW4GXWDQDLVFxlF1ifbMNUqApgW8qXG8RQY4cg5od1Tvh8w+9E24FCkIz0BIVR5kgk94l\nguz5PttQvN4igrje2Lq9xbPvtR3d3gm3AgWhGWiJiqNMkCLHgqT5o95nG/qutwJBQWgGWqLiKBME\nmvNychu4edH1lg5QEJqBlqg42gRxVjGw84LrLV2gIDQDLVFxlAnypC1lPIFtCK63dIGC0Ay0RMVR\nJkhiSWAr8HpLFygIzUBLVBxlgqTVhhSGgtAMtETFUSZI8XVLsgnMrJ0leayBMBSEZqAlKo4yQWak\nrVhJYGbb2v76w+KG4ru3GghDQWgGWqLiKBPEEfgDlxuZEi7eaCAMBaEZaImKo0yQp3cENMYeJpND\n3zYQhoLQDLRExVEmyIwrx/oHbfjvbz2z+ZlvrTMQhoLQDLRExVEmiPCGqcNPZf7uEyNhHc0S3L5l\nwv6q1R6m2TYY2Nws+sfRbDcba1MPtETF8rS0Kq3x00hLAXX0joula0AgPIPQDLRExVF2BgnQIG1v\nGgu7RteAQCgIzUBLVBxlggRoUNpcVMpQxK7RNSAQCkIz0BIVR5kgARqIhnrXNSAQCkIz0BIVR5kg\nARqIhnrXNSAQCkIz0BIVR5kgAg0EQ73rGRAIBaEaaImKwyVIfpxxF2RRHRdroMHtXydoyIOC0Ay0\nRMXhEqSzyqgJCqh8zOtcODxx+EKnTEMJFIRmoCUqpibI5oQnn228J+nOyu4FI7JsJ8kZpCcreUwO\nbL56YeKDmv9p1ytIipeJs19khmF4KMsJzp9zT/QUNFCQ4HqDyQ5ghorpCWJbXv7byY2PTl1zzce/\n8wryYkJJnu3wZttbebbPaAmSN2v3yd13bnjvvsWkkcT8crF5BLtG0EBBgusNJjuAGSqmJ0j8AMxO\nTk0Z85PZ8IVXkMcyYPDqv21OgAu2I7QEGVdPJnUTwMXcBfJdZrirD7l37IIGChJcbzDZAcxQMT1B\nEgEWz2rJ+69V9pJlXkFWJ5S+Ss4giVQFKSWTY9dDIyPIW/ZF2Yvsb7NrAhuHysA/sDuO7h6uQEtW\nHJIgdXOS5hxpnZs43yuIa5F99J+BriB5o5a9vGzUhqLr/8K0qteuWFvNr/I1vKO7+wd2x9HdwxVo\nyYqDFsRH4Sr3PtuF4I1QFQT+9YdFzxyH8mK2JTv0qHd0d//A7vzo7gwoCM1AS1YcuiA1s24au5qW\nHqof8wYMPSr85eKBMvAP7I6ju4cr0JIVm3p0d9EPeAOGHhX+cpEI4h/YnR/dHZ8PYoKnbUQ4kMIB\nTP18kNLmUha2GTD0qPCXi0QQ/8DuOLp7uAItWXHQl1hfiwmDICIChh4V/nKRCOIf2B1Hdw9XoCUr\nNrUgE3nYpnCodz1hKAjNQEtWbGpBTvJwbRx61EjbjN1tiANjXRCe7ZIlorfvSqAgNAMtWbHJBSlf\nnJmZ+cBoZlZwvSV6+46CyLTN2N2GONDcgnj+OGX6PfVVI9LT05f5plNnafRpoSC3Zy29N3/qHmZW\nfL1V1edat6UfBVFsm7G7DXGguQX5ZJYH/v54VYa3N/umZ7+rPtqCUJD45q674dJMttGbewjy1/ax\njVXxTc9NnfQ4CqLYNmN3G+JAkwmy8xePpe/qmH/3nLPeucqbCt0D7WJB4NEPDAiSchjSWl0j2caS\ntM+h+PZfsI2RJz1jaqsdKIhi24zdbYgDzSbIHM/RR/6cDQfu985B8WMTMyur7BkZGeu4KSPIim0G\nBNl4Ve0fp9z5I84J5reJ1Xa2YT93dDI0JKEgim0zdrchDjSbINlwbkFWIfR+zzt3ugY8b02TnEF+\n9qGkHysLAud73K+v62Dnx39KJsXc89iWpN68pWYa3lGo3DZjdxviQLMJkkO0IGeQj+7zzu3JHISq\nKWJBKlKNvAcR8Kbjt2v+w8F95usuKBioXHMZBVFsm7G7DXGgGQXpeHgueQ/CzLn/lH77rBPez68W\n+KbTZh5X79NqY/NW5CzJPuVr4agm6m0zdrchDjSZIFQIFKQRlEctwVFNorG7DXFgrAuS3PDty17Y\n5jQWtqFrVBN8/AHNQEtWHPTjD74WEwZBliZfwT65k22WlJQUF2TsZBs4qkk0/vd4iANj/QxCzhOS\n9U23sq84qkk0drchDoykIP8rJgQpAhG/SRe/FT/BffUhGOJECRSEZqAlKw5akGFiQvPCh1AQwVtx\n5g1I2vCnuFWCIU4UQEFoBlqyYpMLIngrXsJQzo9r4qk8WDGoEYaC0Ay0ZMUmF0T5rXj5ZMeUayfh\n4w+o9oYIH8CMFZtcEMFbccFzO6ev6If+FTPUw1AQmoGWrNjkggjeigue22lvIhMnfsxLtTdE+ABm\nrJimIMHdIaUqiOCtuOC5nY9uIJNXHkFBaPaGCB/AjBXTFCS4O6TUBQlE8NzOXw6f9OAk272ZmZko\nCLXeEOEDmLFimoIEd4eUfkEEDyzM94GCUOsNET6AGSumKUhwd0jpF0TtuZ3u59ev/xgffxDWQEtW\nTFOQ4O6QUhXk1jM6d3P+g0zw8QdhDbRkxTQFCe4OKVVBcrJ6FbbrCpgSKja9UdCOjz8Ia6AlK6Yp\nSHB3SKkKkmG/arzvbUcAwsGrCefPwOkdgscf/H3jxo2f90vwuPk5Yfla7WGabYOB/f2ifxzNdr+x\nNvVAS1bM4Zb2JDEdmoJQQSjIaRbJVsLBq1n6XxI8/uB8dXV1Hd4PQjHQkhVz94O0G74fZGgEETxU\nqnBmwEjWosGrjxyDmh34+IOwBlqyYpqXWGEQJOChUgDjcsrEn2L5Ru11vbF1ews+/iCsgZas2OSC\nBDxUCuC6noA1AaP2KoOC0Ay0ZMUmFyTgoVIAzwV+kx4waq8yKAjNQEtWHLQg4UIoSMBDpVJTJ1wR\n8E26YNReJVAQmoGWrNjkggQ8VKq8/Lj3i3TuU2PBqL1KoCA0Ay1ZsckFCXiolOirD8GovUqgIDQD\nLVmx2QUZKD/4Jfs5r/irj8BRe5VAQWgGWrJikwty6nuOKY7buBtCdD23UwAKQjPQkhWbXJDbmRtr\nl2YEG4aC0Ay0ZMUmFyTJf2Otzud2CkBBaAZasmKTC7JwM5lsfJiZ1fncTgEoCM1AS1ZsakEefXS+\nbcqCKVcuZpuvGi4BBaEZaMmKTS2I6K7aOYm3PLlb43MrIe31Evpb+Dlhf9VqD9NsGwysrxf942i2\n6421qQdasmKWS05pT5JA0QIVRB/zOqsYuEZ/6X8vtN9hIKzDKcHdzs8J+6tWe5hm22Cg0yn6x9Fs\nO421qQdasmKWphZpTxJziZYC6ggFedKWMp7ANno+Xv3wmP/zKwNheIlFM9CSFZv6EgsgsSSg8Y1r\nntjVbCgMBaEZaMmKTS5IWm1AY/XD49J+k18F+kFBaAZasmKTC1J83ZJsAt/0VD6ZqPaUTzEoCM1A\nS1ZsckFmpK1YSWAbe5+dlTQr+5CBMBSEZqAlKza5II7WgMZty9/rAkOgIDQDLVmxyQV5ekdIYSgI\nzUBLVmxyQWZcOTZVblwsnaAgNAMtWbHJBVEbjlcHKAjNQEtWbHJBQgQFoRloyYpNLsg0lmDDUBCa\ngZas2OSClJSUFBdk7DQSgI8/CFegJSs2uSBemm41EoCPPwhXoCUrjgZBTiRJlymDjz8IV6AlKza5\nIMwbkLThTxkJ4B9/sC47O7tEbUNhf9VqD9NsGwwEEP3jaLbBWJt6IFb8tVp/MvgldrBI3oOUlJR7\nFLaVhX/8Qa/L5WppkNDfKl0WEjLHCA1opBxIu8KmAcqB1Cts76Uc2NikvU1kbpgyju7HH1CCdl6d\np4FyIO0KGwcoB1KvsLWHcqBJL7EyeIwE6H78ASVQkNBBQfQTKMi7XnK+OTnYMBSEAQUJGZMKwnD5\n9/a/uoMNkxGkqILy3+4S5by6vbWUA2lXWP0B5UDqFZ75mHJgvZ7/aIXY83UiEmTXtx86TzV/05dU\n48LAc0P0cUjQNP410hVo8XloPwI3NQJBLjw8bjflfBQkZFCQSBIoyNpvLaPeWVCQkEFBIkmgILYr\n4lko5u8dokvF4Mnv0d4morQZ+m1cJKg+EOkKwkegIL7vYCJWDYKYDKr3gyBIrIGCIIgK4RUk8GYR\nc3KozNRFut/ZtuGimQuEvh1bc1tNXWFohFeQwJtFzMjg1lVlpi6yYi/UbTZzgXCsCE7sM3WFoRFe\nQQJvFjEjnsHCMlMXWd8EbblmLhAaO+HkAVNXGBrhFYS/WcS8HCgzeZEXN1Sau0B484VWk1cYCuEV\nhL9ZxLwQQcxcpKdwa4OpC4TeQajZbuoKQyO8ggTeLGJOiCBmLvLMzkGT/xU/OAHn80xdYWiE+VOs\ngJtFzAkRxMxF7lq7fv3rZi4QOl/N3VRv6gpDA78HQRAVUBAEUQEFQRAVUBB5PBMc/QAnJ0Jpmt5d\n4pTuxGQi4twaQeRQ/iC5bQXxp+aOGPnvlQAOG0OV79Vus8VPP6hclWKNiAIoiDyfjRn/nrfXNu/V\nu4ti52Mi4twaQQJB5LYNjB+4cWV1/dKJHnB81EYY8L3aD7d9tS7xuGJVKIhRUBB5li5dmeU7g+Sl\npGxLgfL0F2+46QCUz/i9I734tmueAjj8g4R5F/nlc23jugDyf/mYfUYFwNsTRsx3gvvXyY7nmAiy\nsiiNX8rtQDidkXMrl8IIsummq6dVcNvO2wiwOpNbSWDi2d0JtbZOYsn9beAoZcvlX+0nyeTZHzOz\n7LGZ3dhY5lhsjYgBUBBZBm88ddreywnyr2uP1s0kgiQ+3738Dii/8vWWtOu+KrE5mx27W38zm1/O\n/tc5f/grzuWTPTX2D1qyMqEg9dyJ+C99l1jcUn4HIog96wyXQg51/qpDTVmLuW3XPwAw4x1uJUOc\nm9ud4P7+vP0uZkZWkE++w8yyxya7cbHMsfAMYhgURJbDPwC4ZRcnyB+WAewigoxww+lUKB8LsGIJ\nQErVtgUAPQkD3HJOkEkA/SMr1y4CcH5joODmo56mPp8g3FJ+ByLIVb3ApZBD9dRC19JMbtv6a3oa\nknu4lcy2cW5ud6bR+8q/OeYdI2Ik2u32G/yvXkGqvLeEsscmu3GxzLFQEMOgILI8fvXo0fE/5QRZ\n9DK52CKCTAAoT2X+ByuzAcZX5SSlpKQk13PLOUEeJJMpB1cwj9JObBjInTwmx+UThFvK70AEGQ/A\npZBDuf/ztrvu5QWB6fs3LuJXMtvGubndyaSvhziSF38MHLvOnTv3FfheA84g7LHJblwscywUxDAo\niBz91xY1NBxM7GYFeWY5wO4Ub58WCJI7n3TDUg+33H8GcV9bsZa8gXEOd1fWQO3Udf4zCLuU3wG8\n5xEuhRxqx5QWyPcJ8pcn7tnPr2S2JWcQdncy/9oPmSWzN6q9B2GPTXbjYr3nLBTEKCiIHPu/Q/qk\nJ2UnK8jxUaX1GTKCNIza27Q0HXyCtJE9820bmp6ZNPjliMLWnz8Ca6Zdqpmc6xWkjUy5pUJBuBRy\nqHV3uhqnPsBtC1+MvaGfX8kQ18btTmhM/lNlxStJldynV32+V/vhttqXEz9jtmGPTXbjYllB2iL0\nF41aUBA5Fj3NTJ9awH2KteH6CRsmSgWB9ycmzKn29feFScynWPf+OGn6WYC3JiQ91AiXH0oc+et+\nJmJhEvMpFrtUKAiXQg7VPnfkjHdHb+e2hVuW+FYykHh2d4bK+0Yn3bGf/x4k3/fKfA8yjf0ehD02\n2a2OjfUey1sjYgAURJvyItJRZ+vbNj8zvLUgQwwKok3x9Y2u+1fr2xYFiTFQEB08f8PYX+n8NTcK\nEmOgIAiiAgqCICqgIAiiAgqCICqgIAiiAgqCICqgIAiiAgqCICqgIAiiAgqCICqgIAiiAgqCICqg\nIAiiAgqCICqgIAiiAgqCICqgIAiiAgqCICqgIAiiAgqCICqgIAiiAgqCICoEJ0h3q5Q+mWWh0kY/\nsqOHfmY46rws9zcOlTDU6eqkn9muayvKJigQnCAddRLqoV66MFSa6Uc29dPPrG+in9nRTT+z8RL9\nTFcb/cwmXV2JsgkKoCAUQEHogoLoBQWhCgpiHBSEAigIXVAQvaAgVBliQb4Woj8TBdELCkIVFMQ4\nQQrSKMEJTunCUGkNQ6SbfqazhX7mZRf9zOYm+pm9nYqrRILoz2zR05UaKJugAApCARREFisLIj3h\n4SUWXfASSwvKJiiAglAABZEFBQkEBaELCqIFZRMUQEEogILIgoIEgoLQBQXRgrIJCqAgFEBBZIl5\nQar6XOu29MutQUEEoCCyxLogq+Kbnps66XG5VSiIABREllgXZORJz5jaaofcKhREAAoiS6wLYj93\ndDI0JHnnD5WB+/n16z/mVqEgAlAQWWJdkCWpN2+pmTafzA1uXVUGzn/4V6EgAlAQWWJdEHdBwUDl\nmstkzjNYWAYVm94oaOdWoSACUBBZYl0QgIF6Dzd3oAzOn4HTO8jsjpdeeunYgBSQWWZCBj2RrkAf\n0VKnZ1BxlUgQygfuDL8cDCqCXJydkFw6s4YXhND/Epk4ibyXnBKaoEm6MFTa6Ue2uelnNrXRz+zq\noZ/Z0kI/s7dTcZVIEP2ZbXq60qWh0ENVkLlP9KS4l97lnSeCHDkGNTu4VXiJJQAvsWSJ9Uusb7ZB\nCjQleOeJIK43tm5v4VahIAJQEFliXZBJ7xJB9nxfbhUKIgAFkSXWBSlyLEiaP+p9uVUoiAAURJZY\nFwSa83Jy5W9sREEEoCCyxLwgyqAgAlAQWWJdkMKZExnkVqEgAlAQWWJdkHE5ZeUEuVUoiAAURJZY\nF+S6HsVVKIgAFESWGBTk1jMBjedeGFDaCwURgILIEklB8uNC0sKPUJCcrF5+NjV1whUjJqSmpsrt\nhYIIQEFkiaQgnVU07ACxIBn2q8ZzTpTzoCCaoCCyDKkgmxOefLbxnqQ7K7sXjMiynSRnkJ6s5DE5\nsPnqhYkPumkJcpqFbfTmHoL8tX2ygjRLaIEW6cJQaaMf2eGmn9nSSj+zu4d+ZmsY/on6uhRXiQTR\nn6nrn71RJIhteflvJzc+OnXNNR//zivIiwklebbDm21v5dk+oyVI4C/cYUna51B8+y9kBZH6jGcQ\nuuAZRAuRIPEDMDs5NWXMT2bDF15BHsuAwav/tjkBLtiO0BIk8BfuMLKaTKrtKIgmKIgsQytIIsDi\nWS15/7XKXrLMK8jqhNJXyRkkkaYggb9wh/GfkknxeOOCDBOi/w8jAQWhSqQF0WtMsILUzUmac6R1\nbuJ8ryCuRfbRfwa6ggT+wh3edPx2zX84tqMgmqAgsgypID4KV7n32S4Eb4SqIIJfuHdV5CzJPtWF\ngmiCgsgSGUFqZt00djU1P0SCBPzC3e0e5ya0JaEgmqAgskRGEMqIPsXy/8I9Ls4Wx5AptxcKIgAF\nkSUWBRkoP/gl/znvXOW9UBABKIgsQyrI12LCIsip7zmmOG6rDlyEb9K1iUVBNHu3dm+PQUFuX9EP\n/Usz2Eb54szMzAdGe+cPlYFnz7bX+HfsKIgAFMQqgiQ1kYlzBGdL1tJ786fuAW7o0ZoCOPEhCiIH\nCmIVQRZuJpOND7ON+Oauu+HSTOCGHi0qBudWFEQOFMQagjz66HzblAVTrlzMNlMOQ1qra6R3/kAZ\n7DsLfS+T2UO7d+8+2y0FfHMiQWS21UtPCPsqRQ7SzwxHnf1u+pkul/5tRd1NukCygfEIJXT9OVuN\nCuL545Tp99RXjUhPT1/mm06dVapfkHwetrnxqto/TrnzR7wgH5EzyBYye/LIkSNfdkgB35xIEJlt\n9dIdwr4KdA3Qz+zsop/Z00c/8/Jl/duKupt0gWQD4xFKdHXqKLDZqCCfzPLA3x+v8r7D9k/Pftel\nWxCCs4qBa5zvcb++roMXpHonfI7vQWTBSyxTXmLt/MVj6bs65t8956x3rvKmQvdAu1gQePQDA4I8\naUsZT5DaQgTx7HttRzcKIgcKYk5B5niOPvLnbDhwv3cOih+bmFlZZc/IyFjHTRlBVmwzIEhiiaIt\nAlAQASiIOQXJhnMLsgqh93veudM14HlrmuQM8rMPJd1bWZC0WiVbUBAVUBBzCpJDtCBnkI/u887t\nyRyEqiliQSpSjbwHKb5uSTZBxhYURAUUxLyCdDw8l7wHYebcf0q/fdYJ7+dXC3zTaTOPq/ohEmRG\n2oqVBBlbUBAVUBBTCkIFoSCOwA+XBbagICqgIFYR5OkdAQ2H6KsYFEQJFMQqgsy4cmyqb6w4gS0o\niAooiFUEEYwVJ7AFBVEBBYm8IP8rJiyCKNuCgqiAgkRekGFiwiLINBauVdXnWrelHwXRBAWxiiAl\nJSXFBRk72caq+Kbnpk56XFYQ6UiQAUOPiirVP+RkcGNQGgOHHtWHqDNLF0g2MB6hRDBDjw6NIF6a\nbmVfR570jKmtdsgKIvUZzyB0wTOIFhET5AQ30o/93NHJ0IDD/miDglhFEOYNSNrwp9jGktSbt9RM\nm09BkKCNQUGoYjFBgrtDSlWQEoZybtwfd0HBQOWayyiIJiiIKQUJ7g4pVUH0goIIQEFMKUhwd0ip\nCJLBwzTS9qaxoCCaoCCmFCS4O6RUBHnXS843JzON0uZSFhREExTElIIEd4eUiiAMl39v/yv/RDfB\nDeooiDIoiCkFCe4OKXVBdn37ofP8PL1bblEQ4xgRRG/Hs5ggwd0hpSbIhYfH7fa36N1yi4IYBwXR\nQlMQKgQKsvZbywIflyO65db9/Pr1H6MgcqAg1hDEdkU8C9sU3XLr/Id/SxREAApiDUEaeNim6Jbb\nik1vFLST1wvV1dV10t+Oqf1YUbJALzI/igyVsPxYsZ1+ppEfK4o6nuJ2bQZ+VCnJFC/QPqhmhBLt\nen5UGbkfK/KIbrk9fwZOM/cYvr1x48bj/VLANyeqVGaBXtz6N9Ud6aGfGY46Bwb1byvqeIrbGSlT\nkileoH1QzYiQ6uyIuCDSW277X+Jm8BJLAF5iRf4SK1yoCCK65fbIMajhlUFBBKAglhREdMut642t\n21u4eRREAApiSUFUQEEEoCAoiBAURAAKYkFBGsGpvBcKIgAFsaAgyQ3fvuxFbiUKIgAFsaAgS5Ov\nsHuRW4mCCEBBLCgIwEPKq1AQASiIJQUBGGhwy69AQQSgIJYUxLlweOLwhbJv1UMURNsgeo5J/vhW\nEESjMyt3zeAzaZYVJYI8lOUE58/DMewPCqINCqJFuJQQoiJIEvNjxeYRcqs6nRKaoImfFXVm6QLN\nDXQsCJY2d/D7KtHUQj+zq0f/tqJ+Jl0g7olhyKRZVque/9MNYTJChIog32Vubv9Q9vEHnQ0SLsEl\nflbUmaULNDfQsSBYWtzB76vEpWb6mZ3d+rcV9TPpAnFPDEMmzbKaL8mkiakPkxEiVAR5y74oe5H9\nbblVeIklAC+xLHmJBdVrV6ytll2DgghAQawpiDIoiAAUBAURgoIIQEFQECEoiAAUBAURgoIIQEFQ\nECEoiAAUBAUREnlBghYGBUFBjICCUAAFQUEAPHu2vcaPTIqCCP+/oyAoCNQUwAn+yQooiPD/OwqC\ngkBRMTi3cvMoiPD/OwqCgsC+s9D3MnndkJ2dfUR1S1HnlS7Q3CCICCsj6mfSBeKeGIbMISrLT5f2\nJjTQL8hH5Ayyhbz2ulyuFumPKwN+zUsPmeOEHBmDv+bVi9NJP7Ong35mlPyaV0T1Tvhc33sQeuAt\nt1QxcsutXlxt9DOj8xLLs++1Hd3cPAoiAAWhS3QKEggKIgAFoUtMCnJhzwXaf6e6ujD8g1Z+RD+z\nroF+5Kmj9DMbwlDnJ6fpZ+r7Z6dsggLBCSJDf7bCEEEm49z6SFegj493a29jBgqMPTQ26kBBTAoK\nYg5QEJOCgpgDaoK4Xx2gFRVWLkVJxztTHOkK9FFUGekKwgs1QRAkFkFBEEQFFARBVKAlSODdIibm\nUFk0VOp+Z9uGi1FQZ9+OrbmtUVBnKNASJPBuEdMyuHVVWTRUWrEX6jZHQZ3HiuDEviioMxRoCRJ4\nt4hp8QwWlkVDpfVN0JYbBXU2dsLJA1FQZyjQEoS/W8TkHCiLjkovbqiMijrffKE1KuoMHuDh0U0A\nABV7SURBVFqC8HeLmBwiSBRU6inc2hANdfYOQs32KKgzFGgJEni3iIkhgkRBpWd2DkbFX/SDE3A+\nLwrqDAVqn2IF3C1iYoggUVDprrXr178eBXV2vpq7qT4K6gwF/B4EQVRAQRBEBRQEQVRAQQLwTHD0\nA5ycCKVpeneJU/qRPxMR59YIIofyB8ltK4g/NXfEyH+vBHDYGKp8r3abLX76QeWqFGtEtEFBAvhs\nzPj3vL22ea/eXRQ7HxMR59YIEggit21g/MCNK6vrl070gOOjNsKA79V+uO2rdYnHFXZDQUICBQlg\n6dKVWb4zSF5KyrYUKE9/8YabDkD5jN870otvu+YpgMM/SJh3kV8+1zauCyD/l4/ZZ1QAvD1hxHwn\nuH+d7HiOiSAri9L4pdwOhNMZObdyKYwgm266eloFt+28jQCrM7mVBCae3Z1Qa+skltzfBo5Stlz+\n1X6STJ79MTPLHpvZjY1ljsXWiAQHCuJn8MZTp+29nCD/uvZo3UwiSOLz3cvvgPIrX29Ju+6rEpuz\n2bG79Tez+eXsf53zh7/iXD7ZU2P/oCUrEwpSz52I/9J3icUt5XcggtizznAp5FDnrzrUlLWY23b9\nAwAz3uFWMsS5ud0J7u/P2+9iZmQF+eQ7zCx7bLIbF8scC88goYCC+Dn8A4BbdnGC/GEZwC4iyAg3\nnE6F8rEAK5YApFRtWwDQkzDALecEmQTQP7Jy7SIA5zcGCm4+6mnq8wnCLeV3IIJc1QtcCjlUTy10\nLc3ktq2/pqchuYdbyWwb5+Z2Zxq9r/ybY94xIkai3W6/wf/qFaQqntmEPTbZjYtljoWChAIK4ufx\nq0ePjv8pJ8iil8nFFhFkAkB5KvM/WJkNML4qJyklJSW5nlvOCfIgmUw5uIJsAIkNA7mTx+S4fIJw\nS/kdiCDjAbgUcij3f9521728IDB9/8ZF/Epm2zg3tzuZ9PUQR/Lij4Fj17lz574C32vAGYQ9NtmN\ni2WOhYKEAgrio//aooaGg4ndrCDPLAfYneLt0wJBcueTbljq4Zb7zyDuayvWkjcwzuHuyhqonbrO\nfwZhl/I7gPc8wqWQQ+2Y0gL5PkH+8sQ9+/mVzLbkDMLuTuZf+yGzZPZGtfcg7LHJblys95yFgoQA\nCuJj/3dIn/Sk7GQFOT6qtD5DRpCGUXublqaDT5A2sme+bUPTM5MGvxxR2PrzR2DNtEs1k3O9grSR\nKbdUKAiXQg617k5X49QHuG3hi7E39PMrGeLauN0Jjcl/qqx4JamS+/Sqz/dqP9xW+3LiZ8w27LHJ\nblwsK0hbhP6isQAK4mPR08z0qQXcp1gbrp+wYaJUEHh/YsKcal9/X5jEfIp174+Tpp8FeGtC0kON\ncPmhxJG/7mciFiYxn2KxS4WCcCnkUO1zR854d/R2blu4ZYlvJQOJZ3dnqLxvdNId+/nvQfJ9r8z3\nINPY70HYY5Pd6thY77G8NSLBgYIoUF5EOupsfdvmZ4a3FiRyoCAKFF/f6Lp/tb5tUZDYBQVR4vkb\nxv5K58+4UZDYBQVBEBVQEARRAQVBEBVQEARRAQVBEBVQEARRAQVBEBVQEARRAQVBEBVQEARRAQVB\nEBVQEARRAQVBEBVQEARRAQVBEBVQEARRAQVBEBVQEARRAQVBEBVQEARRAQVBEBVQEARRgYIg3a0S\nLsssC5E26ol91BPp19h9mXZiG/UiO3poJ+r7Q4bec/VAQZCOOgnt3dJlIdJMO7Ae6mlHUq+xrrud\ndmJDI+3EJjftRH1/yNB7rh5QEIqgIJRAQYyDgtABBTEGCkIRFIQSKIhxUBA6DKkgXwvRHYmCGAcF\noQMKYgwUhCIoiAgUhAEF4UFBRKAgDCgIDwoiAgVhQEF4UBARKAgDCsKDgohAQRg6miV09UqXhUgb\n7cAWaKEdSb3G5t4u2oktrbQT2wcUV4kE0R2p5w/ZGHrP1QOeQSiCZxAReAZhQEF4UBARKAgDCsKD\ngohAQRhQEB4URAQKwoCC8KAgIlAQBhSEBwURgYIwoCA8KIgIFIQBBeFBQUTEmCAH3UFFoCA8KIiI\nGBPklZ8u2e2S28izZ9trXfyrd+J+e3t+J7cWBeFBQUTEmCAArfk/z3q9TbJRTQGc+JB/9U4+/QA+\n38OtRUF4UBARMScIwfXPX/2Um63qc63b0k9miorBuZV/9U52l0P7ekaO1tbWpksSOl3SZSHSRjuw\nERppR1Kv8VJPJ+1EZzPtxBa34iqRILoj9fwhGyIkCIF7K7Iqvum5qZMeJ3P7zkLfy/yrd3J0D3z6\nElmS98ILLxQPSvB4pMtChH4iUE8Mw//r6P5DigTRnainxsuRE4Rj5EnPmNpqB5n7iJw0tvCv3snA\nvu17crnt8BKLBy+xRMTiJZYf+7mjk6EhicxV74TPP+RfvZOLlXDyELcdCsKDgoiIPUF8bzsIS1Jv\n3lIzbT6Z8+x7bUd34wb21TvpfnP7P/tQEBEoiIiYE8T/toPQXlAwULnmkmYECsKDgoiIOUH8bzvA\n7R7nJrQlaUagIDwoiIiYE8T/tgPi4mxxDJmaESgIDwoiIuYE8b/tIMzVGYGC8KAgIiInSH5cEDLI\nIRTE7X3bYfATZhSEBwURETlBOquMmqCAyse8hXekMmhGoCA8KIiIIRRkc8KTzzbek3RnZfeCEVm2\nk+QM0pOVPCYHNl+9MPHB4H6FKyNI4cyJDGxj3IrT5QQURDcoiIihFMS2vPy3kxsfnbrmmo9/5xXk\nxYSSPNvhzba38myf0RJkXE6Z34nrZX/Zi4Iog4KIGEpB4gdgdnJqypifzIYvvII8lgGDV/9tcwJc\nsB2hJch1PQGNF18YQEEMgYKIGEpBEgEWz2rJ+69V9pJlXkFWJ5S+Ss4giTQFeS7QifTEERN0vQeR\njnrnH1lxmBDdY+sFNdieIXBkRUoYGFlR3FYiiJEVGUHq5iTNOdI6N3G+VxDXIvvoPwNNQVJTJ1wR\n4EQ5i7YgUrn9ZxCRILr/GxLcf1SMgGcQShg4g+g9o4TyKVbhKvc+24XgjRARKEh5uciJgXqPjggU\nhAcFEREJQWpm3TR2dUhOCBBeYvXmHoL8tdyvEC/OTkgunVmjGYGC8KAgIiIhCGVE36SnfQ7Ft/+C\nbcx9oifFvfQuzQgUhAcFERFzgoysJpNqO9v4ZhukQFOCZgQKwoOCiBhCQb4WExZBxn9KJsXj2cak\nd4kge76PgugmxgTR7N6a/T3mBHnT8ds1/+HYzjaKHAuS5o96HwXRDQpiNEEB0wrSVZGzJPtUF9dq\nzsvJ1TF2BArCg4IYTVDApIKI75FyVjGgILpBQYwmKGBSQUT3SD1pSxlPALmRFeGTvE2tKIgIFMRo\nggJhEcTzxynT76mvGpGenr7MN506q1S/IKJ7pBJL+DnpyIr1eZ7yf6IgIlAQowkKhEWQT2Z54O+P\nV2Uw8/7p2e+q/yZX7n4Q7k16Wi2/QDqy4qFj4OkmS3pdLldLg4ROFz8nEkS6qV5kjhIal+AS7Ujq\nNTa4OmgnNjp1byrqcZK2dIHhBAX0/CHrNQXZ+YvH0nd1zL97zlnvXOVNhe6BdrEg8OgHBgQpX5yZ\nmfnAaLZRfN2SbALIjay475+vbmfev28gG6j+EkwkiGotiLkQ9zhJD9T6b7Z2Qgh0CZtygszxHH3k\nz9lw4H7vHBQ/NjGzssqekZGxjpsygqzYpnoYoSC3Zy29N38qNyr1jLQVKwkgN7Ji4e7BCxu5nfAS\niwcvsYwmKEDnEmtnNpxbkFUIvd/zzp2uAc9b0yRnkJ99aECQ+Oauu+HSTLbh4N+Fy4ys+MV70LwB\nBRGBghhNUICSIDlEC3IG+eg+79yezEGomiIWpCLVyHuQlMOQ1uoayTae3sEvlo6s6Nm3eeN5FEQE\nCmI0QQGKgnQ8PJe8B2Hm3H9Kv33WCe/nVwt802kzj6v6IRJk41W1f5xy54/Yxowrx4Y+aAMKEhIo\niDKaglBB9CnW+R736+s62Hm8YcooKIjRBAXMK4jOL88FoCA8KIjRBAVMK4j/y3OgNC4WChISKIgy\nkRDE/+U5UBoXCwUJCRREGWEv/F8xFORgEAri//IcKI2LhYKEBAqijLAXDhMTkhZ+hIL4vzwHSuNi\noSAhgYIoI+yFQyOI/8tzMDAulrR2FIQSKIgykRDE/+U54Me8xkFBjCYoYFpB/F+eGwBHVuSJsZEV\nRd1ZOi6iZIHhBAWCGFlxiC6xAr88T/EycfaL3RqCSOVWOYMEe0bBMwgdLHQGCe4OKVVBBFdVebN2\nn9x954b37luMgugDBTGaoAAdQYK7Q0pVEMFjoMcxt6TUTQDXWBREHyiI0QQF6AgS3B1SqoIIHgM9\njjkXHbseGlEQnaAgRhMUoPQmPag7pFQFCXgMNLnEGrXs5WWjNhRd/xcURB8oiNEEBegIEtwdUqqC\nBDwGmvCvPyx65jiUF2tEoCA8KIjRBAXoCBLcHVKqgggeA60XFIQnooLo7HwWEiS4O6RUBQl4DHTa\n3jQWFEQ3KIjRBAVM+0VhAKXNpSyaESgIDwpiNEEBkwqieNKQjqzYt2NrbnAjK6IgRkBBlBl6QQQn\njYk8IDey4rEiOLEPBRGBghhNUMCkgpzjYRoneUBuZMXGTjh5AAURgYIYTVDApILY4h0sgRsw45BK\nR1YEePMF5hLL+MiKYfl/gegY5jDkRMkRNA+pmRACXdqb0CBQkMdvvGt9wPNA/OOQSkdW7B2EGkYd\n42Pzitt6wbF5NRB1PqXNYnVs3qEQBAaP/O7mO//nItfyj0MqHVnxgxNwPo/bDi+xePASK7QE336m\nfYgngOf4s6l3sLP+cUilIyt2vpq7iVcYBeFBQUJL8O1nYkH63//VmLvZWcE4pCqgIDwoSGgJvv3M\nKkjf/v83Zl5uE9cSjEOqAgrCg4KEluDbz6SC/N8x9+W1BLQDxyFVAQXhQUFCS/DtZ1JBbMMTWYxF\noCA8KEhoCb79TCqI7xM0YxEoCA8KElqCbz+TChIkKAgPChJagm8/FESpPUxrB5kNgiTWBdHbGakl\noiAKoCA8KAilmlAQFEQfKIgyofdcPdAQRDrqncrIiurtYVo7yGwQJLE2sqKo82mNg0g/MZiRFRUC\nghhZMVzgGYQieAahVFOMnUGktaMglEBBlAm95+oBBaEICkKpJhQEBdEHCqJM6D1XDygIRVAQSjWh\nICiIPlAQZULvuXpAQSiCglCqCQWhKUiQxqAglBNREAVQEB4UhFJN0SaIdGRFfgkDCsKDglCqKdoE\nkY6syC9hQEF4UBBKNUWbINKRFfklDCgIDwpCqaZoE0Q6siK/JPSRFcUbaOygYwOLIup8WuMg0k8M\nZmRFwzX5icDIiopIR1bkl2iOrEgNHFmRDgZGVtRJi5t2omlHVlRCOrIiv4RB9RKLGnjLLR0M3HKr\nkyY37cSou8SSjqzonXBrURAeFIQS0SaIOigIDwpCiZgX5NSntP9kdZdoB17Yc4F2JPUa6z79F+3E\n+gbaiVUHaCfq+0OG3nP1QEEQGUreCUssVfqz3ZEuQZt/fBrpCrSp/VukKwgnKIipQUEiDQpialCQ\nSBMeQc5+HJZYqrhfHYh0Cdp8XB7pCrRp3BXpCsJJeARBkBgBBUEQFVAQBFEhLIIE3i1iUg6Vmb5K\n9zvbNlw0eZF9O7bmtpq8xtAIiyCBd4uYksGtq8pMX2XFXqjbbPIijxXBiX0mrzE0wiJI4N0ipsQz\nWFhm+irrm6At1+RFNnbCyQMmrzE0wiIIf7eIiTlQFgVVXtxQafoi33yh1fQ1hkJYBOHvFjExRBCz\nV+kp3Npg9iJ7B6Fmu8lrDI2wCBJ4t4hJIYKYvcozOwdN/6f84ASczzN5jaERnk+xAu4WMSlEELNX\nuWvt+vWvm7zIzldzN9WbvMbQwO9BEEQFFARBVEBBEEQFFESKZ4KjH+DkRChN07tLnNJP55mIOLdG\nEDmUP0huW0H8qbkjRv57JYDDxlDle7XbbPHTDypXpVgjogIKIuWzMePf8/ba5r16d1HsfExEnFsj\nSCCI3LaB8QM3rqyuXzrRA46P2ggDvlf74bav1iUeV6wKBQkGFETK0qUrs3xnkLyUlG0pUJ7+4g03\nHYDyGb93pBffds1TAId/kDDvIr98rm1cF0D+Lx+zz6gAeHvCiPlOcP862fEcE0FWFqXxS7kdCKcz\ncm7lUhhBNt109bQKbtt5GwFWZ3IrCUw8uzuh1tZJLLm/DRylbLn8q/0kmTz7Y2aWPTazGxvLHIut\nETEICiJh8MZTp+29nCD/uvZo3UwiSOLz3cvvgPIrX29Ju+6rEpuz2bG79Tez+eXsf53zh7/iXD7Z\nU2P/oCUrEwpSz52I/9J3icUt5XcggtizznAp5FDnrzrUlLWY23b9AwAz3uFWMsS5ud0J7u/P2+9i\nZmQF+eQ7zCx7bLIbF8scC88gQYGCSDj8A4BbdnGC/GEZwC4iyAg3nE6F8rEAK5YApFRtWwDQkzDA\nLecEmQTQP7Jy7SIA5zcGCm4+6mnq8wnCLeV3IIJc1QtcCjlUTy10Lc3ktq2/pqchuYdbyWwb5+Z2\nZxq9r/ybY94xIkai3W6/wf/qFaQqntmEPTbZjYtljoWCBAUKIuHxq0ePjv8pJ8iil8nFFhFkAkB5\nKvM/WJkNML4qJyklJSW5nlvOCfIgmUw5uIJsAIkNA7mTx+S4fIJwS/kdiCDjAbgUcij3f9521728\nIDB9/8ZF/Epm2zg3tzuZ9PUQR/Lij4Fj17lz574C32vAGYQ9NtmNi2WOhYIEBQoipv/aooaGg4nd\nrCDPLAfYneLt0wJBcueTbljq4Zb7zyDuayvWkjcwzuHuyhqonbrOfwZhl/I7gPc8wqWQQ+2Y0gL5\nPkH+8sQ9+/mVzLbkDMLuTuZf+yGzZPZGtfcg7LHJblys95yFggQDCiJm/3dIn/Sk7GQFOT6qtD5D\nRpCGUXublqaDT5A2sme+bUPTM5MGvxxR2PrzR2DNtEs1k3O9grSRKbdUKAiXQg617k5X49QHuG3h\ni7E39PMrGeLauN0Jjcl/qqx4JamS+/Sqz/dqP9xW+3LiZ8w27LHJblwsK0hbhP6iUQ0KImbR08z0\nqQXcp1gbrp+wYaJUEHh/YsKcal9/X5jEfIp174+Tpp8FeGtC0kONcPmhxJG/7mciFiYxn2KxS4WC\ncCnkUO1zR854d/R2blu4ZYlvJQOJZ3dnqLxvdNId+/nvQfJ9r8z3INPY70HYY5Pd6thY77G8NSIG\nQUHUKS8iHXW2vm3zM8NbCxIBUBB1iq9vdN2/Wt+2KEgMgoJo8PwNY3+l87fcKEgMgoIgiAooCIKo\ngIIgiAooCIKogIIgiAooCIKogIIgiAooCIKo8P8Bqt368sF12ZUAAAAASUVORK5CYII=\n"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i blastDir -w 800\n",
    "\n",
    "# loading table\n",
    "infile = file.path(blastDir, 'vNT_crDNA-proto_aln_SEEDstats-byPos.txt')\n",
    "\n",
    "tbl = read.delim(infile)\n",
    "tbl = tbl[tbl$region != 'protospacer', ]\n",
    "\n",
    "# mismatches and counts, & normalized mismatches\n",
    "p.mis = ggplot(tbl, aes(pos_rel_SEED, mismatches, group=region, fill=region)) +\n",
    "    geom_bar(stat='identity', position='dodge') +\n",
    "    labs(y='Number\\nof mismatches', x='Alignment position relative to SEED start')\n",
    "p.count = ggplot(tbl, aes(pos_rel_SEED, count, group=region, fill=region)) +  \n",
    "    geom_bar(stat='identity', position='dodge') +\n",
    "    labs(y='Number of alignments\\nwith nucleotide\\nat position', x='Alignment position relative to SEED start')\n",
    "p.norm = ggplot(tbl, aes(pos_rel_SEED, mismatch_norm, group=region, fill=region)) +\n",
    "  geom_bar(stat='identity', position='dodge') +\n",
    "    labs(y='Mismatchs /\\nalignments', x='Alignment position relative to SEED start')\n",
    "\n",
    "grid.arrange(p.count, p.mis, p.norm, ncol=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Notes on plots:**\n",
    "\n",
    "* The top plot shows that some spacer-protospacer are not quiet as long as the rest, which is why the 'Number of alignment with nucleotide at position' drops at the higher positions (> ~30).\n",
    "* There are no mismatches in the middle of the spacer-protospacer alignments."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "hide_input": true,
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
